# Importing Data

There are multiple publicly available pre-compiled microbiome data sets. These data sets begin after the bioinformatics pipeline and are matrices of counts of OTUs per sample. These data sets can exist as `phyloseq` objects, a popular R package for microbiome analysis [@mcmurdie2013], or as separate tables of counts and metadata.

## Global Patterns

The Global Patterns dataset [@caporaso2011] is an available dataset in the `phyloseq` package. These data contain samples from 25 different environmental samples and mock communities. The sampling depth of these samples averages 3.1 million total counts. We will use this dataset to work through the different normalization methods.

The following lines load the relevant packages and data.

```{r}
library(tidyverse)
library(phyloseq)

data("GlobalPatterns")
# examine phyloseq object 
GlobalPatterns
```

## Pre-processing Quality Control and Filtering

In addition to normalization, there are some steps we can perform that ideally remove technical artifacts from the sequencing process that only introduce noise.

These filtering steps commonly consist of filtering out samples with a low total read depth and filtering out taxa that are rarely abundant.

Let's create a filtered version of the Global Patterns dataset. Note that there are only 26 samples, and all have a large library size, so we will not filter out any samples here.

For taxa filtering, we will remove taxa that appear fewer than 5 times in more than half the samples.

```{r, gp_otu_filter}
# Determine which taxa to remove
filter_taxa <- genefilter_sample(GlobalPatterns,
                                 filterfun_sample(function(x) x > 5),
                                 A=0.5*nsamples(GlobalPatterns))
# Remove those taxa from the GlobalPatterns dataset
# Save as an object with the un-normalized counts
gp_raw <- prune_taxa(filter_taxa, GlobalPatterns)
gp_raw
```

This decreases the number of taxa from 19216 to 219. This is not surprising, because this dataset contains samples from widely different locations (gut, soil, etc), and few taxa are shared among all samples and locations. One potential problem with this approach is the widely different locations, so it is possible that the remaining taxa could be some technical artifact, or could be a general 'core' set of taxa shared across the disparate environments.

Additionally, let us save the total sampling depth as the variable `depth` in the metadata for the Global Patterns dataset.

<!-- (Is this original sampling depth or after filtering? Should be initial yes?) -->

```{r}
gp_raw@sam_data$depth <- sample_sums(gp_raw)
```


We can visualize technical artifacts of sapling depth is by looking at  principal coordinates plots using the Bray-Curtis dissimilarity, coloring by sampling depth too see how much variation can be explained by the original sampling depth. 

```{r, rare_depth_ordcolor}
gp_raw_dist <- phyloseq::ordinate(gp_raw, "PCoA", "bray")
plot_ordination(gp_raw,
                gp_raw_dist,
                color = "depth", 
                title = "PCoA on Raw counts")
```

We don't see any extreme patterns with sampling depth, but this might additionally be due to the differences in different locations might have different sampling depths. This comparison might be more interesting when we only have one location we are sampling from.

<!-- ## HMP -->

<!-- Another dataset, not included in analysis yet since it takes very long to load and run analysis on. Could be a more 'realistic' dataset to use.  -->

<!-- About the HMP dataset.....  -->

<!-- This data comes from healthy subjects.  -->

<!-- Data from the Human Microbiome Project from 16S sequences can most easily be accessed using the `HMP16SData` Bioconductor package.  -->

<!-- ```{r, eval = F} -->

<!-- # BiocManager::install("HMP16SData") -->

<!-- library(HMP16SData) -->

<!-- hmp <- as_phyloseq(V13()) -->

<!-- hmp -->

<!-- V35() -->

<!-- # I was having issues using the V35 dataset, even though it is larger (works on old computer...) -->

<!-- ``` -->

<!-- Now do similar filtering steps  -->

<!-- ```{r, eval = F} -->

<!-- # Save original sampling depth as variable  -->

<!-- hmp@sam_data$sampling_depth <- sample_sums(hmp) -->

<!-- # Filter low prevalence OTUs  -->

<!-- filterTaxaByPrevalence <- function(ps, percentSamplesPresentIn = .1){ -->

<!--   #define threshold -->

<!--   prevalenceThreshold <- percentSamplesPresentIn * nsamples(ps) -->

<!--   #apply a function to all samples that determines if a taxa is fully absent from a particular sample -->

<!--   # change to 1?? -->

<!--   toKeep <- apply(data.frame(otu_table(ps)), 1, -->

<!--                   function(taxa) return(sum(taxa > 0) > prevalenceThreshold)) -->

<!--   #this is placed into toKeep which is a TRUE/FALSE vector of whether or not a that taxa is present in enough samples  -->

<!--   ps_filt <- prune_taxa(toKeep, ps) -->

<!--   return(ps_filt) -->

<!-- } -->

<!-- hmp_raw <- filterTaxaByPrevalence(hmp, .05) -->

<!-- hmp_raw -->

<!-- # Filter by low sampling depth  -->

<!-- hist(sample_sums(hmp)) -->

<!-- summary(hmp_raw@sam_data$sampling_depth) -->

<!-- summary(sample_sums(hmp_raw)) -->

<!-- hmp_raw <- prune_samples(sample_sums(hmp_raw) > 2000,hmp_raw) -->

<!-- hmp_raw -->

<!-- ``` -->
