

# Introduction {#intro}

```{r, packages, include = F, echo = F}
library(tidyverse)
library(patchwork)
library(DESeq2)
library(edgeR)
library(metagenomeSeq)
library(Wrench)
library(GMPR)
```

## The importance of normalization

Microbiome data must be normalized before any analysis can be performed. Following sequencing and assigning raw reads into counts per observed and classified identified taxa classes/OTUs/ASVs, the data is in the form of a matrix of counts of the observed abundance in each sample. Statistical analysis on this count matrix is then performed depending on the goal of the experiment. Common analysis goals include community-level analysis (alpha/beta diversity), differential abundance testing (the parallel of differential expression testing in gene expression studies), and network analysis.

<!-- insert table example of what MB data look like??  -->

Ideally, any analysis of composition, differences, connections, etc. should be done based only on true biological aspects. However, technical variation in counts across samples is a given hurdle that must be accounted for. Biases can arise in the sequencing process, sample preparation, contamination, preferential amplification, and can manifest in differences in sparsity and unequal library sizes [@salter2014]. An effective normalization strategy should put all samples on equal footing so interpretations are on biological signals, not technical signals. Currently, there is no known 'best' normalization method that removes all technical artifacts leaving only biological signals.

Due to the sequencing technology, samples will have different library sizes (also known as sequencing depth), so directly comparing counts between samples is not possible. To illustrate this, consider the counts of one taxon across two samples. In Sample A, this taxon has a count of 230, and in Sample B, this taxon has a count of 23. Is this taxon differentially abundant between samples?

```{r, demo_normalization_importance, echo = F}
sampleA <- c(230,5)
sampleB <- c(23,5)
ASVs <- c("ASV1", "ASV2")

data.frame(sampleA, sampleB, ASVs) %>% 
    pivot_longer(sampleA:sampleB, names_to = "sample", values_to = "count") %>% 
    ggplot(aes(y = count, x = ASVs, fill = sample)) + 
        geom_bar(stat = "identity", position = position_dodge()) + 
    ggtitle("Observed raw counts of 2 ASVS in 2 samples", 
            subtitle = "Is there a difference between samples in ASV1? In ASV2?")

```

Normalization in the context of microbiome data often refers to standardizing sequencing depth across samples. One common approach to this is a scaling-based approach, where a scaling factor is calculated for every sample and then the counts for each taxon in a given sample are divided by its scaling factor. Figure B shows the same data as figure A, but where each sample has been transformed into proportions by dividing by the total counts for each sample. The difference in `ASV1` between samples appears much smaller. However, now there appears to be a difference in `ASV2`, even though the counts were originally the same. This is because in sample B `ASV1` consists of a higher proportion of the total count than in sample A.

This demonstrates the importance of normalization, but also the artifacts that can occur depending on the method.

```{r, echo = F}
data.frame(sampleA = sampleA/sum(sampleA), sampleB = sampleB/sum(sampleB), ASVs) %>% 
    pivot_longer(sampleA:sampleB, names_to = "sample", values_to = "count") %>% 
    ggplot(aes(y = count, x = ASVs, fill = sample)) + 
        geom_bar(stat = "identity", position = position_dodge()) + 
    ggtitle("Observed relative abundance of 2 ASVs in 2 samples", 
            subtitle = "Is there a difference between samples in ASV1? In ASV2?")
```

## The compositional nature of microbiome data

Microbiome data are inherently compositional. This means that the count of each sampled taxa is a portion of a larger whole. Each observed taxa is not independent. As we saw in the above example, before normalization, `ASV2` was equal between samples. After converting to proportions, `ASV2` was no longer equal. With compositional data, each sample has a total sum constraint. Numerous traditional statistical methods rely on an independence assumption, which is not met with microbiome data. This can lead to spurious correlations that exist only because of the compositional nature and not any true signal.

<!-- We do not have knowledge of the absolute abundance. -->

<!-- Consider the library size as fixed.  s-->

With the library size as the sum constraint for each sample, if we know in a biological system that after an event occurs (treatment), ASV1 decreases, this will change the composition of the observed ASV2 regardless of what.

```{r, compositional-plot, echo = F}
sampleA_ASV1x <- rnorm(90)
sampleA_ASV1y <- rnorm(90, sd = 2)
sampleA_ASV2x <- rnorm(40, 4, sd = .5)
sampleA_ASV2y <- rnorm(40, 4, 2)
sampleAx <- c(sampleA_ASV1x,sampleA_ASV2x)
sampleAy <- c(sampleA_ASV1y,sampleA_ASV2y)
asv <- c(rep("red",90), rep("blue",40))

data.frame(sampleAx, sampleAy, asv) %>% 
    ggplot(aes(x = sampleAx, y = sampleAy, color = asv)) + 
    geom_point(size = 3) + 
    ggtitle("Population of two ASVs before treatment") + 
    theme_void() +
    theme(legend.position = "none")

sampleB_index <- sample(1:90, 10)
sampleBx <- c(sampleA_ASV1x[sampleB_index],sampleA_ASV2x)
sampleBy <- c(sampleA_ASV1y[sampleB_index],sampleA_ASV2y)
asv2 <- c(rep("red",10), rep("blue",40))
data.frame(sampleBx, sampleBy, asv2) %>% 
    ggplot(aes(x = sampleBx, y = sampleBy, color = asv2)) + 
    geom_point(size = 3) + 
    ggtitle("Population of two ASVs after treatment") + 
    theme_void() +
    theme(legend.position = "none")

# 
# rbind(data.frame(x = sampleAx, y =  sampleAy, color =  asv, treatment = "Before"),
#       data.frame(x = sampleBx, y = sampleBy, color = asv2, treatment = "After")) %>% 
#     ggplot(aes(x = x, y = y, color = color, group = treatment)) + 
#     geom_point(size = 2) + 
#     facet_wrap(vars(treatment)) + 
#     theme_void() + theme(legend. position = "none")

```

Consider again two samples consisting of red and blue points. We can think of the samples as before and after treatment. In the second plot, the number of blue dots has decreased, but the red remains the same. Consider the following potential observation from this population:

```{r, compositional-tibble, echo = F}
sample <- c("Before", "After")
blue <- c(90, 40)
red <- c(10 , 80)
opts <- options(knitr.kable.NA = "")

tibble(sample, blue, red) %>% 
    knitr::kable(col.names = c("Sample", "blue", "red"))
```

In the after sample, the blue counts have decreased, as we would expect from our knowledge of the population. However, the red counts have increased, even though there is no change in the red population. This observed increase in red counts is due to the compositional nature of the sampled points, and not any true difference in the red population.

### Log ratio

The methodology developed by Aitchison in the 1980s is useful for analyzing compositional data. The theory developed there, however, is for data with different characteristics than microbiome data.

Microbiome data is high dimensional, and.

Taking the logarithm of ratios can be an appropriate transformation for compositional data, so standard statistical tests can be appropriate again. This transformation removes the issue of standardizing/normalizing different library sizes. The library size for a given sample will not distort the biological covariance or correlation structure. 
<!-- (cite Aitchison) -->

<!-- Explain logs, log ratios.  -->

This log-ratio method has a drawback, which is the decision of how to define the denominator. Two approaches to this problem can be to use one sample as the reference. This sample should be 'representative'. The log-ratio transformation is then the ratio of every other taxon to that representative sample. Of course, knowledge of what makes a sample representative is hard to come by and often unknown, and subsequent results can be affected by this choice. This method is frequently called the additive log-ratio approach (alr). 
The alternative method is using the data to create a pseudo-reference sample. This pseudo-reference sample is the geometric mean of the counts of all taxa. This is called the centered log-ratio method. 

<!-- A further method is the isometric log-ratio transofrmation (ilr), which is the product of the clr and the transpose of a matrix which consists of elements that are clr-transformed components of an orthonomal basis. -->
<!-- (REWORD).  -->

While promising, these log-ratio methods have drawbacks in current metagenomic literature. Microbiome data are often incredibly sparse, with up to 80-90% of count matrices containing zero counts. For ratio transformations, if we have sparse data, the geometric mean can be zero. Then the ratio is undefined, and further, so is the log. 

One solution to this is adding a small pseudo count to every element in the data. This removes problems occurring from having zero counts in the data, but there is not a clear best choice of what pseudo count to use, and it the choice can impact downstream results. 

<!-- INCLUDE EXAMPLE?  -->

## Zero-inflation of microbiome data

Include section on considerations that are necessary to account for zero-inflation.

<!-- ## Intro to the phyloseq package -->

<!-- ## Phyloseq wrapper function -->

<!-- Many normalization methods, especially those developed for other disciplines, take the raw matrix as an argument, along with a vector of the grouping variable. For convenience, we have included a wrapper function for phyloseq that ... -->

<!-- ## Conventions for this bookdown -->

<!-- #### Objects -->

<!-- Object names are the abbreviation of the dataset source (ie hmp), followed by the normalization method used, the filtered raw counts have the normalization method "raw". -->

<!-- #### Functions -->

<!-- Function names for normalization use "norm", folowed by an underscore, followed by the name or abbreviation for the method -->

<!-- #### Bio stuff  -->

<!-- ASVs/OTUs/taxa. Refer to everything as taxa.  -->
