# Count Scaling Normalization methods

## Total Sum scaling (TSS)

### About TSS

The first method described is Total Sum Scaling (TSS). This method is also referred to as Total Count (TC), converting into proportions, or relative abundance. This is a scaling method to normalize the different library sizes across samples. For every entry in the count matrix, we scale by the total read depth of that sample. This converts the counts into the proportion of abundance present in each given sample. 

Though a more straightforward method, TSS normalization is not without its drawbacks. In microbiome data, it is common to have numerous low or zero-count observations, and that only a few most common OTUs contribute to the majority of the total sum of the sampling depth. These high-count, frequent, taxa could be an artifact of the sequencing step, where high abundance  observations are preferentially sampled. Using these large counts can dominate the scaling factor for each sample. As seen below, we see that the scaling factor for each sample is completely dominated by `ASV1`, if that one taxon were not included in the sample, the scaling factor would be widely different.

```{r, echo = F}
sample <- c("Sample A", "Sample B")
ASV1 <- c(10314, 824)
ASV2 <- c(34, 23)
ASV3 <- c(8,13)
ASV4 <- c(12, 20)
opts <- options(knitr.kable.NA = "")
scaling_factor <- ASV1 +  ASV2 + ASV3 + ASV4
sf_wo_asv1 <- ASV2 + ASV3 + ASV4

tibble(sample, ASV1, ASV2, ASV3, ASV4, scaling_factor, sf_wo_asv1) %>% 
    knitr::kable(col.names = c("Sample", "ASV1", "ASV2", "ASV3", "ASV4", "TSS Scaling Factor", "Scaling factor w/o ASV1"))
```

Because this method does not account for the preferential sequencing over-abundance of `ASV1` it is possible to see an increase in false positives. However, this is a widely used method, and one of the few normalization methods that completely accounts for differing library sizes, which can be an important consideration depending on the analysis goal. Community level-analysis, for example, can be library-size dependent (ordination, some dissimiliarity measures).   


### TSS R code 

#### Function 
Here, we provide a wrapper function that will normalize a `phyloseq` object by Total Sum Scaling. We have the option of keeping the result as proportions (having values 0-1), or transforming to an equal sequencing depth so the results are counts per million.

```{r, norm_tss_func}
norm_TSS <- function(ps, keep_prop = F){
    # keep as proportions or convert to counts per million?
    scale <- ifelse(keep_prop, 1, 1e6)
    # TSS function
    ps_normed <- phyloseq::transform_sample_counts(ps, function(x) x * scale / sum(x))
    return(ps_normed)
}
```


<!-- To perform the normalization, this method calculates scaling factors.   -->

<!-- ```{r, tss_table} -->

<!-- ``` -->



#### TSS implementation on Global Patterns

Using the above function, we apply this normalization to the Global Patterns data. 

```{r, gp_tss}
gp_tss <- norm_TSS(gp_raw)
# rename the depth as the scaling factor 
sample_data(gp_tss)$scaling_factor <- sample_data(gp_tss)$depth
```

To see the differences between the un-normalized, raw data, and the TSS transformed normalized data, one possible way is to look at ordination plots. Microbiome data are high dimensional, so visualization directly of the data is difficult. Here, let us examine the principal coordinates plot using the Bray-Curtis dissimilarity. 

First calculate the distance matrices, using the `phyloseq` function `ordinate()`

```{r, ordinate_tss}
gp_raw_dist <- phyloseq::ordinate(gp_raw, "PCoA", "bray")
gp_tss_dist <- phyloseq::ordinate(gp_tss, "PCoA", "bray")
```

Now plot the two ordinations. Even before normalization, the different communities are clearly clustered. 

(Note to self: perhaps choose a different dataset to use for walk-through, currently using Global Patterns since it is small and quick for computations, but harder to see differences.)

```{r, plot_ordinate_tss}
plot_ordination(gp_raw, gp_raw_dist, color = "SampleType", 
                title = "PCoA on Raw Counts") + 
plot_ordination(gp_tss, gp_tss_dist, color = "SampleType", 
                title = "PCoA on TSS normalized counts") +
  plot_layout(guides = 'collect')
```


We can also compare the values of the distance matrices before and after normalization to see how the normalization method is impacting different types of points.

```{r}
# Function to visualize potential differences and changes after normalization methods
plot_norm_changes <- function(data_norm, data_raw, dist_method = "bray", x_lab = "raw", y_lab = "norm", title = "Plot title"){
    # calculate the Bray-Cutris distance matrix for the raw data, the normalized data, 
    # and calculate the pairwise difference between the original library sizes between samples
    plot <- data.frame(raw = as.numeric(phyloseq::distance(data_raw, dist_method, na.rm = T)), 
                       norm = as.numeric(phyloseq::distance(data_norm, dist_method, na.rm = T)), 
                       diff = as.numeric(dist(get_variable(data_raw, "depth")))) %>% 
    ggplot(aes(x = raw, y = norm, color = diff)) + 
        geom_point() + 
        geom_abline() + 
        ggtitle(title) +
        xlab(x_lab) + ylab(y_lab) +
        labs(color = "Pairwise difference in sampling depth")+
      xlim(c(0,1)) + ylim(c(0,1))
    return(plot)
}

plot_norm_changes(gp_tss, gp_raw,
                  x_lab = "Raw counts", y_lab = "TSS props", 
                  title = "Distance metric comparision between Raw counts and TSS normalized counts ")
    
```

Points below the line are pairs of samples that are marked as more similar after normalization. Points above the line are marked as more different after normalization. Values closer to 1 are 'more different'. Unsurprisingly, pairs that had larger original differences in sampling depth were marked as more different on the raw, un-normalized data, and became marked as more similar after TSS normalization. 

#### TSS on Kostic data


```{r}
k_tss <- norm_TSS(k_raw)
k_raw_dist <- phyloseq::ordinate(k_raw, "PCoA", "bray")
k_tss_dist <- phyloseq::ordinate(k_tss, "PCoA", "bray")

# Ordination with Diagnosis as color 
plot_ordination(k_raw, k_raw_dist, color = "DIAGNOSIS", 
                title = "PCoA on Raw Counts") + 
plot_ordination(k_tss, k_tss_dist, color = "DIAGNOSIS", 
                title = "PCoA on TSS normalized counts") +
  plot_layout(guides = 'collect')

# Ordination with original sampling depth as color
plot_ordination(k_raw, k_raw_dist, color = "depth", 
                title = "PCoA on Raw Counts") + 
plot_ordination(k_tss, k_tss_dist, color = "depth", 
                title = "PCoA on TSS normalized counts") +
  plot_layout(guides = 'collect')

plot_norm_changes(k_tss, k_raw,
                  x_lab = "Raw counts", y_lab = "TSS props", 
                  title = "Distance metric comparision between Raw counts and TSS normalized counts ")
```




<!-- ## HMP TSS -->

<!-- Normalize the HMP data by TSS  -->

<!-- ```{r, eval = F} -->
<!-- hmp_tss <- norm_TSS(hmp_raw) -->
<!-- ``` -->

<!-- Look at ordination plots  -->

<!-- ```{r, eval = F} -->
<!-- hmp_raw_ord <- phyloseq::ordinate(hmp_raw, "PCoA", "bray") -->
<!-- hmp_tss_ord <- phyloseq::ordinate(hmp_tss, "PCoA", "bray") -->
<!-- sample_variables(hmp_raw) -->
<!-- plot_ordination(hmp_raw, -->
<!--                 raw_ord, -->
<!--                 color = "sampling_depth") -->
<!-- plot_ordination(hmp_tss, -->
<!--                 hmp_tss_ord, -->
<!--                 color = "HMP_BODY_SITE") -->
<!-- ``` -->

<!-- Examine how the distance matrices change before/after normalization  -->

<!-- ```{r, eval = F} -->
<!-- plot_norm_changes(hmp_raw, hmp_tss) -->
<!-- ``` -->

<!-- (Harder to see here and takes longer since we have so much more samples. ) -->

## Cumulative sum scaling (CSS)

### About CSS

Cumulative sum scaling is a scaling normalization method, developed for marker gene sequencing, It is intended to account for undersampling and correct biases from preferentially amplified features in a sample-specific manner [@paulson2013]. This method assumes that count distributions should be roughly equivalent and independent to each other up to the given quantile which is chosen as the smallest value at which instability is found. If there is high count variability the assumption may not be met. This method is an extension to UQ scaling normalization where one quantile is specified for all samples. In contrast to Total Sum scaling where the total sum is used as the scaling factor, the scaling factor for CSS is the sum of the counts up to the chosen quantile.  This is not a method that accounts for compositionality. CSS normalization initially showed improvements in separating samples biologically in ordination, it was shown to be an artifact of unequal application of log transformation across methods [@costea2014].


### CSS R code

#### Function 

```{r}
norm_CSS <- function(ps){
  require(metagenomeSeq)
  # Convert to metagenomeSeq data type
  ps.metaG<-phyloseq_to_metagenomeSeq(ps)
  p_stat = cumNormStatFast(ps.metaG)
  ps.metaG = cumNorm(ps.metaG, p = p_stat)
  ps.metaG.norm <- MRcounts(ps.metaG, norm = T)
  # Convert back to phyloseq with normalized counts
  otu <- otu_table(ps.metaG.norm, taxa_are_rows = T)
  sam <- access(ps, "sam_data")
  sam$scaling_factor <- normFactors(ps.metaG)/1e6
  tax <- access(ps, "tax_table")
  phy <- access(ps, "phy_tree")
  ps_CSS <- phyloseq(otu,sam,tax,phy)
  return(ps_CSS)
}
```

#### CSS on Global Patterns 

Perform CSS normalization: 
```{r}
gp_css <- norm_CSS(gp_raw)
```

View how TMM normalization changes distance metrics differently than raw counts. 
```{r}
plot_norm_changes(gp_css, gp_raw, 
                  x_lab = "Raw", y_lab = "CSS",
                  title = "Distance metric comparision between CSS normalization and Raw counts")

plot_norm_changes(gp_css, gp_tss, 
                  x_lab = "TSS", y_lab = "CSS",
                  title = "Distance metric comparision between CSS normalization and TSS normalization")

```

CSS normalization appears to consider pairs as more different than TSS normalization, and pairs with high sequencing depth differences even more so. 

#### CSS on Kostic data

```{r}
k_css <- norm_CSS(k_raw)
plot_ordination(
    k_css,
    phyloseq::ordinate(k_css, "PCoA", "bray") ,
    color = "DIAGNOSIS",
    title = "PCoA on Raw data"
) /
    plot_ordination(
        k_css,
        phyloseq::ordinate(k_css, "PCoA", "bray") ,
        color = "depth",
        title = "PCoA on Raw data"
    )
plot_norm_changes(k_css, k_raw, 
                  x_lab = "Raw", y_lab = "CSS",
                  title = "Distance metric comparision between CSS normalization and Raw counts")


```


