# Wrench

Wrench is a recent normalization method developed for microbiome data [@kumar2018] that uses an empirical Bayes Normalization approach. This method includes compositional bias correction for sparse datasets. This method uses a hurdle log-normal distribution to estimate the normalization factors (the location estimate for the group). For this method, we assume abundances are drawn from a hurdle Log-Gaussian distribution, and the scaling factor used is essentially the location estimate for the group.

<!-- When a differential abundance analysis is performed on this data, fold changes of null features, those not differentially abundant in the absolute scale, are intimately tied to those of features that are perturbed in their absolute abundances, making the former appear differentially abundant.                                           -->

<!-- "In the absence of such control features, effective correc- tion for compositional bias can still be hoped for, as it can be shown that this correction amounts to resolv- ing a linear technical bias [13]. This fact allows one to exploit several widely used non- spike-in normalization approaches [13, 15–17], which approximate the afore- mentioned spike-in strategy by assuming that most fea- tures do not change on average across samples/conditions. For the same reason, such an interpretation can also be given to approaches like centered logarithmic transforms (CLR) from the theory of compositional data, which many analysts favor when working with relative abundances [18–24]. In this paper, we analyze the behavior of these existing scaling normalization techniques in light of com- positional bias. This method develops a compositional bias correction technique for sparse count data based on an empirical Bayes approach that borrows information across features and samples. The scaling factor is the average of ratios of relative abundances across taxa." -->

<!-- ## Wrench Implementation -->

```{r, message = F}
norm_wrench <- function(ps, group_col){
    require(Wrench, quietly = T)
    if( identical(all.equal(length(group_col), 1), TRUE) & nsamples(ps) > 1 ){
        # Assume that group was a sample variable name (must be categorical)
        group = get_variable(ps, group_col)
    }
    otu_tab <- otu_table(ps)
    W <- wrench(otu_tab, group)
    
    compositionalFactors <- W$ccf
    normalizationFactors <- W$nf
    
    normed_otu <- otu_tab/(normalizationFactors/1e6)
    otu <- otu_table(normed_otu, taxa_are_rows = T)
    sam <- access(ps, "sam_data")
    sam$scaling_factor <- normalizationFactors
    tax <- access(ps, "tax_table")
    phy <- access(ps, "phy_tree")
    ps_wrench <- phyloseq(otu,sam,tax,phy)
    
    return(ps_wrench)
}


```

<!-- ## Wrench on Global Patterns -->

<!-- ```{r, eval = F} -->
<!-- gp_wrench <- norm_wrench(gp_raw, group_col = "SampleType") -->
<!-- ``` -->

<!-- ```{r, message = F, results = "hide", eval = F, echo = F} -->
<!-- gp_wrench <- norm_wrench(gp_raw, group_col = "SampleType") -->
<!-- # warnings, authors suggest robust averaging  -->
<!-- # Is this because we have such distinct groups?  -->
<!-- ``` -->

<!-- # ```{r} -->

<!-- # norm_ms_wrench <- function(ps, group_col){ -->

<!-- #   require(metagenomeSeq, quietly = T) -->

<!-- #   browser() -->

<!-- #   ps.metaG <- phyloseq_to_metagenomeSeq(ps) -->

<!-- #   p_stat = cumNormStatFast(ps.metaG) -->

<!-- #   ps.metaG = cumNorm(ps.metaG, p = p_stat) -->

<!-- #    -->

<!-- #   if( identical(all.equal(length(group_col), 1), TRUE) & nsamples(ps) > 1 ){ -->

<!-- #         # Assume that group was a sample variable name (must be categorical) -->

<!-- #         group = get_variable(ps, group_col) -->

<!-- #     } -->

<!-- #    -->

<!-- #   res_wrench <- wrenchNorm(ps.metaG, condition = group) -->

<!-- #    -->

<!-- #   ps.metaG.norm <- MRcounts(res_wrench, norm = T) -->

<!-- #  -->

<!-- #   otu <- otu_table(ps.metaG.norm, taxa_are_rows = T) -->

<!-- #   sam <- access(ps, "sam_data") -->

<!-- #   sam$scaling_factor <- normFactors(ps.metaG)/1e6 -->

<!-- #   tax <- access(ps, "tax_table") -->

<!-- #   phy <- access(ps, "phy_tree") -->

<!-- #   ps_CSS <- phyloseq(otu,sam,tax,phy) -->

<!-- #   return(ps_CSS) -->

<!-- # } -->

<!-- # ``` -->
