<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 DESeq | Tutorial on Microbiome Normalization</title>
  <meta name="description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 DESeq | Tutorial on Microbiome Normalization" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 DESeq | Tutorial on Microbiome Normalization" />
  
  <meta name="twitter:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

<meta name="author" content="Emily Palmer" />


<meta name="date" content="2021-08-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="rarefying.html"/>
<link rel="next" href="edger-tmm.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">MB Normalization</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#the-importance-of-normalization"><i class="fa fa-check"></i><b>1.1</b> The importance of normalization</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-compositional-nature-of-microbiome-data"><i class="fa fa-check"></i><b>1.2</b> The compositional nature of microbiome data</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#log-ratio"><i class="fa fa-check"></i><b>1.2.1</b> Log ratio</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#zero-inflation-of-microbiome-data"><i class="fa fa-check"></i><b>1.3</b> Zero-inflation of microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>2</b> Importing Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-data.html"><a href="importing-data.html#global-patterns"><i class="fa fa-check"></i><b>2.1</b> Global Patterns</a></li>
<li class="chapter" data-level="2.2" data-path="importing-data.html"><a href="importing-data.html#pre-procesing-quality-control-and-filtering"><i class="fa fa-check"></i><b>2.2</b> Pre-procesing Quality Control and Filtering</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html"><i class="fa fa-check"></i><b>3</b> Total Sum scaling (TSS)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#about-tss"><i class="fa fa-check"></i><b>3.1</b> About TSS</a></li>
<li class="chapter" data-level="3.2" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-implementation"><i class="fa fa-check"></i><b>3.2</b> TSS Implementation</a></li>
<li class="chapter" data-level="3.3" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-on-global-patterns"><i class="fa fa-check"></i><b>3.3</b> TSS on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="rarefying.html"><a href="rarefying.html"><i class="fa fa-check"></i><b>4</b> Rarefying</a>
<ul>
<li class="chapter" data-level="4.1" data-path="rarefying.html"><a href="rarefying.html#rarefying-on-global-patterns"><i class="fa fa-check"></i><b>4.1</b> Rarefying on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="deseq.html"><a href="deseq.html"><i class="fa fa-check"></i><b>5</b> DESeq</a>
<ul>
<li class="chapter" data-level="5.1" data-path="deseq.html"><a href="deseq.html#about-deseq"><i class="fa fa-check"></i><b>5.1</b> About DESeq</a></li>
<li class="chapter" data-level="5.2" data-path="deseq.html"><a href="deseq.html#deseq-implementation"><i class="fa fa-check"></i><b>5.2</b> DESeq Implementation</a></li>
<li class="chapter" data-level="5.3" data-path="deseq.html"><a href="deseq.html#deseq-on-global-patterns"><i class="fa fa-check"></i><b>5.3</b> DESeq on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="edger-tmm.html"><a href="edger-tmm.html"><i class="fa fa-check"></i><b>6</b> edgeR TMM</a>
<ul>
<li class="chapter" data-level="6.1" data-path="edger-tmm.html"><a href="edger-tmm.html#edger-tmm-implementatin"><i class="fa fa-check"></i><b>6.1</b> EdgeR TMM implementatin</a></li>
<li class="chapter" data-level="6.2" data-path="edger-tmm.html"><a href="edger-tmm.html#tmm-on-global-patterns"><i class="fa fa-check"></i><b>6.2</b> TMM on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html"><i class="fa fa-check"></i><b>7</b> Cumulative sum scaling (CSS)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-on-global-patterns"><i class="fa fa-check"></i><b>7.1</b> CSS on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="gmpr.html"><a href="gmpr.html"><i class="fa fa-check"></i><b>8</b> GMPR</a>
<ul>
<li class="chapter" data-level="8.1" data-path="gmpr.html"><a href="gmpr.html#global-patterns-gmpr"><i class="fa fa-check"></i><b>8.1</b> Global Patterns GMPR</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="wrench.html"><a href="wrench.html"><i class="fa fa-check"></i><b>9</b> Wrench</a>
<ul>
<li class="chapter" data-level="9.1" data-path="wrench.html"><a href="wrench.html#wrench-implementation"><i class="fa fa-check"></i><b>9.1</b> Wrench Implementation</a></li>
<li class="chapter" data-level="9.2" data-path="wrench.html"><a href="wrench.html#wrench-on-global-patterns"><i class="fa fa-check"></i><b>9.2</b> Wrench on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="log-ratio-approaches.html"><a href="log-ratio-approaches.html"><i class="fa fa-check"></i><b>10</b> Log-Ratio Approaches</a></li>
<li class="chapter" data-level="11" data-path="comparisions.html"><a href="comparisions.html"><i class="fa fa-check"></i><b>11</b> Comparisions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial on Microbiome Normalization</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="deseq" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> DESeq</h1>
<div id="about-deseq" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> About DESeq</h2>
<p>DESeq is a method for adjusting for differing library sizes. This method also can account for differences in library composition. This method has been called (MED) or (RLE) in the literature.</p>
<p>DESeq2 first takes the log (natural logarithm) of every entry in the count matrix. Due to this, all entries with zero will be set to negative infinity. Next, the row average is calculated (geometric average), so we have a vector of average counts for each taxon. Taking the log first should avoid undue influence by extreme outliers. All taxa with an average of infinity are removed. This step will remove all taxa with zero read count in one or more samples. This can be a problem in microbiome data.
Next, we subtract the average log value from the log(counts), this gives a log ratio. This is equivalent to the ratio of the reads in each sample to the average across all samples.
Next, we calculate the median of the log-ratios for each sample. These medians are converted to scaling factors for each sample by exponentiation.
An extension of this method, denoted ‘poscounts’, has been suggested, which instead of taking the geometric mean of the logged counts for each taxon, we take the n-th root of the product of the non-zero counts.</p>
<p>This method assumes that the taxon of median absolute abundance is not differentially abundant, which is more likely true for the RNA-Seq it was developed for, but may not be true for microbiome studies, especially when there are more study groups, or we are analyzing higher taxonomic levels.</p>
<p>An additional option can be used to perform a variance stabilizing transformation on the count matrix before normalizing with the above size factors. This method calculates a dispersion-mean relationship and then transforms the data. The result ideally is an abundance matrix that is approximately homoskedastic or constant variance across the range of mean values. The package also includes an option for a ‘rlog’ transform, which they recommend over the variance stabilizing method in the case when there is a large difference in library sizes.</p>
<p>If differential abundance is of interest to calculate, DESeq uses a negative binomial distribution to model differential abundances.
It is possible to provide the size factors calculated by another method to DESeq to perform differential analysis.</p>
</div>
<div id="deseq-implementation" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> DESeq Implementation</h2>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="deseq.html#cb35-1"></a>norm_DESeq_RLE_poscounts &lt;-<span class="st"> </span><span class="cf">function</span>(ps, <span class="dt">group =</span> <span class="dv">1</span>){</span>
<span id="cb35-2"><a href="deseq.html#cb35-2"></a>  <span class="kw">require</span>(DESeq2, <span class="dt">quietly =</span> T)</span>
<span id="cb35-3"><a href="deseq.html#cb35-3"></a>    <span class="co"># keep arbitrary design for normalization </span></span>
<span id="cb35-4"><a href="deseq.html#cb35-4"></a>    <span class="co"># Convert to DESeq object</span></span>
<span id="cb35-5"><a href="deseq.html#cb35-5"></a>    ps_dds &lt;-<span class="st"> </span><span class="kw">phyloseq_to_deseq2</span>(ps, <span class="op">~</span><span class="dv">1</span>)</span>
<span id="cb35-6"><a href="deseq.html#cb35-6"></a>    <span class="co"># Calculate the size factors (scaling)</span></span>
<span id="cb35-7"><a href="deseq.html#cb35-7"></a>    ps_dds &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(ps_dds, <span class="dt">type =</span> <span class="st">&quot;poscounts&quot;</span>)</span>
<span id="cb35-8"><a href="deseq.html#cb35-8"></a>    <span class="co"># Extract counts</span></span>
<span id="cb35-9"><a href="deseq.html#cb35-9"></a>    counts &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">counts</span>(ps_dds, <span class="dt">normalized =</span> T)</span>
<span id="cb35-10"><a href="deseq.html#cb35-10"></a>    <span class="co"># Convert back to phyloseq</span></span>
<span id="cb35-11"><a href="deseq.html#cb35-11"></a>    otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(counts, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb35-12"><a href="deseq.html#cb35-12"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb35-13"><a href="deseq.html#cb35-13"></a>    sam<span class="op">$</span>scaling_factor &lt;-<span class="st"> </span><span class="kw">sizeFactors</span>(ps_dds)</span>
<span id="cb35-14"><a href="deseq.html#cb35-14"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb35-15"><a href="deseq.html#cb35-15"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb35-16"><a href="deseq.html#cb35-16"></a>    ps_DESeq &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu,sam,tax,phy)</span>
<span id="cb35-17"><a href="deseq.html#cb35-17"></a>    <span class="kw">return</span>(ps_DESeq)</span>
<span id="cb35-18"><a href="deseq.html#cb35-18"></a>}</span>
<span id="cb35-19"><a href="deseq.html#cb35-19"></a></span>
<span id="cb35-20"><a href="deseq.html#cb35-20"></a></span>
<span id="cb35-21"><a href="deseq.html#cb35-21"></a>norm_DESeq_vs &lt;-<span class="st"> </span><span class="cf">function</span>(ps, <span class="dt">group =</span> <span class="dv">1</span>){</span>
<span id="cb35-22"><a href="deseq.html#cb35-22"></a>  <span class="kw">require</span>(DESeq2, <span class="dt">quietly =</span> T)</span>
<span id="cb35-23"><a href="deseq.html#cb35-23"></a>  ps_dds &lt;-<span class="st"> </span><span class="kw">phyloseq_to_deseq2</span>(ps, <span class="op">~</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb35-24"><a href="deseq.html#cb35-24"></a>  ps_dds &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(ps_dds, <span class="dt">type =</span> <span class="st">&quot;poscounts&quot;</span>)</span>
<span id="cb35-25"><a href="deseq.html#cb35-25"></a>  <span class="co"># Variance transformation</span></span>
<span id="cb35-26"><a href="deseq.html#cb35-26"></a>  ps_dds &lt;-<span class="st"> </span><span class="kw">estimateDispersions</span>(ps_dds)</span>
<span id="cb35-27"><a href="deseq.html#cb35-27"></a>  abund &lt;-<span class="st"> </span><span class="kw">getVarianceStabilizedData</span>(ps_dds)</span>
<span id="cb35-28"><a href="deseq.html#cb35-28"></a>  <span class="co"># don’t allow deseq to return negative counts</span></span>
<span id="cb35-29"><a href="deseq.html#cb35-29"></a>  <span class="co"># add the minimum count to all values</span></span>
<span id="cb35-30"><a href="deseq.html#cb35-30"></a>  <span class="co"># another option is to replace negative counts with 0</span></span>
<span id="cb35-31"><a href="deseq.html#cb35-31"></a>  abund &lt;-<span class="st"> </span>abund <span class="op">+</span><span class="st"> </span><span class="kw">abs</span>(<span class="kw">min</span>(abund)) </span>
<span id="cb35-32"><a href="deseq.html#cb35-32"></a>  otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(abund, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb35-33"><a href="deseq.html#cb35-33"></a>  sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb35-34"><a href="deseq.html#cb35-34"></a>  tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb35-35"><a href="deseq.html#cb35-35"></a>  phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb35-36"><a href="deseq.html#cb35-36"></a>  ps_DESeq &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu,sam,tax,phy)</span>
<span id="cb35-37"><a href="deseq.html#cb35-37"></a>  <span class="kw">return</span>(ps_DESeq)</span>
<span id="cb35-38"><a href="deseq.html#cb35-38"></a>}</span></code></pre></div>
</div>
<div id="deseq-on-global-patterns" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> DESeq on Global Patterns</h2>
<p>Perform DESeq RLE normalization as well as DESeq variance stabalised transformation.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="deseq.html#cb36-1"></a>gp_deseq_rle &lt;-<span class="st"> </span><span class="kw">norm_DESeq_RLE_poscounts</span>(gp_raw)</span></code></pre></div>
<pre><code>## converting counts to integer mode</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="deseq.html#cb38-1"></a>gp_deseq_vs &lt;-<span class="st"> </span><span class="kw">norm_DESeq_vs</span>(gp_raw)</span></code></pre></div>
<pre><code>## converting counts to integer mode</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## -- note: fitType=&#39;parametric&#39;, but the dispersion trend was not well captured by the
##    function: y = a/x + b, and a local regression fit was automatically substituted.
##    specify fitType=&#39;local&#39; or &#39;mean&#39; to avoid this message next time.</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<p>Examine principal coordinate plots between raw data and DESeq normalized data.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="deseq.html#cb44-1"></a>gp_raw_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_raw, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) </span>
<span id="cb44-2"><a href="deseq.html#cb44-2"></a>gp_rle_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_deseq_rle, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) </span>
<span id="cb44-3"><a href="deseq.html#cb44-3"></a>gp_vs_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_deseq_vs, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) </span>
<span id="cb44-4"><a href="deseq.html#cb44-4"></a></span>
<span id="cb44-5"><a href="deseq.html#cb44-5"></a><span class="kw">plot_ordination</span>(gp_raw, gp_raw_dist, <span class="dt">color =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;Raw&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb44-6"><a href="deseq.html#cb44-6"></a><span class="kw">plot_ordination</span>(gp_deseq_rle, gp_rle_dist, <span class="dt">color =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;RLE&quot;</span>) <span class="op">+</span></span>
<span id="cb44-7"><a href="deseq.html#cb44-7"></a><span class="kw">plot_ordination</span>(gp_deseq_vs, gp_vs_dist, <span class="dt">color =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;RLE&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb44-8"><a href="deseq.html#cb44-8"></a><span class="st">  </span><span class="kw">plot_layout</span>(<span class="dt">guides =</span> <span class="st">&#39;collect&#39;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/deseq-ordination-1.png" width="672" /></p>
<p>See how dissimilarity matrices differ from raw count dissimilarity matrices.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="deseq.html#cb45-1"></a><span class="kw">plot_norm_changes</span>(gp_deseq_rle, gp_raw)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="deseq.html#cb46-1"></a><span class="kw">plot_norm_changes</span>(gp_deseq_vs, gp_raw)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-6-2.png" width="672" /></p>
<!-- Look at differences between scaling factors -->
<!-- do we need to exponentiate?  -->
<!-- ```{r} -->
<!-- data.frame(deseq = sample_data(gp_deseq_rle)$depth/sample_data(gp_deseq_rle)$scaling_factor, -->
<!--            tss = sample_data(gp_tss)$scaling_factor) %>%  -->
<!--   ggplot(aes(x = deseq, y = tss)) + geom_point() -->
<!-- ``` -->

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="rarefying.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="edger-tmm.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MB_normalization_review.pdf", "MB_normalization_review.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
