<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Importing Data | Tutorial on Normalization of Microbiome Data</title>
  <meta name="description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Importing Data | Tutorial on Normalization of Microbiome Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Importing Data | Tutorial on Normalization of Microbiome Data" />
  
  <meta name="twitter:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="total-sum-scaling-tss.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">MB Normalization</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#the-importance-of-normalization"><i class="fa fa-check"></i><b>1.1</b> The importance of normalization</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-compositional-nature-of-microbiome-data"><i class="fa fa-check"></i><b>1.2</b> The compositional nature of microbiome data</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#log-ratio-methods"><i class="fa fa-check"></i><b>1.2.1</b> Log ratio methods</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>2</b> Importing Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-data.html"><a href="importing-data.html#global-patterns"><i class="fa fa-check"></i><b>2.1</b> Global Patterns</a></li>
<li class="chapter" data-level="2.2" data-path="importing-data.html"><a href="importing-data.html#pre-processing-quality-control-and-filtering"><i class="fa fa-check"></i><b>2.2</b> Pre-processing Quality Control and Filtering</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html"><i class="fa fa-check"></i><b>3</b> Total Sum scaling (TSS)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#about-tss"><i class="fa fa-check"></i><b>3.1</b> About TSS</a></li>
<li class="chapter" data-level="3.2" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-implementation"><i class="fa fa-check"></i><b>3.2</b> TSS Implementation</a></li>
<li class="chapter" data-level="3.3" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-on-global-patterns"><i class="fa fa-check"></i><b>3.3</b> TSS on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="rarefying.html"><a href="rarefying.html"><i class="fa fa-check"></i><b>4</b> Rarefying</a>
<ul>
<li class="chapter" data-level="4.1" data-path="rarefying.html"><a href="rarefying.html#about-rarefying"><i class="fa fa-check"></i><b>4.1</b> About Rarefying</a></li>
<li class="chapter" data-level="4.2" data-path="rarefying.html"><a href="rarefying.html#rarefying-implementation"><i class="fa fa-check"></i><b>4.2</b> Rarefying implementation</a></li>
<li class="chapter" data-level="4.3" data-path="rarefying.html"><a href="rarefying.html#rarefying-on-global-patterns"><i class="fa fa-check"></i><b>4.3</b> Rarefying on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="deseq.html"><a href="deseq.html"><i class="fa fa-check"></i><b>5</b> DESeq</a>
<ul>
<li class="chapter" data-level="5.1" data-path="deseq.html"><a href="deseq.html#about-deseq"><i class="fa fa-check"></i><b>5.1</b> About DESeq</a></li>
<li class="chapter" data-level="5.2" data-path="deseq.html"><a href="deseq.html#deseq-implementation"><i class="fa fa-check"></i><b>5.2</b> DESeq Implementation</a></li>
<li class="chapter" data-level="5.3" data-path="deseq.html"><a href="deseq.html#deseq-on-global-patterns"><i class="fa fa-check"></i><b>5.3</b> DESeq on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tmm-edger.html"><a href="tmm-edger.html"><i class="fa fa-check"></i><b>6</b> TMM (edgeR)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="tmm-edger.html"><a href="tmm-edger.html#edger-tmm-implementation"><i class="fa fa-check"></i><b>6.1</b> EdgeR TMM implementation</a></li>
<li class="chapter" data-level="6.2" data-path="tmm-edger.html"><a href="tmm-edger.html#tmm-on-global-patterns"><i class="fa fa-check"></i><b>6.2</b> TMM on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html"><i class="fa fa-check"></i><b>7</b> Cumulative sum scaling (CSS)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-implementation"><i class="fa fa-check"></i><b>7.1</b> CSS implementation</a></li>
<li class="chapter" data-level="7.2" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-on-global-patterns"><i class="fa fa-check"></i><b>7.2</b> CSS on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="gmpr.html"><a href="gmpr.html"><i class="fa fa-check"></i><b>8</b> GMPR</a>
<ul>
<li class="chapter" data-level="8.1" data-path="gmpr.html"><a href="gmpr.html#gmpr-implementation"><i class="fa fa-check"></i><b>8.1</b> GMPR Implementation</a></li>
<li class="chapter" data-level="8.2" data-path="gmpr.html"><a href="gmpr.html#global-patterns-gmpr"><i class="fa fa-check"></i><b>8.2</b> Global Patterns GMPR</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="wrench.html"><a href="wrench.html"><i class="fa fa-check"></i><b>9</b> Wrench</a>
<ul>
<li class="chapter" data-level="9.1" data-path="wrench.html"><a href="wrench.html#wrench-implementation"><i class="fa fa-check"></i><b>9.1</b> Wrench Implementation</a></li>
<li class="chapter" data-level="9.2" data-path="wrench.html"><a href="wrench.html#wrench-on-global-patterns"><i class="fa fa-check"></i><b>9.2</b> Wrench on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="log-ratio-approaches.html"><a href="log-ratio-approaches.html"><i class="fa fa-check"></i><b>10</b> Log-Ratio Approaches</a></li>
<li class="chapter" data-level="11" data-path="comparisions.html"><a href="comparisions.html"><i class="fa fa-check"></i><b>11</b> Comparisions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial on Normalization of Microbiome Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="importing-data" class="section level1" number="2">
<h1><span class="header-section-number">Chapter 2</span> Importing Data</h1>
<p>There are multiple publicly available pre-compiled microbiome data sets. These data sets begin after the bioinformatics pipeline and are matrices of counts of OTUs per sample. These data sets can exist as <code>phyloseq</code> objects, a popular R package for microbiome analysis <span class="citation">(McMurdie and Holmes <a href="#ref-mcmurdie2013" role="doc-biblioref">2013</a>)</span>, or as separate tables of counts and metadata.</p>
<div id="global-patterns" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Global Patterns</h2>
<p>The Global Patterns dataset <span class="citation">(Caporaso et al. <a href="#ref-caporaso2011" role="doc-biblioref">2011</a>)</span> is an available dataset in the <code>phyloseq</code> package. These data contain samples from 25 different environmental samples and mock communities. The sampling depth of these samples averages 3.1 million total counts. We will use this dataset to work through the different normalization methods.</p>
<p>The following lines load the relevant packages and data.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="importing-data.html#cb1-1"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="importing-data.html#cb1-2"></a><span class="kw">library</span>(phyloseq)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;phyloseq&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:SummarizedExperiment&#39;:
## 
##     distance</code></pre>
<pre><code>## The following object is masked from &#39;package:Biobase&#39;:
## 
##     sampleNames</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomicRanges&#39;:
## 
##     distance</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     distance</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="importing-data.html#cb7-1"></a><span class="kw">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>)</span>
<span id="cb7-2"><a href="importing-data.html#cb7-2"></a><span class="co"># examine phyloseq object </span></span>
<span id="cb7-3"><a href="importing-data.html#cb7-3"></a>GlobalPatterns</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
</div>
<div id="pre-processing-quality-control-and-filtering" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Pre-processing Quality Control and Filtering</h2>
<p>In addition to normalization, there are some steps we can perform that ideally remove technical artifacts from the sequencing process that only introduce noise.</p>
<p>These filtering steps commonly consist of filtering out samples with a low total read depth and filtering out taxa that are rarely abundant.</p>
<p>Let’s create a filtered version of the Global Patterns dataset. Note that there are only 26 samples, and all have a large library size, so we will not filter out any samples here.</p>
<p>For taxa filtering, we will remove taxa that appear fewer than 5 times in more than half the samples.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="importing-data.html#cb9-1"></a><span class="co"># Determine which taxa to remove</span></span>
<span id="cb9-2"><a href="importing-data.html#cb9-2"></a>filter_taxa &lt;-<span class="st"> </span><span class="kw">genefilter_sample</span>(GlobalPatterns,</span>
<span id="cb9-3"><a href="importing-data.html#cb9-3"></a>                                 <span class="kw">filterfun_sample</span>(<span class="cf">function</span>(x) x <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>),</span>
<span id="cb9-4"><a href="importing-data.html#cb9-4"></a>                                 <span class="dt">A=</span><span class="fl">0.5</span><span class="op">*</span><span class="kw">nsamples</span>(GlobalPatterns))</span>
<span id="cb9-5"><a href="importing-data.html#cb9-5"></a><span class="co"># Remove those taxa from the GlobalPatterns dataset</span></span>
<span id="cb9-6"><a href="importing-data.html#cb9-6"></a><span class="co"># Save as an object with the un-normalized counts</span></span>
<span id="cb9-7"><a href="importing-data.html#cb9-7"></a>gp_raw &lt;-<span class="st"> </span><span class="kw">prune_taxa</span>(filter_taxa, GlobalPatterns)</span>
<span id="cb9-8"><a href="importing-data.html#cb9-8"></a>gp_raw</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 219 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 219 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 219 tips and 218 internal nodes ]</code></pre>
<p>This decreases the number of taxa from 19216 to 219. This is not surprising, because this dataset contains samples from widely different locations (gut, soil, etc), and few taxa are shared among all samples and locations. One potential problem with this approach is the widely different locations, so it is possible that the remaining taxa could be some technical artifact, or could be a general ‘core’ set of taxa shared across the disparate environments.</p>
<p>Additionally, let us save the total sampling depth as the variable <code>depth</code> in the metadata for the Global Patterns dataset.</p>
<!-- (Is this original sampling depth or after filtering? Should be initial yes?) -->
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="importing-data.html#cb11-1"></a>gp_raw<span class="op">@</span>sam_data<span class="op">$</span>depth &lt;-<span class="st"> </span><span class="kw">sample_sums</span>(gp_raw)</span></code></pre></div>
<p>We can visualize technical artifacts of sapling depth is by looking at principal coordinates plots using the Bray-Curtis dissimilarity, coloring by sampling depth too see how much variation can be explained by the original sampling depth.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="importing-data.html#cb12-1"></a>gp_raw_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_raw, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb12-2"><a href="importing-data.html#cb12-2"></a><span class="kw">plot_ordination</span>(gp_raw,</span>
<span id="cb12-3"><a href="importing-data.html#cb12-3"></a>                gp_raw_dist,</span>
<span id="cb12-4"><a href="importing-data.html#cb12-4"></a>                <span class="dt">color =</span> <span class="st">&quot;depth&quot;</span>, </span>
<span id="cb12-5"><a href="importing-data.html#cb12-5"></a>                <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw counts&quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/rare_depth_ordcolor-1.png" width="672" /></p>
<p>We don’t see any extreme patterns with sampling depth, but this might additionally be due to the differences in different locations might have different sampling depths. This comparison might be more interesting when we only have one location we are sampling from.</p>
<!-- ## HMP -->
<!-- Another dataset, not included in analysis yet since it takes very long to load and run analysis on. Could be a more 'realistic' dataset to use.  -->
<!-- About the HMP dataset.....  -->
<!-- This data comes from healthy subjects.  -->
<!-- Data from the Human Microbiome Project from 16S sequences can most easily be accessed using the `HMP16SData` Bioconductor package.  -->
<!-- ```{r, eval = F} -->
<!-- # BiocManager::install("HMP16SData") -->
<!-- library(HMP16SData) -->
<!-- hmp <- as_phyloseq(V13()) -->
<!-- hmp -->
<!-- V35() -->
<!-- # I was having issues using the V35 dataset, even though it is larger (works on old computer...) -->
<!-- ``` -->
<!-- Now do similar filtering steps  -->
<!-- ```{r, eval = F} -->
<!-- # Save original sampling depth as variable  -->
<!-- hmp@sam_data$sampling_depth <- sample_sums(hmp) -->
<!-- # Filter low prevalence OTUs  -->
<!-- filterTaxaByPrevalence <- function(ps, percentSamplesPresentIn = .1){ -->
<!--   #define threshold -->
<!--   prevalenceThreshold <- percentSamplesPresentIn * nsamples(ps) -->
<!--   #apply a function to all samples that determines if a taxa is fully absent from a particular sample -->
<!--   # change to 1?? -->
<!--   toKeep <- apply(data.frame(otu_table(ps)), 1, -->
<!--                   function(taxa) return(sum(taxa > 0) > prevalenceThreshold)) -->
<!--   #this is placed into toKeep which is a TRUE/FALSE vector of whether or not a that taxa is present in enough samples  -->
<!--   ps_filt <- prune_taxa(toKeep, ps) -->
<!--   return(ps_filt) -->
<!-- } -->
<!-- hmp_raw <- filterTaxaByPrevalence(hmp, .05) -->
<!-- hmp_raw -->
<!-- # Filter by low sampling depth  -->
<!-- hist(sample_sums(hmp)) -->
<!-- summary(hmp_raw@sam_data$sampling_depth) -->
<!-- summary(sample_sums(hmp_raw)) -->
<!-- hmp_raw <- prune_samples(sample_sums(hmp_raw) > 2000,hmp_raw) -->
<!-- hmp_raw -->
<!-- ``` -->

</div>
</div>
<h3>References</h3>
<div id="refs" class="references hanging-indent">
<div id="ref-caporaso2011">
<p>Caporaso, J. Gregory, Christian L. Lauber, William A. Walters, Donna Berg-Lyons, Catherine A. Lozupone, Peter J. Turnbaugh, Noah Fierer, and Rob Knight. 2011. “Global Patterns of 16S rRNA Diversity at a Depth of Millions of Sequences Per Sample.” <em>Proceedings of the National Academy of Sciences</em> 108 (Supplement 1): 4516–22. <a href="https://doi.org/10.1073/pnas.1000080107">https://doi.org/10.1073/pnas.1000080107</a>.</p>
</div>
<div id="ref-mcmurdie2013">
<p>McMurdie, Paul J., and Susan Holmes. 2013. “Phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PLOS ONE</em> 8 (4): e61217. <a href="https://doi.org/10.1371/journal.pone.0061217">https://doi.org/10.1371/journal.pone.0061217</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="total-sum-scaling-tss.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MB_normalization_review.pdf", "MB_normalization_review.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
