<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Importing Data | Tutorial on Normalization of Microbiome Data</title>
  <meta name="description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Importing Data | Tutorial on Normalization of Microbiome Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Importing Data | Tutorial on Normalization of Microbiome Data" />
  
  <meta name="twitter:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="count-scaling-normalization-methods.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">MB Normalization</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#the-importance-of-normalization"><i class="fa fa-check"></i><b>1.1</b> The importance of normalization</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-compositional-nature-of-microbiome-data"><i class="fa fa-check"></i><b>1.2</b> The compositional nature of microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>2</b> Importing Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-data.html"><a href="importing-data.html#global-patterns"><i class="fa fa-check"></i><b>2.1</b> Global Patterns</a></li>
<li class="chapter" data-level="2.2" data-path="importing-data.html"><a href="importing-data.html#colorectal-cancer"><i class="fa fa-check"></i><b>2.2</b> Colorectal cancer</a></li>
<li class="chapter" data-level="2.3" data-path="importing-data.html"><a href="importing-data.html#pre-processing-quality-control-and-filtering"><i class="fa fa-check"></i><b>2.3</b> Pre-processing Quality Control and Filtering</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html"><i class="fa fa-check"></i><b>3</b> Count Scaling Normalization methods</a>
<ul>
<li class="chapter" data-level="3.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#total-sum-scaling-tss"><i class="fa fa-check"></i><b>3.1</b> Total Sum scaling (TSS)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#about-tss"><i class="fa fa-check"></i><b>3.1.1</b> About TSS</a></li>
<li class="chapter" data-level="3.1.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#tss-r-code"><i class="fa fa-check"></i><b>3.1.2</b> TSS R code</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#cumulative-sum-scaling-css"><i class="fa fa-check"></i><b>3.2</b> Cumulative sum scaling (CSS)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#about-css"><i class="fa fa-check"></i><b>3.2.1</b> About CSS</a></li>
<li class="chapter" data-level="3.2.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#css-r-code"><i class="fa fa-check"></i><b>3.2.2</b> CSS R code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html"><i class="fa fa-check"></i><b>4</b> Compositionality Aware Scaling Methods</a>
<ul>
<li class="chapter" data-level="4.1" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#deseq"><i class="fa fa-check"></i><b>4.1</b> DESeq</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#about-deseq"><i class="fa fa-check"></i><b>4.1.1</b> About DESeq</a></li>
<li class="chapter" data-level="4.1.2" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#deseq-r-code"><i class="fa fa-check"></i><b>4.1.2</b> DESeq R Code</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#gmpr"><i class="fa fa-check"></i><b>4.2</b> GMPR</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#about-gmpr"><i class="fa fa-check"></i><b>4.2.1</b> About GMPR</a></li>
<li class="chapter" data-level="4.2.2" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#gmpr-walkthough"><i class="fa fa-check"></i><b>4.2.2</b> GMPR Walkthough</a></li>
<li class="chapter" data-level="4.2.3" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#gmpr-r-code"><i class="fa fa-check"></i><b>4.2.3</b> GMPR R Code</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#tmm-edger"><i class="fa fa-check"></i><b>4.3</b> TMM (edgeR)</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#about-tmm"><i class="fa fa-check"></i><b>4.3.1</b> About TMM</a></li>
<li class="chapter" data-level="4.3.2" data-path="compositionality-aware-scaling-methods.html"><a href="compositionality-aware-scaling-methods.html#edger-tmm-t-code"><i class="fa fa-check"></i><b>4.3.2</b> EdgeR TMM T code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="subsampling-methods.html"><a href="subsampling-methods.html"><i class="fa fa-check"></i><b>5</b> Subsampling Methods</a>
<ul>
<li class="chapter" data-level="5.1" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying"><i class="fa fa-check"></i><b>5.1</b> Rarefying</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="subsampling-methods.html"><a href="subsampling-methods.html#about-rarefying"><i class="fa fa-check"></i><b>5.1.1</b> About Rarefying</a></li>
<li class="chapter" data-level="5.1.2" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-walkthrough"><i class="fa fa-check"></i><b>5.1.2</b> Rarefying walkthrough</a></li>
<li class="chapter" data-level="5.1.3" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-r-code"><i class="fa fa-check"></i><b>5.1.3</b> Rarefying R code</a></li>
<li class="chapter" data-level="5.1.4" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-on-kostic-data"><i class="fa fa-check"></i><b>5.1.4</b> Rarefying on Kostic Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="other-methods.html"><a href="other-methods.html"><i class="fa fa-check"></i><b>6</b> Other Methods</a>
<ul>
<li class="chapter" data-level="6.1" data-path="other-methods.html"><a href="other-methods.html#wrench"><i class="fa fa-check"></i><b>6.1</b> Wrench</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html"><i class="fa fa-check"></i><b>7</b> Log Ratio Transformation Methods</a>
<ul>
<li class="chapter" data-level="7.1" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#ancom-ii"><i class="fa fa-check"></i><b>7.1</b> ANCOM II</a></li>
<li class="chapter" data-level="7.2" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#ancom-bc"><i class="fa fa-check"></i><b>7.2</b> ANCOM BC</a></li>
<li class="chapter" data-level="7.3" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#aldex2"><i class="fa fa-check"></i><b>7.3</b> ALDEx2</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial on Normalization of Microbiome Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="importing-data" class="section level1" number="2">
<h1><span class="header-section-number">Chapter 2</span> Importing Data</h1>
<p>There are multiple publicly available pre-compiled microbiome data sets of feature tables. These data sets begin after the bioinformatics pipeline and are matrices of counts of OTUs per sample. These data sets can exist as <code>phyloseq</code> objects, a popular R package for microbiome analysis <span class="citation">(McMurdie and Holmes <a href="#ref-mcmurdie2013" role="doc-biblioref">2013</a>)</span>, or as separate tables of feature counts and metadata. For this tutorial, we will provide code that assumes a <code>phyloseq</code> object as input, and outputs a normalized <code>phyloseq</code> object. Converting between separate tables and <code>phyloseq</code> objects is straightforward using the <code>otu_table()</code>, <code>sample_data()</code>, <code>taxa_table()</code>, and <code>phyloseq()</code> functions. For more information on the <code>phyloseq</code> package, see <a href="https://joey711.github.io/phyloseq/" class="uri">https://joey711.github.io/phyloseq/</a>. In this tutorial we will use two data sets that are included in the <code>phyloseq</code> package.</p>
<div id="global-patterns" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Global Patterns</h2>
<p>The Global Patterns dataset <span class="citation">(Caporaso et al. <a href="#ref-caporaso2011" role="doc-biblioref">2011</a>)</span> is a dataset available in the <code>phyloseq</code> package. These data contain samples from 25 different environmental samples and mock communities. The sampling depth of these samples averages 3.1 million total counts. We will use this dataset to work through the different normalization methods.</p>
<p>The following lines load the relevant packages and data.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="importing-data.html#cb1-1"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="importing-data.html#cb1-2"></a><span class="kw">library</span>(phyloseq)</span>
<span id="cb1-3"><a href="importing-data.html#cb1-3"></a></span>
<span id="cb1-4"><a href="importing-data.html#cb1-4"></a><span class="kw">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>)</span>
<span id="cb1-5"><a href="importing-data.html#cb1-5"></a><span class="co"># examine phyloseq object </span></span>
<span id="cb1-6"><a href="importing-data.html#cb1-6"></a>GlobalPatterns</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
</div>
<div id="colorectal-cancer" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Colorectal cancer</h2>
<p>Another publicly available dataset is from a study on colorectal cancer <span class="citation">(Kostic et al. <a href="#ref-kostic2012" role="doc-biblioref">2012</a>)</span>. See the abstract of this study below:</p>
<blockquote>
<p>The tumor microenvironment of colorectal carcinoma is a complex community of genomically altered cancer cells, nonneoplastic cells, and a diverse collection of microorganisms. Each of these components may contribute to carcinogenesis; however, the role of the microbiota is the least well understood. We have characterized the composition of the microbiota in colorectal carcinoma using whole genome sequences from nine tumor/normal pairs. Fusobacterium sequences were enriched in carcinomas, confirmed by quantitative PCR and 16S rDNA sequence analysis of 95 carcinoma/normal DNA pairs, while the Bacteroidetes and Firmicutes phyla were depleted in tumors. Fusobacteria were also visualized within colorectal tumors using FISH. These findings reveal alterations in the colorectal cancer microbiota; however, the precise role of Fusobacteria in colorectal carcinoma pathogenesis requires further investigation.</p>
</blockquote>
<p>This dataset is downloaded with the <code>phyloseq</code> package as a .biom file and can be loaded in using the following code, showing how to load qiime files as <code>phyloseq</code> objects. These data contain observations from 2505 taxa across 185 samples, which are categorized as ‘Tumor’ or ‘Healthy’. We remove samples that are not categorized into these two categories.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="importing-data.html#cb3-1"></a><span class="co"># Load the data from a system file that is downloaded if the phyloseq package is installed</span></span>
<span id="cb3-2"><a href="importing-data.html#cb3-2"></a>filepath &lt;-<span class="st">  </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;study_1457_split_library_seqs_and_mapping.zip&quot;</span>, <span class="dt">package=</span><span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb3-3"><a href="importing-data.html#cb3-3"></a>kostic &lt;-<span class="st">  </span><span class="kw">microbio_me_qiime</span>(filepath)</span></code></pre></div>
<pre><code>## Found biom-format file, now parsing it... 
## Done parsing biom... 
## Importing Sample Metdadata from mapping file...
## Merging the imported objects... 
## Successfully merged, phyloseq-class created. 
##  Returning...</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="importing-data.html#cb5-1"></a><span class="co"># Some samples are labeled as &quot;none&quot; as a dignosis; remove these samples. </span></span>
<span id="cb5-2"><a href="importing-data.html#cb5-2"></a><span class="co"># Save as an object with the un-normalized counts</span></span>
<span id="cb5-3"><a href="importing-data.html#cb5-3"></a>k_raw &lt;-<span class="st"> </span><span class="kw">subset_samples</span>(kostic, DIAGNOSIS <span class="op">!=</span><span class="st"> &quot;None&quot;</span>)</span>
<span id="cb5-4"><a href="importing-data.html#cb5-4"></a>k_raw</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 2505 taxa and 185 samples ]
## sample_data() Sample Data:       [ 185 samples by 71 sample variables ]
## tax_table()   Taxonomy Table:    [ 2505 taxa by 7 taxonomic ranks ]</code></pre>
</div>
<div id="pre-processing-quality-control-and-filtering" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Pre-processing Quality Control and Filtering</h2>
<p>In addition to normalization, there are some steps we can perform that ideally remove technical artifacts from the sequencing process that only introduce noise.</p>
<p>These filtering steps commonly consist of filtering out samples with a low total read depth and filtering out taxa that are rarely abundant.</p>
<p>Let’s create a filtered version of the Global Patterns dataset. Note that there are only 26 samples, and all have a large library size, so we will not filter out any samples here.</p>
<p>For taxa filtering, we will remove taxa that appear fewer than 5 times in more than half the samples.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="importing-data.html#cb7-1"></a><span class="co"># Determine which taxa to remove</span></span>
<span id="cb7-2"><a href="importing-data.html#cb7-2"></a>filter_taxa &lt;-<span class="st"> </span><span class="kw">genefilter_sample</span>(GlobalPatterns,</span>
<span id="cb7-3"><a href="importing-data.html#cb7-3"></a>                                 <span class="kw">filterfun_sample</span>(<span class="cf">function</span>(x) x <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>),</span>
<span id="cb7-4"><a href="importing-data.html#cb7-4"></a>                                 <span class="dt">A=</span><span class="fl">0.5</span><span class="op">*</span><span class="kw">nsamples</span>(GlobalPatterns))</span>
<span id="cb7-5"><a href="importing-data.html#cb7-5"></a><span class="co"># Remove those taxa from the GlobalPatterns dataset</span></span>
<span id="cb7-6"><a href="importing-data.html#cb7-6"></a><span class="co"># Save as an object with the un-normalized counts</span></span>
<span id="cb7-7"><a href="importing-data.html#cb7-7"></a>gp_raw &lt;-<span class="st"> </span><span class="kw">prune_taxa</span>(filter_taxa, GlobalPatterns)</span>
<span id="cb7-8"><a href="importing-data.html#cb7-8"></a>gp_raw</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 219 taxa and 26 samples ]
## sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 219 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 219 tips and 218 internal nodes ]</code></pre>
<p>This decreases the number of taxa from 19216 to 219. This is not surprising, because this dataset contains samples from widely different locations (gut, soil, etc), and few taxa are shared among all samples and locations. One potential problem with this approach is the widely different locations, so it is possible that the remaining taxa could be some technical artifact, or could be a general ‘core’ set of taxa shared across the disparate environments.</p>
<p>Additionally, let us save the total sampling depth as the variable <code>depth</code> in the metadata for the Global Patterns dataset.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="importing-data.html#cb9-1"></a>gp_raw<span class="op">@</span>sam_data<span class="op">$</span>depth &lt;-<span class="st"> </span><span class="kw">sample_sums</span>(gp_raw)</span></code></pre></div>
<p>We can visualize technical artifacts of sapling depth is by looking at principal coordinates plots using the Bray-Curtis dissimilarity, coloring by sampling depth too see how much variation can be explained by the original sampling depth.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="importing-data.html#cb10-1"></a>gp_raw_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_raw, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb10-2"><a href="importing-data.html#cb10-2"></a><span class="kw">plot_ordination</span>(gp_raw,</span>
<span id="cb10-3"><a href="importing-data.html#cb10-3"></a>                gp_raw_dist,</span>
<span id="cb10-4"><a href="importing-data.html#cb10-4"></a>                <span class="dt">color =</span> <span class="st">&quot;depth&quot;</span>, </span>
<span id="cb10-5"><a href="importing-data.html#cb10-5"></a>                <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw counts&quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/rare_depth_ordcolor-1.png" width="672" /></p>
<p>We don’t see any extreme patterns with sampling depth, but this might additionally be due to the differences in different locations might have different sampling depths. This comparison might be more interesting when we only have one location we are sampling from.</p>
<p>Repeat this same process for the <code>kostic</code> data. Use a prevalence filter to only include taxa that have nonzero counts in over 10% of samples. This results in 478 taxa across 185 samples.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="importing-data.html#cb11-1"></a><span class="co"># Prevalence filter</span></span>
<span id="cb11-2"><a href="importing-data.html#cb11-2"></a>prevalenceThreshold &lt;-<span class="st"> </span><span class="fl">0.10</span> <span class="op">*</span><span class="st"> </span><span class="kw">nsamples</span>(k_raw)</span>
<span id="cb11-3"><a href="importing-data.html#cb11-3"></a>toKeep &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="kw">data.frame</span>(<span class="kw">otu_table</span>(k_raw)), <span class="dv">1</span>, <span class="cf">function</span>(taxa)</span>
<span id="cb11-4"><a href="importing-data.html#cb11-4"></a>    <span class="kw">return</span>(<span class="kw">sum</span>(taxa <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span>prevalenceThreshold))</span>
<span id="cb11-5"><a href="importing-data.html#cb11-5"></a>k_raw &lt;-<span class="st"> </span><span class="kw">prune_taxa</span>(toKeep, k_raw)</span>
<span id="cb11-6"><a href="importing-data.html#cb11-6"></a>k_raw</span></code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 478 taxa and 185 samples ]
## sample_data() Sample Data:       [ 185 samples by 71 sample variables ]
## tax_table()   Taxonomy Table:    [ 478 taxa by 7 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="importing-data.html#cb13-1"></a><span class="co"># Save sampling depth as a sample variable</span></span>
<span id="cb13-2"><a href="importing-data.html#cb13-2"></a>k_raw<span class="op">@</span>sam_data<span class="op">$</span>depth &lt;-<span class="st"> </span><span class="kw">sample_sums</span>(k_raw)</span>
<span id="cb13-3"><a href="importing-data.html#cb13-3"></a></span>
<span id="cb13-4"><a href="importing-data.html#cb13-4"></a>k_raw_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(k_raw, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb13-5"><a href="importing-data.html#cb13-5"></a><span class="kw">plot_ordination</span>(k_raw,</span>
<span id="cb13-6"><a href="importing-data.html#cb13-6"></a>                k_raw_dist,</span>
<span id="cb13-7"><a href="importing-data.html#cb13-7"></a>                <span class="dt">color =</span> <span class="st">&quot;depth&quot;</span>, </span>
<span id="cb13-8"><a href="importing-data.html#cb13-8"></a>                <span class="dt">shape =</span> <span class="st">&quot;DIAGNOSIS&quot;</span>,</span>
<span id="cb13-9"><a href="importing-data.html#cb13-9"></a>                <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw counts&quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>When considering a normalization method, it is important to know if between sample (DA methods) or within sample (community) analysis is of interest.</p>
<!-- ## HMP -->
<!-- Another dataset, not included in analysis yet since it takes very long to load and run analysis on. Could be a more 'realistic' dataset to use.This data comes from healthy subjects.  Data from the Human Microbiome Project from 16S sequences can most easily be accessed using the `HMP16SData` Bioconductor package.  -->
<!-- ```{r, eval = F} -->
<!-- # BiocManager::install("HMP16SData") -->
<!-- library(HMP16SData) -->
<!-- hmp <- as_phyloseq(V13()) -->

</div>
</div>
<h3>References</h3>
<div id="refs" class="references hanging-indent">
<div id="ref-caporaso2011">
<p>Caporaso, J. Gregory, Christian L. Lauber, William A. Walters, Donna Berg-Lyons, Catherine A. Lozupone, Peter J. Turnbaugh, Noah Fierer, and Rob Knight. 2011. “Global Patterns of 16S rRNA Diversity at a Depth of Millions of Sequences Per Sample.” <em>Proceedings of the National Academy of Sciences</em> 108 (Supplement 1): 4516–22. <a href="https://doi.org/10.1073/pnas.1000080107">https://doi.org/10.1073/pnas.1000080107</a>.</p>
</div>
<div id="ref-kostic2012">
<p>Kostic, Aleksandar D., Dirk Gevers, Chandra Sekhar Pedamallu, Monia Michaud, Fujiko Duke, Ashlee M. Earl, Akinyemi I. Ojesina, et al. 2012. “Genomic Analysis Identifies Association of Fusobacterium with Colorectal Carcinoma.” <em>Genome Research</em> 22 (2): 292–98. <a href="https://doi.org/10.1101/gr.126573.111">https://doi.org/10.1101/gr.126573.111</a>.</p>
</div>
<div id="ref-mcmurdie2013">
<p>McMurdie, Paul J., and Susan Holmes. 2013. “Phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PLOS ONE</em> 8 (4): e61217. <a href="https://doi.org/10.1371/journal.pone.0061217">https://doi.org/10.1371/journal.pone.0061217</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="count-scaling-normalization-methods.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MB_normalization_review.pdf", "MB_normalization_review.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
