<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Tutorial on Normalization of Microbiome Data</title>
  <meta name="description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Tutorial on Normalization of Microbiome Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Tutorial on Normalization of Microbiome Data" />
  
  <meta name="twitter:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="importing-data.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">MB Normalization</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#the-importance-of-normalization"><i class="fa fa-check"></i><b>1.1</b> The importance of normalization</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-compositional-nature-of-microbiome-data"><i class="fa fa-check"></i><b>1.2</b> The compositional nature of microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>2</b> Importing Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-data.html"><a href="importing-data.html#global-patterns"><i class="fa fa-check"></i><b>2.1</b> Global Patterns</a></li>
<li class="chapter" data-level="2.2" data-path="importing-data.html"><a href="importing-data.html#colorectal-cancer"><i class="fa fa-check"></i><b>2.2</b> Colorectal cancer</a></li>
<li class="chapter" data-level="2.3" data-path="importing-data.html"><a href="importing-data.html#pre-processing-quality-control-and-filtering"><i class="fa fa-check"></i><b>2.3</b> Pre-processing Quality Control and Filtering</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html"><i class="fa fa-check"></i><b>3</b> Count Scaling Normalization methods</a>
<ul>
<li class="chapter" data-level="3.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#total-sum-scaling-tss"><i class="fa fa-check"></i><b>3.1</b> Total Sum scaling (TSS)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#about-tss"><i class="fa fa-check"></i><b>3.1.1</b> About TSS</a></li>
<li class="chapter" data-level="3.1.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#tss-r-code"><i class="fa fa-check"></i><b>3.1.2</b> TSS R code</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#cumulative-sum-scaling-css"><i class="fa fa-check"></i><b>3.2</b> Cumulative sum scaling (CSS)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#about-css"><i class="fa fa-check"></i><b>3.2.1</b> About CSS</a></li>
<li class="chapter" data-level="3.2.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#css-r-code"><i class="fa fa-check"></i><b>3.2.2</b> CSS R code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="subsampling-methods.html"><a href="subsampling-methods.html"><i class="fa fa-check"></i><b>4</b> Subsampling Methods</a>
<ul>
<li class="chapter" data-level="4.1" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying"><i class="fa fa-check"></i><b>4.1</b> Rarefying</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="subsampling-methods.html"><a href="subsampling-methods.html#about-rarefying"><i class="fa fa-check"></i><b>4.1.1</b> About Rarefying</a></li>
<li class="chapter" data-level="4.1.2" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-walkthrough"><i class="fa fa-check"></i><b>4.1.2</b> Rarefying walkthrough</a></li>
<li class="chapter" data-level="4.1.3" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-r-code"><i class="fa fa-check"></i><b>4.1.3</b> Rarefying R code</a></li>
<li class="chapter" data-level="4.1.4" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-on-kostic-data"><i class="fa fa-check"></i><b>4.1.4</b> Rarefying on Kostic Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html"><i class="fa fa-check"></i><b>5</b> Library Size and Composition Scaling Methods</a>
<ul>
<li class="chapter" data-level="5.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#deseq"><i class="fa fa-check"></i><b>5.1</b> DESeq</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-deseq"><i class="fa fa-check"></i><b>5.1.1</b> About DESeq</a></li>
<li class="chapter" data-level="5.1.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#deseq-r-code"><i class="fa fa-check"></i><b>5.1.2</b> DESeq R Code</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr"><i class="fa fa-check"></i><b>5.2</b> GMPR</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-gmpr"><i class="fa fa-check"></i><b>5.2.1</b> About GMPR</a></li>
<li class="chapter" data-level="5.2.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr-walkthough"><i class="fa fa-check"></i><b>5.2.2</b> GMPR Walkthough</a></li>
<li class="chapter" data-level="5.2.3" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr-r-code"><i class="fa fa-check"></i><b>5.2.3</b> GMPR R Code</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#tmm-edger"><i class="fa fa-check"></i><b>5.3</b> TMM (edgeR)</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-tmm"><i class="fa fa-check"></i><b>5.3.1</b> About TMM</a></li>
<li class="chapter" data-level="5.3.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#edger-tmm-t-code"><i class="fa fa-check"></i><b>5.3.2</b> EdgeR TMM T code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="wrench.html"><a href="wrench.html"><i class="fa fa-check"></i><b>6</b> Wrench</a></li>
<li class="chapter" data-level="7" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html"><i class="fa fa-check"></i><b>7</b> Log Ratio Transformation Methods</a>
<ul>
<li class="chapter" data-level="7.1" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#ancom-ii"><i class="fa fa-check"></i><b>7.1</b> ANCOM II</a></li>
<li class="chapter" data-level="7.2" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#ancom-bc"><i class="fa fa-check"></i><b>7.2</b> ANCOM BC</a></li>
<li class="chapter" data-level="7.3" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#aldex2"><i class="fa fa-check"></i><b>7.3</b> ALDEx2</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="comparisions.html"><a href="comparisions.html"><i class="fa fa-check"></i><b>8</b> Comparisions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial on Normalization of Microbiome Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Tutorial on Normalization of Microbiome Data</h1>
</div>
<div id="intro" class="section level1" number="1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<div id="the-importance-of-normalization" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> The importance of normalization</h2>
<p>Microbiome data must be normalized before any statistical analysis can be performed. Following the process of sequencing and assigning raw reads into counts per observed and classified identified taxa classes/OTUs/ASVs, microbiome data consist of a matrix of read counts, referred to as a feature table of raw counts. Normalization is the process of transforming raw read count data into data that can be compared between samples. Statistical analysis on this count matrix is then performed depending on the goal of the experiment. Common analysis goals include community-level analysis (alpha/beta diversity), differential abundance testing (the parallel of differential expression testing in gene expression studies), and network analysis.</p>
<!-- Terminology "feature table" and taxonomy table, and metadata" , and potentially phyloenetic tree -->
<!-- insert table example of what MB data look like??  what is a raw read/what are asvs or otus -->
<p>Analysis of composition, differences, connections, etc., should be done based only on true biological aspects. However, technical variation in counts across samples is a given hurdle that must be accounted for. Biases can arise in the sequencing process, sample preparation, contamination, preferential amplification, and can manifest in differences in sparsity and unequal sequencing depths <span class="citation">(Salter et al. <a href="#ref-salter2014" role="doc-biblioref">2014</a>)</span>. An effective normalization strategy should put all samples on equal footing so interpretations are on biological signals, not technical signals such as sequencing depth. Currently, there is no known ‘best’ normalization method that removes all technical artifacts leaving only biological signals.</p>
<p>Due to the sequencing technology, samples will have different sequencing depths, or the sum of all the counts in a sample. Directly comparing raw counts between samples is not possible. To illustrate this, consider the counts of Taxon 1, across two samples shown below. In Sample A, this taxon has a count of 230, and in Sample B, this taxon has a count of 23. Is this taxon differentially abundant between samples? As we see below, the way we normalize the data can change how we would answer this question.</p>
<table>
<caption><span id="tab:unnamed-chunk-1">Table 1.1: </span>The raw counts of taxon 1 are different, and the raw counts of taxon 2 are equal. Sampling depth is not accounted for.</caption>
<thead>
<tr class="header">
<th align="left">Taxa</th>
<th align="right">Sample A</th>
<th align="right">Sample B</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Taxon 1</td>
<td align="right">230</td>
<td align="right">23</td>
</tr>
<tr class="even">
<td align="left">Taxon 2</td>
<td align="right">5</td>
<td align="right">5</td>
</tr>
</tbody>
</table>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>One goal for normalization of microbiome data is the standardization of sequencing depth across samples. One common approach to this is a scaling-based approach, where a scaling factor is calculated for every sample and the counts for each taxon are divided by the scaling factor for that sample. Figure B shows the same data as Figure A, but where each sample has been transformed into proportions by dividing by the total counts for each sample. The difference in Taxon 1 between samples appears much smaller. However, now there appears to be a difference in Taxon 2, even though the counts were originally the same. This is because in sample B, Taxon 1 consists of a higher proportion of the total count than in sample A.</p>
<p>This demonstrates the importance of normalization, but also the artifacts that can occur depending on the method.</p>
</div>
<div id="the-compositional-nature-of-microbiome-data" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> The compositional nature of microbiome data</h2>
<p>Microbiome data are inherently compositional. The counts of the collection of taxa that make up each sample are constrained by the total sum, or sequencing depth for that sample. This means that the count of each sampled taxon is a portion of a larger whole. Each observed taxon is not independent. As we saw in the above example, before normalization, Taxon 2 was equal between samples. After converting to proportions, Taxon 2 no longer appears equal. If there is a difference between two samples it is unclear if that difference is because of a true difference in that taxon or if that taxon is changing because of differences in another taxon. Numerous traditional statistical methods rely on an independence assumption, which is not met with microbiome data. This can lead to spurious correlations that exist only because of the compositional nature and not any true signal.</p>
<p>With library size as the sum constraint for each sample, if we know in a biological system that after an event occurs (treatment), the red taxon decreases, this will change the composition of the sampled blue taxon regardless of its change or lack thereof in the underlying population.</p>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>Consider again two samples consisting of red and blue points. We can think of the samples as before and after treatment. In the second plot, the number of red dots in the population and in the observed sample has decreased, but the blue remains the same.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-4">Table 1.2: </span>Counts of sampled red and blue taxa before and after
</caption>
<thead>
<tr>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:right;">
blue
</th>
<th style="text-align:right;">
red
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Before
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
20
</td>
</tr>
<tr>
<td style="text-align:left;">
After
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
4
</td>
</tr>
</tbody>
</table>
<table>
<caption>
<span id="tab:unnamed-chunk-4">Table 1.2: </span>Proportions of sampled red and blue taxa before and after
</caption>
<thead>
<tr>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:right;">
blue
</th>
<th style="text-align:right;">
red
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Before
</td>
<td style="text-align:right;">
0.2857143
</td>
<td style="text-align:right;">
0.6666667
</td>
</tr>
<tr>
<td style="text-align:left;">
After
</td>
<td style="text-align:right;">
0.7142857
</td>
<td style="text-align:right;">
0.3333333
</td>
</tr>
</tbody>
</table>
<!-- In the after sample, the blue counts have decreased, as we would expect from our knowledge of the population. However, the red counts have increased, even though there is no change in the red population.  -->
<p>This observed increase in the proportion of blue is due to the compositional nature of the sampled points, and not any true difference in the blue population.</p>
<!-- INCLUDE EXAMPLE?  -->
<!-- ## Zero-inflation of microbiome data -->
<!-- Include section on considerations that are necessary to account for zero-inflation. -->
<!-- ## Intro to the phyloseq package -->
<!-- ## Phyloseq wrapper function -->
<!-- Many normalization methods, especially those developed for other disciplines, take the raw matrix as an argument, along with a vector of the grouping variable. For convenience, we have included a wrapper function for phyloseq that ... -->
<!-- ## Conventions for this bookdown -->
<!-- #### Objects -->
<!-- Object names are the abbreviation of the dataset source (ie hmp), followed by the normalization method used, the filtered raw counts have the normalization method "raw". -->
<!-- #### Functions -->
<!-- Function names for normalization use "norm", folowed by an underscore, followed by the name or abbreviation for the method -->
<!-- #### Bio stuff  -->
<!-- ASVs/OTUs/taxa. Refer to everything as taxa.  -->

</div>
</div>
<h3>References</h3>
<div id="refs" class="references hanging-indent">
<div id="ref-salter2014">
<p>Salter, Susannah J., Michael J. Cox, Elena M. Turek, Szymon T. Calus, William O. Cookson, Miriam F. Moffatt, Paul Turner, Julian Parkhill, Nicholas J. Loman, and Alan W. Walker. 2014. “Reagent and Laboratory Contamination Can Critically Impact Sequence-Based Microbiome Analyses.” <em>BMC Biology</em> 12 (1): 87. <a href="https://doi.org/10.1186/s12915-014-0087-z">https://doi.org/10.1186/s12915-014-0087-z</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="importing-data.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MB_normalization_review.pdf", "MB_normalization_review.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
