<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Library Size and Composition Scaling Methods | Tutorial on Normalization of Microbiome Data</title>
  <meta name="description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Library Size and Composition Scaling Methods | Tutorial on Normalization of Microbiome Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Library Size and Composition Scaling Methods | Tutorial on Normalization of Microbiome Data" />
  
  <meta name="twitter:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="subsampling-methods.html"/>
<link rel="next" href="wrench.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">MB Normalization</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#the-importance-of-normalization"><i class="fa fa-check"></i><b>1.1</b> The importance of normalization</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-compositional-nature-of-microbiome-data"><i class="fa fa-check"></i><b>1.2</b> The compositional nature of microbiome data</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#log-ratio-transformation-methods"><i class="fa fa-check"></i><b>1.2.1</b> Log ratio transformation methods</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>2</b> Importing Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-data.html"><a href="importing-data.html#global-patterns"><i class="fa fa-check"></i><b>2.1</b> Global Patterns</a></li>
<li class="chapter" data-level="2.2" data-path="importing-data.html"><a href="importing-data.html#colorectal-cancer"><i class="fa fa-check"></i><b>2.2</b> Colorectal cancer</a></li>
<li class="chapter" data-level="2.3" data-path="importing-data.html"><a href="importing-data.html#pre-processing-quality-control-and-filtering"><i class="fa fa-check"></i><b>2.3</b> Pre-processing Quality Control and Filtering</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html"><i class="fa fa-check"></i><b>3</b> Count Scaling Normalization methods</a>
<ul>
<li class="chapter" data-level="3.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#total-sum-scaling-tss"><i class="fa fa-check"></i><b>3.1</b> Total Sum scaling (TSS)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#about-tss"><i class="fa fa-check"></i><b>3.1.1</b> About TSS</a></li>
<li class="chapter" data-level="3.1.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#tss-r-code"><i class="fa fa-check"></i><b>3.1.2</b> TSS R code</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#cumulative-sum-scaling-css"><i class="fa fa-check"></i><b>3.2</b> Cumulative sum scaling (CSS)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#about-css"><i class="fa fa-check"></i><b>3.2.1</b> About CSS</a></li>
<li class="chapter" data-level="3.2.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#css-r-code"><i class="fa fa-check"></i><b>3.2.2</b> CSS R code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="subsampling-methods.html"><a href="subsampling-methods.html"><i class="fa fa-check"></i><b>4</b> Subsampling Methods</a>
<ul>
<li class="chapter" data-level="4.1" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying"><i class="fa fa-check"></i><b>4.1</b> Rarefying</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="subsampling-methods.html"><a href="subsampling-methods.html#about-rarefying"><i class="fa fa-check"></i><b>4.1.1</b> About Rarefying</a></li>
<li class="chapter" data-level="4.1.2" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-walkthrough"><i class="fa fa-check"></i><b>4.1.2</b> Rarefying walkthrough</a></li>
<li class="chapter" data-level="4.1.3" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-r-code"><i class="fa fa-check"></i><b>4.1.3</b> Rarefying R code</a></li>
<li class="chapter" data-level="4.1.4" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-on-kostic-data"><i class="fa fa-check"></i><b>4.1.4</b> Rarefying on Kostic Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html"><i class="fa fa-check"></i><b>5</b> Library Size and Composition Scaling Methods</a>
<ul>
<li class="chapter" data-level="5.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#deseq"><i class="fa fa-check"></i><b>5.1</b> DESeq</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-deseq"><i class="fa fa-check"></i><b>5.1.1</b> About DESeq</a></li>
<li class="chapter" data-level="5.1.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#deseq-r-code"><i class="fa fa-check"></i><b>5.1.2</b> DESeq R Code</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr"><i class="fa fa-check"></i><b>5.2</b> GMPR</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-gmpr"><i class="fa fa-check"></i><b>5.2.1</b> About GMPR</a></li>
<li class="chapter" data-level="5.2.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr-walkthough"><i class="fa fa-check"></i><b>5.2.2</b> GMPR Walkthough</a></li>
<li class="chapter" data-level="5.2.3" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr-r-code"><i class="fa fa-check"></i><b>5.2.3</b> GMPR R Code</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#tmm-edger"><i class="fa fa-check"></i><b>5.3</b> TMM (edgeR)</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-tmm"><i class="fa fa-check"></i><b>5.3.1</b> About TMM</a></li>
<li class="chapter" data-level="5.3.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#edger-tmm-t-code"><i class="fa fa-check"></i><b>5.3.2</b> EdgeR TMM T code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="wrench.html"><a href="wrench.html"><i class="fa fa-check"></i><b>6</b> Wrench</a></li>
<li class="chapter" data-level="7" data-path="log-ratio-transformation-methods-1.html"><a href="log-ratio-transformation-methods-1.html"><i class="fa fa-check"></i><b>7</b> Log Ratio Transformation Methods</a>
<ul>
<li class="chapter" data-level="7.1" data-path="log-ratio-transformation-methods-1.html"><a href="log-ratio-transformation-methods-1.html#ancom-ii"><i class="fa fa-check"></i><b>7.1</b> ANCOM II</a></li>
<li class="chapter" data-level="7.2" data-path="log-ratio-transformation-methods-1.html"><a href="log-ratio-transformation-methods-1.html#ancom-bc"><i class="fa fa-check"></i><b>7.2</b> ANCOM BC</a></li>
<li class="chapter" data-level="7.3" data-path="log-ratio-transformation-methods-1.html"><a href="log-ratio-transformation-methods-1.html#aldex2"><i class="fa fa-check"></i><b>7.3</b> ALDEx2</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="comparisions.html"><a href="comparisions.html"><i class="fa fa-check"></i><b>8</b> Comparisions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial on Normalization of Microbiome Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="library-size-and-composition-scaling-methods" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Library Size and Composition Scaling Methods</h1>
<div id="deseq" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> DESeq</h2>
<div id="about-deseq" class="section level3" number="5.1.1">
<h3><span class="header-section-number">5.1.1</span> About DESeq</h3>
<p>The DESeq2 package includes a normalization method for adjusting for differing library sizes across samples <span class="citation">(Anders and Huber <a href="#ref-anders2010" role="doc-biblioref">2010</a>)</span>. This method also can account for differences in library composition. A scaling factor to normalize each sample takes into account both library size and library composition. This method has also been called MED, RLE, or DESeq in the literature.</p>
<p>DESeq2 first takes the natural logarithm of every entry in the count matrix. Due to this, all entries with zero will be set to negative infinity. Next, the row average is calculated (geometric average), so we have a vector of average counts for each taxon. Taking the log first should avoid undue influence by extreme outliers. All taxa with an average of infinity are removed. This step will remove all taxa with zero read count in one or more samples. This can be a problem in microbiome data. Next, we subtract the average log value from the log(counts), this gives a log ratio. This is equivalent to the ratio of the reads in each sample to the average across all samples. Next, we calculate the median of the log-ratios for each sample. These medians are converted to scaling factors for each sample by exponentiation. An extension of this method, denoted ‘poscounts’, has been suggested, which instead of taking the geometric mean of the logged counts for each taxon, we take the n-th root of the product of the non-zero counts.</p>
<p>This method assumes that the taxon of median absolute abundance is not differentially abundant, which is more likely true for the RNA-Seq it was developed for, but may not be true for microbiome studies, especially when there are more study groups, or we are analyzing higher taxonomic levels.</p>
<p>An additional option can be used to perform a variance stabilizing transformation on the count matrix before normalizing with the above size factors. This method calculates a dispersion-mean relationship and then transforms the data. The result ideally is an abundance matrix that is approximately homoskedastic or constant variance across the range of mean values. The package also includes an option for a ‘rlog’ transform, which they recommend over the variance stabilizing method in the case when there is a large difference in library sizes.</p>
<p>If differential abundance is of interest to calculate, DESeq uses a negative binomial distribution to model differential abundances. It is possible to provide the size factors calculated by another method to DESeq to perform differential analysis.</p>
<div id="deseq-walkthrough" class="section level4" number="5.1.1.1">
<h4><span class="header-section-number">5.1.1.1</span> DESeq walkthrough</h4>
<p>Consider again this example dataset.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-20">Table 5.1: </span>Example dataset - Raw Counts
</caption>
<thead>
<tr>
<th style="text-align:left;">
Taxa
</th>
<th style="text-align:right;">
Sample 1
</th>
<th style="text-align:right;">
Sample2
</th>
<th style="text-align:right;">
Sample 3
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Taxon 1
</td>
<td style="text-align:right;">
132
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 2
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
48
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 4
</td>
<td style="text-align:right;">
23
</td>
<td style="text-align:right;">
15
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 5
</td>
<td style="text-align:right;">
71
</td>
<td style="text-align:right;">
80
</td>
<td style="text-align:right;">
5
</td>
</tr>
</tbody>
</table>
<p>The first step in DESeq normalization is to take the natural log of each entry in the count matrix.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-21">Table 5.2: </span>Step 1: Logged Raw Counts
</caption>
<thead>
<tr>
<th style="text-align:left;">
Taxa
</th>
<th style="text-align:right;">
Sample 1
</th>
<th style="text-align:right;">
Sample2
</th>
<th style="text-align:right;">
Sample 3
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Taxon 1
</td>
<td style="text-align:right;">
4.882802
</td>
<td style="text-align:right;">
4.6347290
</td>
<td style="text-align:right;">
2.3978953
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 2
</td>
<td style="text-align:right;">
1.945910
</td>
<td style="text-align:right;">
3.8712010
</td>
<td style="text-align:right;">
1.0986123
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 3
</td>
<td style="text-align:right;">
-Inf
</td>
<td style="text-align:right;">
0.6931472
</td>
<td style="text-align:right;">
0.0000000
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 4
</td>
<td style="text-align:right;">
3.135494
</td>
<td style="text-align:right;">
2.7080502
</td>
<td style="text-align:right;">
0.6931472
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 5
</td>
<td style="text-align:right;">
4.262680
</td>
<td style="text-align:right;">
4.3820266
</td>
<td style="text-align:right;">
1.6094379
</td>
</tr>
</tbody>
</table>
<p>Notice the one entry with zero counts is marked as negative infinity.</p>
<p>Next, the average of the logged values is taken across the samples. This results in an average log value for each taxa. This averaged log value is helpful for normalizing because it is not overwhelmed by outliers.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-22">Table 5.3: </span>Step 2: Average Logged Counts
</caption>
<thead>
<tr>
<th style="text-align:left;">
Taxa
</th>
<th style="text-align:right;">
Average Log Counts
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Taxon 1
</td>
<td style="text-align:right;">
3.971809
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 2
</td>
<td style="text-align:right;">
2.305241
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 3
</td>
<td style="text-align:right;">
-Inf
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 4
</td>
<td style="text-align:right;">
2.178897
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 5
</td>
<td style="text-align:right;">
3.418048
</td>
</tr>
</tbody>
</table>
<p>The row for the taxa that contained a zero on one of the samples is negative infinity. This taxon is now excluded from the following steps. This step results in removing any taxa that have zero counts from being considered to contribute to calculating the scaling factors.</p>
<!-- The idea here is to keep housekeeping genes  -->
<p>The next step is to subtract the average log values (step 2) from the log of the raw counts (step 1), only including rows that were not filtered. This step shows which samples have counts in a sample higher or lower than the average.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-23">Table 5.4: </span>Step 3: Log Ratio of reads in each sample to average across all samples
</caption>
<thead>
<tr>
<th style="text-align:left;">
Taxa
</th>
<th style="text-align:left;">
Sample 1
</th>
<th style="text-align:left;">
Sample 2
</th>
<th style="text-align:left;">
Sample 3
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Taxon 1
</td>
<td style="text-align:left;">
4.88 - 3.97 = 0.91
</td>
<td style="text-align:left;">
4.63 - 3.97 = 0.66
</td>
<td style="text-align:left;">
2.40 - 3.97 = -1.57
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 2
</td>
<td style="text-align:left;">
1.95 - 2.31 = -0.36
</td>
<td style="text-align:left;">
3.97 - 2.31 = 1.66
</td>
<td style="text-align:left;">
1.10 - 2.31 = -1.31
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 4
</td>
<td style="text-align:left;">
3.14 - 2.18 = 0.96
</td>
<td style="text-align:left;">
2.71 - 2.18 = 0.53
</td>
<td style="text-align:left;">
0.69 - 2.18 = -1.49
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 5
</td>
<td style="text-align:left;">
4.26 - 3.42 = 0.84
</td>
<td style="text-align:left;">
4.38 - 3.42 = 0.96
</td>
<td style="text-align:left;">
1.61- 3.42 = -1.81
</td>
</tr>
</tbody>
</table>
<p>Next, to calculate the scaling factors for each sample, we take the median of the log ratios calculated in the above step. Like using logs, calculating medians avoids the influence of outlier taxa that put undue influence on the scaling factor.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-24">Table 5.5: </span>Step 4: Calculating medians per sample
</caption>
<thead>
<tr>
<th style="text-align:left;">
Taxa
</th>
<th style="text-align:right;">
Sample 1
</th>
<th style="text-align:right;">
Sample 2
</th>
<th style="text-align:right;">
Sample 3
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Taxon 1
</td>
<td style="text-align:right;">
0.910000
</td>
<td style="text-align:right;">
0.660000
</td>
<td style="text-align:right;">
-1.5700000
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 2
</td>
<td style="text-align:right;">
-0.360000
</td>
<td style="text-align:right;">
1.660000
</td>
<td style="text-align:right;">
-1.3100000
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 4
</td>
<td style="text-align:right;">
0.960000
</td>
<td style="text-align:right;">
0.530000
</td>
<td style="text-align:right;">
-1.4900000
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 5
</td>
<td style="text-align:right;">
0.840000
</td>
<td style="text-align:right;">
0.960000
</td>
<td style="text-align:right;">
-1.8100000
</td>
</tr>
<tr>
<td style="text-align:left;">
Median
</td>
<td style="text-align:right;">
0.875000
</td>
<td style="text-align:right;">
0.810000
</td>
<td style="text-align:right;">
-1.5300000
</td>
</tr>
<tr>
<td style="text-align:left;">
Exponentiated Median
</td>
<td style="text-align:right;">
2.398875
</td>
<td style="text-align:right;">
2.247908
</td>
<td style="text-align:right;">
0.2165357
</td>
</tr>
</tbody>
</table>
<p>Finally, we normalize the data, by exponentiating the median log ratios for each sample, which are the final scaling factors. We then divide all the raw counts in a sample by the sample’s scaling factor.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-25">Table 5.6: </span>Step 5: DESeq Normalized Counts
</caption>
<thead>
<tr>
<th style="text-align:left;">
Taxa
</th>
<th style="text-align:right;">
Sample 1
</th>
<th style="text-align:right;">
Sample2
</th>
<th style="text-align:right;">
Sample 3
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Taxon 1
</td>
<td style="text-align:right;">
55.025787
</td>
<td style="text-align:right;">
45.8203808
</td>
<td style="text-align:right;">
50.799945
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 2
</td>
<td style="text-align:right;">
2.918034
</td>
<td style="text-align:right;">
21.3531872
</td>
<td style="text-align:right;">
13.854530
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 3
</td>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
0.8897161
</td>
<td style="text-align:right;">
4.618177
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 4
</td>
<td style="text-align:right;">
9.587827
</td>
<td style="text-align:right;">
6.6728710
</td>
<td style="text-align:right;">
9.236354
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 5
</td>
<td style="text-align:right;">
29.597203
</td>
<td style="text-align:right;">
35.5886453
</td>
<td style="text-align:right;">
23.090884
</td>
</tr>
</tbody>
</table>
<p>Since Taxon 3 had a zero count in sample 1, it was excluded from the calculation of scale factors. The above example dataset may not be characteristic of microbiome datasets. Microbiome datasets are zero-inflated, meaning that there are numerous zero counts in the raw count matrix. Even up to 80-90% of the counts in a microbiome datset can be zero. Because of this, if the zero-inflation in the datseta is not accounted for, very few, or even perhaps none of the taxa will contribute to calculating the scaling factor. One option so that all of the taxa are included in the calculation is to add a pseudocount so none of the counts are zero. Another option is the <code>poscounts</code> option, which is encouraged for microbiome data. Instead of taking the average of the logged counts, it takes the <span class="math inline">\(n\)</span>th root of the non-zero counts. This replaces step 2 in this example.</p>
</div>
</div>
<div id="deseq-r-code" class="section level3" number="5.1.2">
<h3><span class="header-section-number">5.1.2</span> DESeq R Code</h3>
<div id="function-2" class="section level4" number="5.1.2.1">
<h4><span class="header-section-number">5.1.2.1</span> Function</h4>
<p>Here we provide two normalization functions implemented in R using DESeq methods. The first calculates the RLE normalization using the <code>poscounts</code> option for microbiome data. The second calculates the variance stabilizing transformation.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="library-size-and-composition-scaling-methods.html#cb50-1"></a>norm_DESeq_RLE_poscounts &lt;-<span class="st"> </span><span class="cf">function</span>(ps, <span class="dt">group =</span> <span class="dv">1</span>){</span>
<span id="cb50-2"><a href="library-size-and-composition-scaling-methods.html#cb50-2"></a>  <span class="kw">require</span>(DESeq2, <span class="dt">quietly =</span> T)</span>
<span id="cb50-3"><a href="library-size-and-composition-scaling-methods.html#cb50-3"></a>    <span class="co"># keep arbitrary design for normalization </span></span>
<span id="cb50-4"><a href="library-size-and-composition-scaling-methods.html#cb50-4"></a>    <span class="co"># Convert to DESeq object</span></span>
<span id="cb50-5"><a href="library-size-and-composition-scaling-methods.html#cb50-5"></a>    ps_dds &lt;-<span class="st"> </span><span class="kw">phyloseq_to_deseq2</span>(ps, <span class="op">~</span><span class="dv">1</span>)</span>
<span id="cb50-6"><a href="library-size-and-composition-scaling-methods.html#cb50-6"></a>    <span class="co"># Calculate the size factors (scaling)</span></span>
<span id="cb50-7"><a href="library-size-and-composition-scaling-methods.html#cb50-7"></a>    ps_dds &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(ps_dds, <span class="dt">type =</span> <span class="st">&quot;poscounts&quot;</span>)</span>
<span id="cb50-8"><a href="library-size-and-composition-scaling-methods.html#cb50-8"></a>    <span class="co"># Extract counts</span></span>
<span id="cb50-9"><a href="library-size-and-composition-scaling-methods.html#cb50-9"></a>    counts &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">counts</span>(ps_dds, <span class="dt">normalized =</span> T)</span>
<span id="cb50-10"><a href="library-size-and-composition-scaling-methods.html#cb50-10"></a>    <span class="co"># Convert back to phyloseq</span></span>
<span id="cb50-11"><a href="library-size-and-composition-scaling-methods.html#cb50-11"></a>    otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(counts, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb50-12"><a href="library-size-and-composition-scaling-methods.html#cb50-12"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb50-13"><a href="library-size-and-composition-scaling-methods.html#cb50-13"></a>    sam<span class="op">$</span>scaling_factor &lt;-<span class="st"> </span><span class="kw">sizeFactors</span>(ps_dds)</span>
<span id="cb50-14"><a href="library-size-and-composition-scaling-methods.html#cb50-14"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb50-15"><a href="library-size-and-composition-scaling-methods.html#cb50-15"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb50-16"><a href="library-size-and-composition-scaling-methods.html#cb50-16"></a>    ps_DESeq &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu,sam,tax,phy)</span>
<span id="cb50-17"><a href="library-size-and-composition-scaling-methods.html#cb50-17"></a>    <span class="kw">return</span>(ps_DESeq)</span>
<span id="cb50-18"><a href="library-size-and-composition-scaling-methods.html#cb50-18"></a>}</span>
<span id="cb50-19"><a href="library-size-and-composition-scaling-methods.html#cb50-19"></a></span>
<span id="cb50-20"><a href="library-size-and-composition-scaling-methods.html#cb50-20"></a></span>
<span id="cb50-21"><a href="library-size-and-composition-scaling-methods.html#cb50-21"></a></span>
<span id="cb50-22"><a href="library-size-and-composition-scaling-methods.html#cb50-22"></a>norm_DESeq_vs &lt;-<span class="st"> </span><span class="cf">function</span>(ps, <span class="dt">group =</span> <span class="dv">1</span>){</span>
<span id="cb50-23"><a href="library-size-and-composition-scaling-methods.html#cb50-23"></a>  <span class="kw">require</span>(DESeq2, <span class="dt">quietly =</span> T)</span>
<span id="cb50-24"><a href="library-size-and-composition-scaling-methods.html#cb50-24"></a>  ps_dds &lt;-<span class="st"> </span><span class="kw">phyloseq_to_deseq2</span>(ps, <span class="op">~</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb50-25"><a href="library-size-and-composition-scaling-methods.html#cb50-25"></a>  ps_dds &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(ps_dds, <span class="dt">type =</span> <span class="st">&quot;poscounts&quot;</span>)</span>
<span id="cb50-26"><a href="library-size-and-composition-scaling-methods.html#cb50-26"></a>  <span class="co"># Variance transformation</span></span>
<span id="cb50-27"><a href="library-size-and-composition-scaling-methods.html#cb50-27"></a>  ps_dds &lt;-<span class="st"> </span><span class="kw">estimateDispersions</span>(ps_dds)</span>
<span id="cb50-28"><a href="library-size-and-composition-scaling-methods.html#cb50-28"></a>  abund &lt;-<span class="st"> </span><span class="kw">getVarianceStabilizedData</span>(ps_dds)</span>
<span id="cb50-29"><a href="library-size-and-composition-scaling-methods.html#cb50-29"></a>  <span class="co"># don’t allow deseq to return negative counts</span></span>
<span id="cb50-30"><a href="library-size-and-composition-scaling-methods.html#cb50-30"></a>  <span class="co"># add the minimum count to all values</span></span>
<span id="cb50-31"><a href="library-size-and-composition-scaling-methods.html#cb50-31"></a>  <span class="co"># another option is to replace negative counts with 0</span></span>
<span id="cb50-32"><a href="library-size-and-composition-scaling-methods.html#cb50-32"></a>  abund &lt;-<span class="st"> </span>abund <span class="op">+</span><span class="st"> </span><span class="kw">abs</span>(<span class="kw">min</span>(abund)) </span>
<span id="cb50-33"><a href="library-size-and-composition-scaling-methods.html#cb50-33"></a>  otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(abund, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb50-34"><a href="library-size-and-composition-scaling-methods.html#cb50-34"></a>  sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb50-35"><a href="library-size-and-composition-scaling-methods.html#cb50-35"></a>  tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb50-36"><a href="library-size-and-composition-scaling-methods.html#cb50-36"></a>  phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb50-37"><a href="library-size-and-composition-scaling-methods.html#cb50-37"></a>  ps_DESeq &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu,sam,tax,phy)</span>
<span id="cb50-38"><a href="library-size-and-composition-scaling-methods.html#cb50-38"></a>  <span class="kw">return</span>(ps_DESeq)</span>
<span id="cb50-39"><a href="library-size-and-composition-scaling-methods.html#cb50-39"></a>}</span></code></pre></div>
</div>
<div id="deseq-implemented-on-global-patterns" class="section level4" number="5.1.2.2">
<h4><span class="header-section-number">5.1.2.2</span> DESeq implemented on Global Patterns</h4>
<p>Perform DESeq RLE normalization as well as DESeq variance stabilized transformation on Global Patterns:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="library-size-and-composition-scaling-methods.html#cb51-1"></a>gp_deseq_rle &lt;-<span class="st"> </span><span class="kw">norm_DESeq_RLE_poscounts</span>(gp_raw)</span>
<span id="cb51-2"><a href="library-size-and-composition-scaling-methods.html#cb51-2"></a>gp_deseq_vs &lt;-<span class="st"> </span><span class="kw">norm_DESeq_vs</span>(gp_raw)</span></code></pre></div>
<p>Examine principal coordinate plots between raw data and both DESeq normalized data.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="library-size-and-composition-scaling-methods.html#cb52-1"></a><span class="co"># First calculate distance matrices </span></span>
<span id="cb52-2"><a href="library-size-and-composition-scaling-methods.html#cb52-2"></a>gp_rle_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_deseq_rle, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) </span>
<span id="cb52-3"><a href="library-size-and-composition-scaling-methods.html#cb52-3"></a>gp_vs_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_deseq_vs, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) </span>
<span id="cb52-4"><a href="library-size-and-composition-scaling-methods.html#cb52-4"></a></span>
<span id="cb52-5"><a href="library-size-and-composition-scaling-methods.html#cb52-5"></a><span class="co"># Plot ordinations</span></span>
<span id="cb52-6"><a href="library-size-and-composition-scaling-methods.html#cb52-6"></a><span class="kw">plot_ordination</span>(gp_raw, gp_raw_dist, <span class="dt">color =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw data&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb52-7"><a href="library-size-and-composition-scaling-methods.html#cb52-7"></a><span class="kw">plot_ordination</span>(gp_deseq_rle, gp_rle_dist, <span class="dt">color =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PCoA on RLE&quot;</span>) <span class="op">+</span></span>
<span id="cb52-8"><a href="library-size-and-composition-scaling-methods.html#cb52-8"></a><span class="kw">plot_ordination</span>(gp_deseq_vs, gp_vs_dist, <span class="dt">color =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PCoA on VST&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb52-9"><a href="library-size-and-composition-scaling-methods.html#cb52-9"></a><span class="st">  </span><span class="kw">plot_layout</span>(<span class="dt">guides =</span> <span class="st">&#39;collect&#39;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/deseq-ordination-1.png" width="672" /></p>
<p>See how dissimilarity matrices differ from raw counts and each DESeq transformation.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="library-size-and-composition-scaling-methods.html#cb53-1"></a><span class="kw">plot_norm_changes</span>(gp_deseq_rle, gp_raw, </span>
<span id="cb53-2"><a href="library-size-and-composition-scaling-methods.html#cb53-2"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;Raw&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;RLE&quot;</span>, </span>
<span id="cb53-3"><a href="library-size-and-composition-scaling-methods.html#cb53-3"></a>                  <span class="dt">title =</span> <span class="st">&quot;Distance metric comparision between RLE normalization and Raw counts &quot;</span>) <span class="op">/</span><span class="st"> </span></span>
<span id="cb53-4"><a href="library-size-and-composition-scaling-methods.html#cb53-4"></a><span class="kw">plot_norm_changes</span>(gp_deseq_vs, gp_raw, </span>
<span id="cb53-5"><a href="library-size-and-composition-scaling-methods.html#cb53-5"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;Raw&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;VST&quot;</span>, </span>
<span id="cb53-6"><a href="library-size-and-composition-scaling-methods.html#cb53-6"></a>                  <span class="dt">title =</span> <span class="st">&quot;Distance metric comparision between VST normalization and Raw counts &quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb53-7"><a href="library-size-and-composition-scaling-methods.html#cb53-7"></a><span class="st">  </span><span class="kw">plot_layout</span>(<span class="dt">guides =</span> <span class="st">&#39;collect&#39;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-27-1.png" width="672" /></p>
</div>
<div id="deseq-implemented-on-kostic-dataset" class="section level4" number="5.1.2.3">
<h4><span class="header-section-number">5.1.2.3</span> DESeq implemented on Kostic dataset</h4>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="library-size-and-composition-scaling-methods.html#cb54-1"></a>k_deseq_rle &lt;-<span class="st"> </span><span class="kw">norm_DESeq_RLE_poscounts</span>(k_raw)</span>
<span id="cb54-2"><a href="library-size-and-composition-scaling-methods.html#cb54-2"></a>k_deseq_vs &lt;-<span class="st"> </span><span class="kw">norm_DESeq_vs</span>(k_raw)</span>
<span id="cb54-3"><a href="library-size-and-composition-scaling-methods.html#cb54-3"></a></span>
<span id="cb54-4"><a href="library-size-and-composition-scaling-methods.html#cb54-4"></a><span class="co"># First calculate distance matrices </span></span>
<span id="cb54-5"><a href="library-size-and-composition-scaling-methods.html#cb54-5"></a>k_rle_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(k_deseq_rle, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) </span>
<span id="cb54-6"><a href="library-size-and-composition-scaling-methods.html#cb54-6"></a>k_vs_dist &lt;-<span class="st"> </span>phyloseq<span class="op">::</span><span class="kw">ordinate</span>(k_deseq_vs, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) </span>
<span id="cb54-7"><a href="library-size-and-composition-scaling-methods.html#cb54-7"></a></span>
<span id="cb54-8"><a href="library-size-and-composition-scaling-methods.html#cb54-8"></a><span class="co"># Plot ordinations</span></span>
<span id="cb54-9"><a href="library-size-and-composition-scaling-methods.html#cb54-9"></a><span class="kw">plot_ordination</span>(k_raw, k_raw_dist, <span class="dt">color =</span> <span class="st">&quot;DIAGNOSIS&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw data&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb54-10"><a href="library-size-and-composition-scaling-methods.html#cb54-10"></a><span class="kw">plot_ordination</span>(k_deseq_rle, k_rle_dist, <span class="dt">color =</span> <span class="st">&quot;DIAGNOSIS&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PCoA on RLE&quot;</span>) <span class="op">+</span></span>
<span id="cb54-11"><a href="library-size-and-composition-scaling-methods.html#cb54-11"></a><span class="kw">plot_ordination</span>(k_deseq_vs, k_vs_dist, <span class="dt">color =</span> <span class="st">&quot;DIAGNOSIS&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PCoA on VST&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb54-12"><a href="library-size-and-composition-scaling-methods.html#cb54-12"></a><span class="st">  </span><span class="kw">plot_layout</span>(<span class="dt">guides =</span> <span class="st">&#39;collect&#39;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="library-size-and-composition-scaling-methods.html#cb55-1"></a><span class="kw">plot_norm_changes</span>(k_deseq_rle, k_raw, </span>
<span id="cb55-2"><a href="library-size-and-composition-scaling-methods.html#cb55-2"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;Raw&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;RLE&quot;</span>, </span>
<span id="cb55-3"><a href="library-size-and-composition-scaling-methods.html#cb55-3"></a>                  <span class="dt">title =</span> <span class="st">&quot;Distance metric comparision between RLE normalization and Raw counts &quot;</span>) <span class="op">/</span><span class="st"> </span></span>
<span id="cb55-4"><a href="library-size-and-composition-scaling-methods.html#cb55-4"></a><span class="kw">plot_norm_changes</span>(k_deseq_vs, k_raw, </span>
<span id="cb55-5"><a href="library-size-and-composition-scaling-methods.html#cb55-5"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;Raw&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;VST&quot;</span>, </span>
<span id="cb55-6"><a href="library-size-and-composition-scaling-methods.html#cb55-6"></a>                  <span class="dt">title =</span> <span class="st">&quot;Distance metric comparision between VST normalization and Raw counts &quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb55-7"><a href="library-size-and-composition-scaling-methods.html#cb55-7"></a><span class="st">  </span><span class="kw">plot_layout</span>(<span class="dt">guides =</span> <span class="st">&#39;collect&#39;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-28-2.png" width="672" /></p>
</div>
</div>
</div>
<div id="gmpr" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> GMPR</h2>
<div id="about-gmpr" class="section level3" number="5.2.1">
<h3><span class="header-section-number">5.2.1</span> About GMPR</h3>
<p>A recent extension of the RLE DESeq method is the Geometric mean of Pairwise ratios (GMPR) approach <span class="citation">(Chen et al. <a href="#ref-chen2018" role="doc-biblioref">2018</a>)</span>. This method reverses the steps of RLE, and instead calculates the median count ratio of the non-zero counts between pairs of samples as although only a small number of taxa are likely to be shared for every sample, it is more likely that there are many shared taxa between pairs. It then uses the pairwise results to calculate the size factor for each sample. This method has slow computation, but is robust to differential and outlier taxa. The size factors can be inputted to DESeq and a VST transformation applied additionally. It is a newer method, and has unfortunately not been included in many benchmarking studies, although initial results show it to be more powerful than DESeq, not surprisingly, as it uses more data, as zero counts do not need to be discarded. It assumes there is a large invariant portion of the count data, similar to other methods.</p>
</div>
<div id="gmpr-walkthough" class="section level3" number="5.2.2">
<h3><span class="header-section-number">5.2.2</span> GMPR Walkthough</h3>
<p>Consider the following zero-inflated dataset. Notice that there are no taxa that are present in every sample.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-29">Table 5.7: </span>Example dataset - Raw Counts
</caption>
<thead>
<tr>
<th style="text-align:left;">
Taxa
</th>
<th style="text-align:right;">
Sample 1
</th>
<th style="text-align:right;">
Sample2
</th>
<th style="text-align:right;">
Sample 3
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Taxon 1
</td>
<td style="text-align:right;">
132
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
48
</td>
<td style="text-align:right;">
74
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 4
</td>
<td style="text-align:right;">
23
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
35
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 5
</td>
<td style="text-align:right;">
71
</td>
<td style="text-align:right;">
80
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 6
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
96
</td>
<td style="text-align:right;">
82
</td>
</tr>
</tbody>
</table>
<p>The first step in GMPR normalization is to calculate the pairwise median count ratio between samples. We first calculate the scaling factor for sample 1. The pairwise comparisons we need to make are between sample 1 and sample 2 as well as sample 1 and sample 3. For both pairs, we calculate the ratio of the counts of taxa that the pair shares. Between sample 1 and sample 2, the shared taxa are 1 and 5. Between sample 1 and sample 3, taxon 4 is the only one shared.</p>
<table>
<caption>
<span id="tab:unnamed-chunk-30">Table 5.8: </span>Example dataset - Raw Counts
</caption>
<thead>
<tr>
<th style="text-align:left;">
Taxa
</th>
<th style="text-align:right;">
Sample 1
</th>
<th style="text-align:right;">
Sample2
</th>
<th style="text-align:right;">
Sample 3
</th>
<th style="text-align:left;">
Count ratio between 1 and 2
</th>
<th style="text-align:left;">
Count ratio between 1 and 3
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Taxon 1
</td>
<td style="text-align:right;">
132
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
132/103
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
48
</td>
<td style="text-align:right;">
74
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 4
</td>
<td style="text-align:right;">
23
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
35
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
23/35
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 5
</td>
<td style="text-align:right;">
71
</td>
<td style="text-align:right;">
80
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
71/80
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
Taxon 6
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
96
</td>
<td style="text-align:right;">
82
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
</tr>
</tbody>
</table>
<p>Then we calculate the median of the ratios between each pair. Between sample 1 and 2, the two ratios between shared taxa are (132/103, 71/80). Then the median of those is 1.085. Between sample 1 and 3, there is only one shared taxa, so the median is 0.657.</p>
<p>Finally to find the scaling factor for sample 1, we calculate the geometric mean of the two medians of the pairwise shared taxa ratios. In this case it is the geometric mean of 1.085 and 0.657, which equals 0.844. This is the scaling factor for sample 1.
We now repeat this process to find the scaling factor for sample 1 and sample 3.</p>
</div>
<div id="gmpr-r-code" class="section level3" number="5.2.3">
<h3><span class="header-section-number">5.2.3</span> GMPR R Code</h3>
<div id="function-3" class="section level4" number="5.2.3.1">
<h4><span class="header-section-number">5.2.3.1</span> Function</h4>
<p>The following provides the R code to implement GMPR normalization in R. We can additionally specify the number of taxa that should be shared between paired samples as well as the minimum count labeled as nonzero. The default values are 4 and 2 respectively.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="library-size-and-composition-scaling-methods.html#cb56-1"></a>norm_GMPR &lt;-<span class="st"> </span><span class="cf">function</span>(ps, <span class="dt">intersect_no =</span> <span class="dv">4</span>, <span class="dt">min_ct =</span> <span class="dv">2</span>){</span>
<span id="cb56-2"><a href="library-size-and-composition-scaling-methods.html#cb56-2"></a>  <span class="kw">require</span>(GMPR, <span class="dt">quietly =</span> T)</span>
<span id="cb56-3"><a href="library-size-and-composition-scaling-methods.html#cb56-3"></a>    <span class="co"># Convert data to correct format for GMPR function </span></span>
<span id="cb56-4"><a href="library-size-and-composition-scaling-methods.html#cb56-4"></a>    otu &lt;-<span class="st"> </span><span class="kw">as</span>(<span class="kw">otu_table</span>(ps), <span class="st">&quot;matrix&quot;</span>)</span>
<span id="cb56-5"><a href="library-size-and-composition-scaling-methods.html#cb56-5"></a>    <span class="cf">if</span>(<span class="kw">taxa_are_rows</span>(ps)){otu &lt;-<span class="st"> </span><span class="kw">t</span>(otu)}</span>
<span id="cb56-6"><a href="library-size-and-composition-scaling-methods.html#cb56-6"></a>    otu_df =<span class="st"> </span><span class="kw">as.data.frame</span>(otu)</span>
<span id="cb56-7"><a href="library-size-and-composition-scaling-methods.html#cb56-7"></a>    otu.tab &lt;-<span class="st"> </span><span class="kw">matrix</span>(otu, <span class="dt">ncol =</span> <span class="kw">ncol</span>(otu))</span>
<span id="cb56-8"><a href="library-size-and-composition-scaling-methods.html#cb56-8"></a>    <span class="co"># calculate scaling factor</span></span>
<span id="cb56-9"><a href="library-size-and-composition-scaling-methods.html#cb56-9"></a>    <span class="co"># OTU matrix must be a data frame where OTUs are arranged in columns and samples as rows</span></span>
<span id="cb56-10"><a href="library-size-and-composition-scaling-methods.html#cb56-10"></a>    gmpr.size.factor &lt;-<span class="st"> </span>GMPR<span class="op">::</span><span class="kw">GMPR</span>(<span class="dt">OTUmatrix =</span> otu_df,</span>
<span id="cb56-11"><a href="library-size-and-composition-scaling-methods.html#cb56-11"></a>                                   <span class="dt">intersect_no =</span> intersect_no,</span>
<span id="cb56-12"><a href="library-size-and-composition-scaling-methods.html#cb56-12"></a>                                   <span class="dt">min_ct =</span> min_ct)</span>
<span id="cb56-13"><a href="library-size-and-composition-scaling-methods.html#cb56-13"></a>    <span class="co"># normalize</span></span>
<span id="cb56-14"><a href="library-size-and-composition-scaling-methods.html#cb56-14"></a>    otu.tab.norm &lt;-<span class="st"> </span><span class="kw">t</span>(otu <span class="op">/</span><span class="st"> </span>(gmpr.size.factor))</span>
<span id="cb56-15"><a href="library-size-and-composition-scaling-methods.html#cb56-15"></a>    <span class="co"># convert back to PS</span></span>
<span id="cb56-16"><a href="library-size-and-composition-scaling-methods.html#cb56-16"></a>    otu_ps &lt;-<span class="st"> </span><span class="kw">otu_table</span>(otu.tab.norm, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb56-17"><a href="library-size-and-composition-scaling-methods.html#cb56-17"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb56-18"><a href="library-size-and-composition-scaling-methods.html#cb56-18"></a>    sam<span class="op">$</span>scaling_factor &lt;-<span class="st"> </span>gmpr.size.factor</span>
<span id="cb56-19"><a href="library-size-and-composition-scaling-methods.html#cb56-19"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb56-20"><a href="library-size-and-composition-scaling-methods.html#cb56-20"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb56-21"><a href="library-size-and-composition-scaling-methods.html#cb56-21"></a>    ps_GMPR &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu_ps,sam,tax,phy)</span>
<span id="cb56-22"><a href="library-size-and-composition-scaling-methods.html#cb56-22"></a>    <span class="kw">return</span>(ps_GMPR)</span>
<span id="cb56-23"><a href="library-size-and-composition-scaling-methods.html#cb56-23"></a>}</span></code></pre></div>
</div>
<div id="gmpr-implementation-on-global-patterns" class="section level4" number="5.2.3.2">
<h4><span class="header-section-number">5.2.3.2</span> GMPR implementation on Global Patterns</h4>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="library-size-and-composition-scaling-methods.html#cb57-1"></a><span class="co"># Normalize</span></span>
<span id="cb57-2"><a href="library-size-and-composition-scaling-methods.html#cb57-2"></a>gp_gmpr &lt;-<span class="st"> </span><span class="kw">norm_GMPR</span>(gp_raw)</span>
<span id="cb57-3"><a href="library-size-and-composition-scaling-methods.html#cb57-3"></a></span>
<span id="cb57-4"><a href="library-size-and-composition-scaling-methods.html#cb57-4"></a><span class="co"># Compare to other methods like before</span></span>
<span id="cb57-5"><a href="library-size-and-composition-scaling-methods.html#cb57-5"></a><span class="kw">plot_norm_changes</span>(gp_gmpr, gp_raw,</span>
<span id="cb57-6"><a href="library-size-and-composition-scaling-methods.html#cb57-6"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;GMPR&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;Raw Counts&quot;</span>,</span>
<span id="cb57-7"><a href="library-size-and-composition-scaling-methods.html#cb57-7"></a>                  <span class="dt">title =</span> <span class="st">&quot;GMPR normalization vs Raw Counts&quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-31-1.png" width="672" /></p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="library-size-and-composition-scaling-methods.html#cb58-1"></a><span class="kw">plot_norm_changes</span>(gp_gmpr, gp_tss,</span>
<span id="cb58-2"><a href="library-size-and-composition-scaling-methods.html#cb58-2"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;TSS&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;GMPR&quot;</span>,</span>
<span id="cb58-3"><a href="library-size-and-composition-scaling-methods.html#cb58-3"></a>                  <span class="dt">title =</span> <span class="st">&quot;TSS vs GMPR&quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-31-2.png" width="672" /></p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="library-size-and-composition-scaling-methods.html#cb59-1"></a><span class="kw">plot_norm_changes</span>(gp_gmpr, gp_deseq_rle,</span>
<span id="cb59-2"><a href="library-size-and-composition-scaling-methods.html#cb59-2"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;DESeq&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;GMPR&quot;</span>,</span>
<span id="cb59-3"><a href="library-size-and-composition-scaling-methods.html#cb59-3"></a>                  <span class="dt">title =</span> <span class="st">&quot;DESeq vs GMPR&quot;</span> )</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-31-3.png" width="672" /></p>
</div>
</div>
</div>
<div id="tmm-edger" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> TMM (edgeR)</h2>
<div id="about-tmm" class="section level3" number="5.3.1">
<h3><span class="header-section-number">5.3.1</span> About TMM</h3>
<p>TMM (Trimmed median of m-values) is another method borrowed from RNA-Seq analysis, and implemented in <code>edgeR</code> <span class="citation">(Robinson and Oshlack <a href="#ref-robinson2010" role="doc-biblioref">2010</a>)</span>. This method uses, or calculates a reference sample, and compares all other samples to the reference sample. The size factor is the mean of the log-ratios after excluding the highest count taxa and taxa with the largest fold change. As taxa with zero counts are excluded, a pseudo count is needed. Additionally, there is the <code>TMMwsp</code> option which is encouraged as it is more robust to zero counts. Positive counts are reused to increase the number of features when we compared. The singleton positive counts are paired up in decreasing order of size and then a modified TMM method is applied to the re-ordered libraries.</p>
</div>
<div id="edger-tmm-t-code" class="section level3" number="5.3.2">
<h3><span class="header-section-number">5.3.2</span> EdgeR TMM T code</h3>
<div id="function-4" class="section level4" number="5.3.2.1">
<h4><span class="header-section-number">5.3.2.1</span> Function</h4>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="library-size-and-composition-scaling-methods.html#cb60-1"></a>norm_TMM &lt;-<span class="st"> </span><span class="cf">function</span>(physeq, <span class="dt">group =</span> <span class="dv">1</span>, <span class="dt">method=</span><span class="st">&quot;TMM&quot;</span>, <span class="dt">pseudocount =</span> <span class="dv">1</span>, ...){</span>
<span id="cb60-2"><a href="library-size-and-composition-scaling-methods.html#cb60-2"></a>    <span class="kw">require</span>(<span class="st">&quot;edgeR&quot;</span>, <span class="dt">quietly =</span> T)</span>
<span id="cb60-3"><a href="library-size-and-composition-scaling-methods.html#cb60-3"></a>    <span class="kw">require</span>(<span class="st">&quot;phyloseq&quot;</span>, <span class="dt">quietly =</span> T)</span>
<span id="cb60-4"><a href="library-size-and-composition-scaling-methods.html#cb60-4"></a>    <span class="co"># Enforce orientation.</span></span>
<span id="cb60-5"><a href="library-size-and-composition-scaling-methods.html#cb60-5"></a>    <span class="cf">if</span>( <span class="op">!</span><span class="kw">taxa_are_rows</span>(physeq) ){ physeq &lt;-<span class="st"> </span><span class="kw">t</span>(physeq) }</span>
<span id="cb60-6"><a href="library-size-and-composition-scaling-methods.html#cb60-6"></a>    x =<span class="st"> </span><span class="kw">as</span>(<span class="kw">otu_table</span>(physeq), <span class="st">&quot;matrix&quot;</span>)</span>
<span id="cb60-7"><a href="library-size-and-composition-scaling-methods.html#cb60-7"></a>    <span class="co"># Add one to protect against overflow, log(0) issues.</span></span>
<span id="cb60-8"><a href="library-size-and-composition-scaling-methods.html#cb60-8"></a>    x =<span class="st"> </span>x <span class="op">+</span><span class="st"> </span>pseudocount</span>
<span id="cb60-9"><a href="library-size-and-composition-scaling-methods.html#cb60-9"></a>    <span class="co"># Check `group` argument</span></span>
<span id="cb60-10"><a href="library-size-and-composition-scaling-methods.html#cb60-10"></a>    <span class="cf">if</span>( <span class="kw">identical</span>(<span class="kw">all.equal</span>(<span class="kw">length</span>(group), <span class="dv">1</span>), <span class="ot">TRUE</span>) <span class="op">&amp;</span><span class="st"> </span><span class="kw">nsamples</span>(physeq) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span> ){</span>
<span id="cb60-11"><a href="library-size-and-composition-scaling-methods.html#cb60-11"></a>        <span class="co"># Assume that group was a sample variable name (must be categorical)</span></span>
<span id="cb60-12"><a href="library-size-and-composition-scaling-methods.html#cb60-12"></a>        group =<span class="st"> </span><span class="kw">get_variable</span>(physeq, group)</span>
<span id="cb60-13"><a href="library-size-and-composition-scaling-methods.html#cb60-13"></a>    }</span>
<span id="cb60-14"><a href="library-size-and-composition-scaling-methods.html#cb60-14"></a>    <span class="co"># Define gene annotations (`genes`) as tax_table</span></span>
<span id="cb60-15"><a href="library-size-and-composition-scaling-methods.html#cb60-15"></a>    taxonomy =<span class="st"> </span><span class="kw">tax_table</span>(physeq, <span class="dt">errorIfNULL=</span><span class="ot">FALSE</span>)</span>
<span id="cb60-16"><a href="library-size-and-composition-scaling-methods.html#cb60-16"></a>    <span class="cf">if</span>( <span class="op">!</span><span class="kw">is.null</span>(taxonomy) ){</span>
<span id="cb60-17"><a href="library-size-and-composition-scaling-methods.html#cb60-17"></a>        taxonomy =<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">as</span>(taxonomy, <span class="st">&quot;matrix&quot;</span>))</span>
<span id="cb60-18"><a href="library-size-and-composition-scaling-methods.html#cb60-18"></a>    } </span>
<span id="cb60-19"><a href="library-size-and-composition-scaling-methods.html#cb60-19"></a>    <span class="co"># Now turn into a DGEList</span></span>
<span id="cb60-20"><a href="library-size-and-composition-scaling-methods.html#cb60-20"></a>    y =<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">counts=</span>x, <span class="dt">group=</span>group, <span class="dt">genes=</span>taxonomy, <span class="dt">remove.zeros =</span> <span class="ot">TRUE</span>)</span>
<span id="cb60-21"><a href="library-size-and-composition-scaling-methods.html#cb60-21"></a>    <span class="co"># Calculate the normalization factors</span></span>
<span id="cb60-22"><a href="library-size-and-composition-scaling-methods.html#cb60-22"></a>    d =<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">calcNormFactors</span>(y, <span class="dt">method=</span>method)</span>
<span id="cb60-23"><a href="library-size-and-composition-scaling-methods.html#cb60-23"></a>    <span class="co"># Check for division by zero inside `calcNormFactors`</span></span>
<span id="cb60-24"><a href="library-size-and-composition-scaling-methods.html#cb60-24"></a>    <span class="cf">if</span>( <span class="op">!</span><span class="kw">all</span>(<span class="kw">is.finite</span>(d<span class="op">$</span>samples<span class="op">$</span>norm.factors)) ){</span>
<span id="cb60-25"><a href="library-size-and-composition-scaling-methods.html#cb60-25"></a>        <span class="kw">stop</span>(<span class="st">&quot;Something wrong with edgeR::calcNormFactors on this data,</span></span>
<span id="cb60-26"><a href="library-size-and-composition-scaling-methods.html#cb60-26"></a><span class="st">         non-finite $norm.factors, consider changing `method` argument&quot;</span>)</span>
<span id="cb60-27"><a href="library-size-and-composition-scaling-methods.html#cb60-27"></a>    }</span>
<span id="cb60-28"><a href="library-size-and-composition-scaling-methods.html#cb60-28"></a>    scalingFactor &lt;-<span class="st"> </span>d<span class="op">$</span>samples<span class="op">$</span>norm.factors <span class="op">*</span><span class="st"> </span>d<span class="op">$</span>samples<span class="op">$</span>lib.size <span class="op">/</span><span class="st"> </span><span class="fl">1e6</span></span>
<span id="cb60-29"><a href="library-size-and-composition-scaling-methods.html#cb60-29"></a>    dataNormalized &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(<span class="kw">otu_table</span>(physeq)) <span class="op">/</span><span class="st"> </span>scalingFactor)</span>
<span id="cb60-30"><a href="library-size-and-composition-scaling-methods.html#cb60-30"></a>    <span class="co">#dataNormalized &lt;- cpm(d)</span></span>
<span id="cb60-31"><a href="library-size-and-composition-scaling-methods.html#cb60-31"></a>    otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(dataNormalized, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb60-32"><a href="library-size-and-composition-scaling-methods.html#cb60-32"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(physeq, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb60-33"><a href="library-size-and-composition-scaling-methods.html#cb60-33"></a>    sam<span class="op">$</span>scaling_factor &lt;-<span class="st"> </span>scalingFactor</span>
<span id="cb60-34"><a href="library-size-and-composition-scaling-methods.html#cb60-34"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(physeq, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb60-35"><a href="library-size-and-composition-scaling-methods.html#cb60-35"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(physeq, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb60-36"><a href="library-size-and-composition-scaling-methods.html#cb60-36"></a>    ps_edgeR &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu,sam,tax,phy)</span>
<span id="cb60-37"><a href="library-size-and-composition-scaling-methods.html#cb60-37"></a>    </span>
<span id="cb60-38"><a href="library-size-and-composition-scaling-methods.html#cb60-38"></a>    <span class="kw">return</span>(ps_edgeR)</span>
<span id="cb60-39"><a href="library-size-and-composition-scaling-methods.html#cb60-39"></a>    </span>
<span id="cb60-40"><a href="library-size-and-composition-scaling-methods.html#cb60-40"></a>}</span></code></pre></div>
</div>
<div id="tmm-implementation-on-global-patterns" class="section level4" number="5.3.2.2">
<h4><span class="header-section-number">5.3.2.2</span> TMM implementation on Global Patterns</h4>
<p>Perform normalization:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="library-size-and-composition-scaling-methods.html#cb61-1"></a>gp_tmm &lt;-<span class="st"> </span><span class="kw">norm_TMM</span>(gp_raw)</span></code></pre></div>
<p>View PCoA plots</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="library-size-and-composition-scaling-methods.html#cb62-1"></a><span class="kw">plot_ordination</span>(gp_tmm, phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_tmm, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) , <span class="dt">color =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw data&quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-34-1.png" width="672" /></p>
<p>View how TMM normalization changes distance metrics differently than raw counts.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="library-size-and-composition-scaling-methods.html#cb63-1"></a><span class="kw">plot_norm_changes</span>(gp_tmm, gp_raw,</span>
<span id="cb63-2"><a href="library-size-and-composition-scaling-methods.html#cb63-2"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;Raw&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;TMM&quot;</span>,</span>
<span id="cb63-3"><a href="library-size-and-composition-scaling-methods.html#cb63-3"></a>                  <span class="dt">title =</span> <span class="st">&quot;Distance metric comparision between TMM normalization and Raw counts &quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-35-1.png" width="672" /></p>
</div>
<div id="tmm-implementation-on-kostic-data" class="section level4" number="5.3.2.3">
<h4><span class="header-section-number">5.3.2.3</span> TMM implementation on Kostic data</h4>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="library-size-and-composition-scaling-methods.html#cb64-1"></a>k_tmm &lt;-<span class="st"> </span><span class="kw">norm_TMM</span>(k_raw)</span>
<span id="cb64-2"><a href="library-size-and-composition-scaling-methods.html#cb64-2"></a><span class="kw">plot_ordination</span>(</span>
<span id="cb64-3"><a href="library-size-and-composition-scaling-methods.html#cb64-3"></a>    k_tmm,</span>
<span id="cb64-4"><a href="library-size-and-composition-scaling-methods.html#cb64-4"></a>    phyloseq<span class="op">::</span><span class="kw">ordinate</span>(k_tmm, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) ,</span>
<span id="cb64-5"><a href="library-size-and-composition-scaling-methods.html#cb64-5"></a>    <span class="dt">color =</span> <span class="st">&quot;DIAGNOSIS&quot;</span>,</span>
<span id="cb64-6"><a href="library-size-and-composition-scaling-methods.html#cb64-6"></a>    <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw data&quot;</span></span>
<span id="cb64-7"><a href="library-size-and-composition-scaling-methods.html#cb64-7"></a>) <span class="op">/</span></span>
<span id="cb64-8"><a href="library-size-and-composition-scaling-methods.html#cb64-8"></a><span class="st">    </span><span class="kw">plot_ordination</span>(</span>
<span id="cb64-9"><a href="library-size-and-composition-scaling-methods.html#cb64-9"></a>        k_tmm,</span>
<span id="cb64-10"><a href="library-size-and-composition-scaling-methods.html#cb64-10"></a>        phyloseq<span class="op">::</span><span class="kw">ordinate</span>(k_tmm, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) ,</span>
<span id="cb64-11"><a href="library-size-and-composition-scaling-methods.html#cb64-11"></a>        <span class="dt">color =</span> <span class="st">&quot;depth&quot;</span>,</span>
<span id="cb64-12"><a href="library-size-and-composition-scaling-methods.html#cb64-12"></a>        <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw data&quot;</span></span>
<span id="cb64-13"><a href="library-size-and-composition-scaling-methods.html#cb64-13"></a>    )</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/edgeR-kostic-1.png" width="672" /></p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="library-size-and-composition-scaling-methods.html#cb65-1"></a><span class="kw">plot_norm_changes</span>(k_tmm, k_raw,</span>
<span id="cb65-2"><a href="library-size-and-composition-scaling-methods.html#cb65-2"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;Raw&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;TMM&quot;</span>,</span>
<span id="cb65-3"><a href="library-size-and-composition-scaling-methods.html#cb65-3"></a>                  <span class="dt">title =</span> <span class="st">&quot;Distance metric comparision between TMM normalization and Raw counts &quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/edgeR-kostic-2.png" width="672" /></p>

</div>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references hanging-indent">
<div id="ref-anders2010">
<p>Anders, Simon, and Wolfgang Huber. 2010. “Differential Expression Analysis for Sequence Count Data.” <em>Genome Biology</em> 11 (10): R106. <a href="https://doi.org/10.1186/gb-2010-11-10-r106">https://doi.org/10.1186/gb-2010-11-10-r106</a>.</p>
</div>
<div id="ref-chen2018">
<p>Chen, Li, James Reeve, Lujun Zhang, Shengbing Huang, Xuefeng Wang, and Jun Chen. 2018. “GMPR: A Robust Normalization Method for Zero-Inflated Count Data with Application to Microbiome Sequencing Data.” <em>PeerJ</em> 6 (April): e4600. <a href="https://doi.org/10.7717/peerj.4600">https://doi.org/10.7717/peerj.4600</a>.</p>
</div>
<div id="ref-robinson2010">
<p>Robinson, Mark D., and Alicia Oshlack. 2010. “A Scaling Normalization Method for Differential Expression Analysis of Rna-Seq Data.” <em>Genome Biology</em> 11 (3): R25. <a href="https://doi.org/10.1186/gb-2010-11-3-r25">https://doi.org/10.1186/gb-2010-11-3-r25</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="subsampling-methods.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="wrench.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MB_normalization_review.pdf", "MB_normalization_review.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
