<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Log Ratio Methods | Tutorial on Normalization of Microbiome Data</title>
  <meta name="description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Log Ratio Methods | Tutorial on Normalization of Microbiome Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Log Ratio Methods | Tutorial on Normalization of Microbiome Data" />
  
  <meta name="twitter:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="wrench.html"/>
<link rel="next" href="comparisions.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">MB Normalization</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#the-importance-of-normalization"><i class="fa fa-check"></i><b>1.1</b> The importance of normalization</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-compositional-nature-of-microbiome-data"><i class="fa fa-check"></i><b>1.2</b> The compositional nature of microbiome data</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#log-ratio-methods"><i class="fa fa-check"></i><b>1.2.1</b> Log ratio methods</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>2</b> Importing Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-data.html"><a href="importing-data.html#global-patterns"><i class="fa fa-check"></i><b>2.1</b> Global Patterns</a></li>
<li class="chapter" data-level="2.2" data-path="importing-data.html"><a href="importing-data.html#colorectal-cancer"><i class="fa fa-check"></i><b>2.2</b> Colorectal cancer</a></li>
<li class="chapter" data-level="2.3" data-path="importing-data.html"><a href="importing-data.html#pre-processing-quality-control-and-filtering"><i class="fa fa-check"></i><b>2.3</b> Pre-processing Quality Control and Filtering</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html"><i class="fa fa-check"></i><b>3</b> Total Sum scaling (TSS)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#about-tss"><i class="fa fa-check"></i><b>3.1</b> About TSS</a></li>
<li class="chapter" data-level="3.2" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-implementation"><i class="fa fa-check"></i><b>3.2</b> TSS Implementation</a></li>
<li class="chapter" data-level="3.3" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-on-global-patterns"><i class="fa fa-check"></i><b>3.3</b> TSS on Global Patterns</a></li>
<li class="chapter" data-level="3.4" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-on-kostic-data"><i class="fa fa-check"></i><b>3.4</b> TSS on Kostic data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="rarefying.html"><a href="rarefying.html"><i class="fa fa-check"></i><b>4</b> Rarefying</a>
<ul>
<li class="chapter" data-level="4.1" data-path="rarefying.html"><a href="rarefying.html#about-rarefying"><i class="fa fa-check"></i><b>4.1</b> About Rarefying</a></li>
<li class="chapter" data-level="4.2" data-path="rarefying.html"><a href="rarefying.html#rarefying-implementation"><i class="fa fa-check"></i><b>4.2</b> Rarefying implementation</a></li>
<li class="chapter" data-level="4.3" data-path="rarefying.html"><a href="rarefying.html#rarefying-on-global-patterns"><i class="fa fa-check"></i><b>4.3</b> Rarefying on Global Patterns</a></li>
<li class="chapter" data-level="4.4" data-path="rarefying.html"><a href="rarefying.html#rarefying-on-kostic-data"><i class="fa fa-check"></i><b>4.4</b> Rarefying on Kostic Data</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="deseq.html"><a href="deseq.html"><i class="fa fa-check"></i><b>5</b> DESeq</a>
<ul>
<li class="chapter" data-level="5.1" data-path="deseq.html"><a href="deseq.html#about-deseq"><i class="fa fa-check"></i><b>5.1</b> About DESeq</a></li>
<li class="chapter" data-level="5.2" data-path="deseq.html"><a href="deseq.html#deseq-implementation"><i class="fa fa-check"></i><b>5.2</b> DESeq Implementation</a></li>
<li class="chapter" data-level="5.3" data-path="deseq.html"><a href="deseq.html#deseq-on-global-patterns"><i class="fa fa-check"></i><b>5.3</b> DESeq on Global Patterns</a></li>
<li class="chapter" data-level="5.4" data-path="deseq.html"><a href="deseq.html#deseq-on-kostic-data"><i class="fa fa-check"></i><b>5.4</b> DESeq on Kostic data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tmm-edger.html"><a href="tmm-edger.html"><i class="fa fa-check"></i><b>6</b> TMM (edgeR)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="tmm-edger.html"><a href="tmm-edger.html#edger-tmm-implementation"><i class="fa fa-check"></i><b>6.1</b> EdgeR TMM implementation</a></li>
<li class="chapter" data-level="6.2" data-path="tmm-edger.html"><a href="tmm-edger.html#tmm-on-global-patterns"><i class="fa fa-check"></i><b>6.2</b> TMM on Global Patterns</a></li>
<li class="chapter" data-level="6.3" data-path="tmm-edger.html"><a href="tmm-edger.html#tmm-on-kostic-data"><i class="fa fa-check"></i><b>6.3</b> TMM on Kostic data</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html"><i class="fa fa-check"></i><b>7</b> Cumulative sum scaling (CSS)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-implementation"><i class="fa fa-check"></i><b>7.1</b> CSS implementation</a></li>
<li class="chapter" data-level="7.2" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-on-global-patterns"><i class="fa fa-check"></i><b>7.2</b> CSS on Global Patterns</a></li>
<li class="chapter" data-level="7.3" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-on-kostic-data"><i class="fa fa-check"></i><b>7.3</b> CSS on Kostic data</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="gmpr.html"><a href="gmpr.html"><i class="fa fa-check"></i><b>8</b> GMPR</a>
<ul>
<li class="chapter" data-level="8.1" data-path="gmpr.html"><a href="gmpr.html#gmpr-implementation"><i class="fa fa-check"></i><b>8.1</b> GMPR Implementation</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="wrench.html"><a href="wrench.html"><i class="fa fa-check"></i><b>9</b> Wrench</a>
<ul>
<li class="chapter" data-level="9.1" data-path="wrench.html"><a href="wrench.html#wrench-implementation"><i class="fa fa-check"></i><b>9.1</b> Wrench Implementation</a></li>
<li class="chapter" data-level="9.2" data-path="wrench.html"><a href="wrench.html#wrench-on-global-patterns"><i class="fa fa-check"></i><b>9.2</b> Wrench on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="log-ratio-methods-1.html"><a href="log-ratio-methods-1.html"><i class="fa fa-check"></i><b>10</b> Log Ratio Methods</a>
<ul>
<li class="chapter" data-level="10.1" data-path="log-ratio-methods-1.html"><a href="log-ratio-methods-1.html#ancom-ii"><i class="fa fa-check"></i><b>10.1</b> ANCOM II</a></li>
<li class="chapter" data-level="10.2" data-path="log-ratio-methods-1.html"><a href="log-ratio-methods-1.html#ancom-bc"><i class="fa fa-check"></i><b>10.2</b> ANCOM BC</a></li>
<li class="chapter" data-level="10.3" data-path="log-ratio-methods-1.html"><a href="log-ratio-methods-1.html#aldex2"><i class="fa fa-check"></i><b>10.3</b> ALDEx2</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="comparisions.html"><a href="comparisions.html"><i class="fa fa-check"></i><b>11</b> Comparisions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial on Normalization of Microbiome Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="log-ratio-methods-1" class="section level1" number="10">
<h1><span class="header-section-number">Chapter 10</span> Log Ratio Methods</h1>
<div id="ancom-ii" class="section level2" number="10.1">
<h2><span class="header-section-number">10.1</span> ANCOM II</h2>
<p>ANCOM-II <span class="citation">(Kaul et al. <a href="#ref-kaul2017" role="doc-biblioref">2017</a>)</span> is a log-ratio approach for accounting for zero-inflation in log-ratio transformations. It uses a log-ratio transformation instead of library size normalization. If there are <span class="math inline">\(m\)</span> total taxa, this method performs <span class="math inline">\(m-1\)</span> differential abundance tests, each one using a different taxa as the reference taxa. Before the alr log-ratio transformation is performed, this method identifies and adjusts for three types of zeros that can exist. These three types of zeros are outlier zeros, structural zeros, and sampling zeros. Outlier zeros are zeros caused from some extraneous reasons separate from below detection limits due to sampling depth. To account for these zeros, they are replaced by NA in the data. Secondly, structural zeros are zeros which are zero due to the nature of the groups. If a taxa is structurally zero in a group, it is marked as automatically differentially abundant, and removed from further analysis. Finally, sampling zeros are other zeros that are zero perhaps caused by sampling depth. These are imputed by a small pseudocount. This method thus takes into account zero-inflation and compositionality. <!-- After this process, no numerical zeros remain, so log and log-ratio transformations can be performed. This method thus takes into account zero-inflation and compositionality.  --></p>
<p>After this correction of zeros, the data are log ratio transformed using each possible taxa as the reference sample. Tests for differential abundance are performed on each of iterations of the log-ratio transformed data, and the count of times the resulting test is significant is used to determining overall significance.</p>
<!-- This method is shown to be conservative,  -->
<!-- Sensitive, good control of FDR -->
<p>It is possible that no outlier or structural zeros are detected, so as a normalization/transformation technique, this method can be equivalent to alr or clr transoformation on raw data with a pseudocount.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="log-ratio-methods-1.html#cb85-1"></a><span class="co">## Load the Ancom function (it&#39;s not a package)</span></span>
<span id="cb85-2"><a href="log-ratio-methods-1.html#cb85-2"></a>devtools<span class="op">::</span><span class="kw">source_url</span>(<span class="st">&quot;https://raw.githubusercontent.com/FrederickHuangLin/ANCOM/master/scripts/ancom_v2.1.R&quot;</span>)</span>
<span id="cb85-3"><a href="log-ratio-methods-1.html#cb85-3"></a></span>
<span id="cb85-4"><a href="log-ratio-methods-1.html#cb85-4"></a>norm_ANCOM2 &lt;-<span class="st"> </span><span class="cf">function</span>(ps, group_var, <span class="dt">outlier.replace =</span> <span class="ot">TRUE</span>) {</span>
<span id="cb85-5"><a href="log-ratio-methods-1.html#cb85-5"></a>    feature_table &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">otu_table</span>(ps))</span>
<span id="cb85-6"><a href="log-ratio-methods-1.html#cb85-6"></a>    groups &lt;-<span class="st"> </span><span class="kw">as.factor</span>(<span class="kw">get_variable</span>(ps, group_var))</span>
<span id="cb85-7"><a href="log-ratio-methods-1.html#cb85-7"></a>    meta_data &lt;-</span>
<span id="cb85-8"><a href="log-ratio-methods-1.html#cb85-8"></a><span class="st">        </span><span class="kw">data.frame</span>(<span class="dt">sample_id =</span> <span class="kw">paste0</span>(<span class="st">&quot;sample&quot;</span>, <span class="kw">seq</span>(<span class="kw">dim</span>(raw)[<span class="dv">2</span>])), <span class="dt">groups =</span> groups)</span>
<span id="cb85-9"><a href="log-ratio-methods-1.html#cb85-9"></a>    <span class="co">#rownames(meta_data) &lt;- meta_data$sample_id</span></span>
<span id="cb85-10"><a href="log-ratio-methods-1.html#cb85-10"></a>    <span class="kw">colnames</span>(feature_table) &lt;-<span class="st"> </span>meta_data<span class="op">$</span>sample_id</span>
<span id="cb85-11"><a href="log-ratio-methods-1.html#cb85-11"></a>    </span>
<span id="cb85-12"><a href="log-ratio-methods-1.html#cb85-12"></a>    <span class="co"># The main ANCOM2 zero-classification function </span></span>
<span id="cb85-13"><a href="log-ratio-methods-1.html#cb85-13"></a>    prepro &lt;-</span>
<span id="cb85-14"><a href="log-ratio-methods-1.html#cb85-14"></a><span class="st">        </span><span class="kw">feature_table_pre_process</span>(</span>
<span id="cb85-15"><a href="log-ratio-methods-1.html#cb85-15"></a>            feature_table,</span>
<span id="cb85-16"><a href="log-ratio-methods-1.html#cb85-16"></a>            meta_data,</span>
<span id="cb85-17"><a href="log-ratio-methods-1.html#cb85-17"></a>            <span class="dt">sample_var =</span> <span class="st">&quot;sample_id&quot;</span>,</span>
<span id="cb85-18"><a href="log-ratio-methods-1.html#cb85-18"></a>            <span class="dt">group_var =</span> <span class="st">&quot;groups&quot;</span>,</span>
<span id="cb85-19"><a href="log-ratio-methods-1.html#cb85-19"></a>            <span class="dt">out_cut =</span> <span class="fl">0.05</span>,</span>
<span id="cb85-20"><a href="log-ratio-methods-1.html#cb85-20"></a>            <span class="dt">zero_cut =</span> <span class="fl">0.9</span>,</span>
<span id="cb85-21"><a href="log-ratio-methods-1.html#cb85-21"></a>            <span class="dt">lib_cut =</span> <span class="dv">0</span>,</span>
<span id="cb85-22"><a href="log-ratio-methods-1.html#cb85-22"></a>            <span class="dt">neg_lb =</span> <span class="ot">FALSE</span></span>
<span id="cb85-23"><a href="log-ratio-methods-1.html#cb85-23"></a>        )</span>
<span id="cb85-24"><a href="log-ratio-methods-1.html#cb85-24"></a>    feature_table &lt;-<span class="st"> </span>prepro<span class="op">$</span>feature_table</span>
<span id="cb85-25"><a href="log-ratio-methods-1.html#cb85-25"></a>    meta_data &lt;-<span class="st"> </span>prepro<span class="op">$</span>meta_data</span>
<span id="cb85-26"><a href="log-ratio-methods-1.html#cb85-26"></a>    struc_zero &lt;-<span class="st"> </span>prepro<span class="op">$</span>structure_zeros</span>
<span id="cb85-27"><a href="log-ratio-methods-1.html#cb85-27"></a>    </span>
<span id="cb85-28"><a href="log-ratio-methods-1.html#cb85-28"></a>    <span class="co"># OTU table transformation:</span></span>
<span id="cb85-29"><a href="log-ratio-methods-1.html#cb85-29"></a>    <span class="co"># (1) Discard taxa with structural zeros (if any); (2) Add pseudocount (1) and take logarithm.</span></span>
<span id="cb85-30"><a href="log-ratio-methods-1.html#cb85-30"></a>    <span class="cf">if</span> (<span class="op">!</span><span class="kw">is.null</span>(struc_zero)) {</span>
<span id="cb85-31"><a href="log-ratio-methods-1.html#cb85-31"></a>        num_struc_zero =<span class="st"> </span><span class="kw">apply</span>(struc_zero, <span class="dv">1</span>, sum)</span>
<span id="cb85-32"><a href="log-ratio-methods-1.html#cb85-32"></a>        comp_table =<span class="st"> </span>feature_table[num_struc_zero <span class="op">==</span><span class="st"> </span><span class="dv">0</span>,]</span>
<span id="cb85-33"><a href="log-ratio-methods-1.html#cb85-33"></a>    } <span class="cf">else</span>{</span>
<span id="cb85-34"><a href="log-ratio-methods-1.html#cb85-34"></a>        comp_table =<span class="st"> </span>feature_table</span>
<span id="cb85-35"><a href="log-ratio-methods-1.html#cb85-35"></a>    }</span>
<span id="cb85-36"><a href="log-ratio-methods-1.html#cb85-36"></a>    comp_table &lt;-<span class="st"> </span><span class="kw">log</span>(<span class="kw">as.matrix</span>(comp_table) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb85-37"><a href="log-ratio-methods-1.html#cb85-37"></a>    </span>
<span id="cb85-38"><a href="log-ratio-methods-1.html#cb85-38"></a>    <span class="co"># Replace outlier zeros with the mean of the rest genes in each</span></span>
<span id="cb85-39"><a href="log-ratio-methods-1.html#cb85-39"></a>    <span class="cf">if</span> (outlier.replace) {</span>
<span id="cb85-40"><a href="log-ratio-methods-1.html#cb85-40"></a>        <span class="cf">for</span> (col <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(comp_table)) {</span>
<span id="cb85-41"><a href="log-ratio-methods-1.html#cb85-41"></a>            comp_table[<span class="kw">is.na</span>(comp_table[, col]), col] &lt;-</span>
<span id="cb85-42"><a href="log-ratio-methods-1.html#cb85-42"></a><span class="st">                </span><span class="kw">mean</span>(comp_table[, col], <span class="dt">na.rm =</span> T)</span>
<span id="cb85-43"><a href="log-ratio-methods-1.html#cb85-43"></a>        }</span>
<span id="cb85-44"><a href="log-ratio-methods-1.html#cb85-44"></a>    }</span>
<span id="cb85-45"><a href="log-ratio-methods-1.html#cb85-45"></a>    <span class="co"># Calculate ratio</span></span>
<span id="cb85-46"><a href="log-ratio-methods-1.html#cb85-46"></a>    n_taxa &lt;-<span class="st"> </span><span class="kw">dim</span>(comp_table)[<span class="dv">1</span>]</span>
<span id="cb85-47"><a href="log-ratio-methods-1.html#cb85-47"></a>    taxa_id &lt;-<span class="st"> </span><span class="kw">rownames</span>(comp_table)</span>
<span id="cb85-48"><a href="log-ratio-methods-1.html#cb85-48"></a>    n_samp &lt;-<span class="st"> </span><span class="kw">dim</span>(comp_table)[<span class="dv">2</span>]</span>
<span id="cb85-49"><a href="log-ratio-methods-1.html#cb85-49"></a>    log_geo_mean &lt;-</span>
<span id="cb85-50"><a href="log-ratio-methods-1.html#cb85-50"></a><span class="st">        </span><span class="kw">apply</span>(comp_table, <span class="dv">2</span>, <span class="cf">function</span>(x) {</span>
<span id="cb85-51"><a href="log-ratio-methods-1.html#cb85-51"></a>            <span class="kw">mean</span>(x, <span class="dt">na.rm =</span> T)</span>
<span id="cb85-52"><a href="log-ratio-methods-1.html#cb85-52"></a>        })</span>
<span id="cb85-53"><a href="log-ratio-methods-1.html#cb85-53"></a>    mu_group_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">mean</span>(log_geo_mean[groups <span class="op">==</span><span class="st"> </span><span class="kw">levels</span>(groups)[<span class="dv">1</span>]])</span>
<span id="cb85-54"><a href="log-ratio-methods-1.html#cb85-54"></a>    mu_group_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw">mean</span>(log_geo_mean[groups <span class="op">==</span><span class="st"> </span><span class="kw">levels</span>(groups)[<span class="dv">2</span>]])</span>
<span id="cb85-55"><a href="log-ratio-methods-1.html#cb85-55"></a>    mu &lt;-<span class="st"> </span><span class="kw">rep</span>(mu_group_<span class="dv">1</span>, n_samp)</span>
<span id="cb85-56"><a href="log-ratio-methods-1.html#cb85-56"></a>    mu[groups <span class="op">==</span><span class="st"> </span><span class="kw">levels</span>(groups)[<span class="dv">2</span>]] &lt;-<span class="st"> </span>mu_group_<span class="dv">2</span></span>
<span id="cb85-57"><a href="log-ratio-methods-1.html#cb85-57"></a>    bias &lt;-<span class="st"> </span>log_geo_mean <span class="op">-</span><span class="st"> </span>mu</span>
<span id="cb85-58"><a href="log-ratio-methods-1.html#cb85-58"></a>    dat.normed &lt;-</span>
<span id="cb85-59"><a href="log-ratio-methods-1.html#cb85-59"></a><span class="st">        </span>comp_table <span class="op">-</span><span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rep</span>(bias, n_taxa), <span class="dt">nrow =</span> n_taxa, <span class="dt">byrow =</span> T)</span>
<span id="cb85-60"><a href="log-ratio-methods-1.html#cb85-60"></a>    <span class="kw">colnames</span>(dat.normed) &lt;-<span class="st"> </span><span class="kw">colnames</span>(raw)</span>
<span id="cb85-61"><a href="log-ratio-methods-1.html#cb85-61"></a>    </span>
<span id="cb85-62"><a href="log-ratio-methods-1.html#cb85-62"></a>    </span>
<span id="cb85-63"><a href="log-ratio-methods-1.html#cb85-63"></a>    <span class="co"># Convert to phyloseq</span></span>
<span id="cb85-64"><a href="log-ratio-methods-1.html#cb85-64"></a>    otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(dat.normed, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb85-65"><a href="log-ratio-methods-1.html#cb85-65"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb85-66"><a href="log-ratio-methods-1.html#cb85-66"></a>    sam<span class="op">$</span>scaling_factor &lt;-<span class="st"> </span><span class="kw">exp</span>(bias)</span>
<span id="cb85-67"><a href="log-ratio-methods-1.html#cb85-67"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb85-68"><a href="log-ratio-methods-1.html#cb85-68"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb85-69"><a href="log-ratio-methods-1.html#cb85-69"></a>    ps_ancom2 &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu, sam, tax, phy)</span>
<span id="cb85-70"><a href="log-ratio-methods-1.html#cb85-70"></a>    <span class="kw">return</span>(ps_ancom2)</span>
<span id="cb85-71"><a href="log-ratio-methods-1.html#cb85-71"></a>}</span>
<span id="cb85-72"><a href="log-ratio-methods-1.html#cb85-72"></a></span>
<span id="cb85-73"><a href="log-ratio-methods-1.html#cb85-73"></a></span>
<span id="cb85-74"><a href="log-ratio-methods-1.html#cb85-74"></a>k_ancom2 &lt;-<span class="st"> </span><span class="kw">norm_ANCOM2</span>(k_raw, <span class="st">&quot;DIAGNOSIS&quot;</span>)</span>
<span id="cb85-75"><a href="log-ratio-methods-1.html#cb85-75"></a>k_ancom2</span></code></pre></div>
</div>
<div id="ancom-bc" class="section level2" number="10.2">
<h2><span class="header-section-number">10.2</span> ANCOM BC</h2>
<p>ANCOM BC <span class="citation">(Lin and Peddada <a href="#ref-lin2020" role="doc-biblioref">2020</a>)</span> includes a normalization method that accounts for zero inflation and compositionality.</p>
<p>This method corrects bias by estimating a sampling fraction, which is…. It accounts for sampling fraction by introducing a sample-specific offset in a linear regression framework, estimated form observed data. The offset term is the bias correction, and the linear regression in log scale is analogous to the log ratio transformation to deal with compositionality.</p>
<p>This method also identifies and accounts for outlier zeros, structural zeros, and sampling zeros, using the same methodology as ANCOM2.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="log-ratio-methods-1.html#cb86-1"></a><span class="kw">library</span>(ANCOMBC)</span>
<span id="cb86-2"><a href="log-ratio-methods-1.html#cb86-2"></a>norm_ANCOMBC &lt;-<span class="st"> </span><span class="cf">function</span>(ps, group_var) {</span>
<span id="cb86-3"><a href="log-ratio-methods-1.html#cb86-3"></a>    out =<span class="st"> </span><span class="kw">ancombc</span>(ps,</span>
<span id="cb86-4"><a href="log-ratio-methods-1.html#cb86-4"></a>                  <span class="dt">formula =</span> group_var,</span>
<span id="cb86-5"><a href="log-ratio-methods-1.html#cb86-5"></a>                  <span class="dt">group =</span> group_var,</span>
<span id="cb86-6"><a href="log-ratio-methods-1.html#cb86-6"></a>                  <span class="dt">struc_zero =</span> T, </span>
<span id="cb86-7"><a href="log-ratio-methods-1.html#cb86-7"></a>                  <span class="dt">neg_lb =</span> T, </span>
<span id="cb86-8"><a href="log-ratio-methods-1.html#cb86-8"></a>                  <span class="dt">zero_cut =</span> <span class="fl">.9</span>)</span>
<span id="cb86-9"><a href="log-ratio-methods-1.html#cb86-9"></a>    <span class="co">#it is recommended to set neg_lb = TRUE when the sample size per group is relatively large (e.g. &gt; 30)</span></span>
<span id="cb86-10"><a href="log-ratio-methods-1.html#cb86-10"></a>    sampling.fraction &lt;-<span class="st"> </span><span class="kw">exp</span>(out<span class="op">$</span>samp_frac)</span>
<span id="cb86-11"><a href="log-ratio-methods-1.html#cb86-11"></a>    <span class="co"># Normalize counts</span></span>
<span id="cb86-12"><a href="log-ratio-methods-1.html#cb86-12"></a>    <span class="co"># feature_table &lt;- otu_table(ps, taxa_are_rows = T)</span></span>
<span id="cb86-13"><a href="log-ratio-methods-1.html#cb86-13"></a>    feature_table &lt;-<span class="st"> </span>out<span class="op">$</span>feature_table</span>
<span id="cb86-14"><a href="log-ratio-methods-1.html#cb86-14"></a>    <span class="co"># otu_norm &lt;- t(t(otu) / sampling.fraction)</span></span>
<span id="cb86-15"><a href="log-ratio-methods-1.html#cb86-15"></a>    otu_norm &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(feature_table) <span class="op">/</span><span class="st"> </span>sampling.fraction)</span>
<span id="cb86-16"><a href="log-ratio-methods-1.html#cb86-16"></a>    otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(otu_norm, T)</span>
<span id="cb86-17"><a href="log-ratio-methods-1.html#cb86-17"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb86-18"><a href="log-ratio-methods-1.html#cb86-18"></a>    sam<span class="op">$</span>scaling_factor &lt;-<span class="st"> </span>sampling.fraction</span>
<span id="cb86-19"><a href="log-ratio-methods-1.html#cb86-19"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb86-20"><a href="log-ratio-methods-1.html#cb86-20"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb86-21"><a href="log-ratio-methods-1.html#cb86-21"></a>    ps_ancombc &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu, sam, tax, phy)</span>
<span id="cb86-22"><a href="log-ratio-methods-1.html#cb86-22"></a>    <span class="kw">return</span>(ps_ancombc)</span>
<span id="cb86-23"><a href="log-ratio-methods-1.html#cb86-23"></a>}</span>
<span id="cb86-24"><a href="log-ratio-methods-1.html#cb86-24"></a></span>
<span id="cb86-25"><a href="log-ratio-methods-1.html#cb86-25"></a>k_ancombc &lt;-<span class="st"> </span><span class="kw">norm_ANCOMBC</span>(k_raw, <span class="dt">group_var =</span> <span class="st">&quot;DIAGNOSIS&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in data_prep(phyloseq, group, zero_cut, lib_cut, global = global):
## The multi-group comparison will be deactivated as the group variable has &lt; 3
## categories.</code></pre>
</div>
<div id="aldex2" class="section level2" number="10.3">
<h2><span class="header-section-number">10.3</span> ALDEx2</h2>
<!-- https://www.bioconductor.org/packages/devel/bioc/vignettes/ALDEx2/inst/doc/ALDEx2_vignette.html -->
<p>ALDEx2 <span class="citation">(Fernandes et al. <a href="#ref-fernandes2014" role="doc-biblioref">2014</a>)</span> is a method for differential abundance analysis developed for use with microbiome or any other sequencing setting. It uses a clr log-ratio transformation instead of library size normalization. To deal with sparsity, it creates randomized instances via Monte Carlo draws based on Dirichlet-Multinomial distribution with probabilities from the raw count data, These are zero-free, so log-ratio transformations are possible.</p>
<p>The idea is to estimate biological and sampling variation to calculate expected false discovery rate given variation. The method estimates technical variation within each sample using MC draws from Dirichlet distribution. For differential abundance, statistical tests are performed on each instance, p-values averaged, then corrected for multiple comparisons This method has been shown to be conservative compared to many of the library size normalization methods.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="log-ratio-methods-1.html#cb88-1"></a><span class="kw">library</span>(ALDEx2)</span></code></pre></div>
<pre><code>## Loading required package: zCompositions</code></pre>
<pre><code>## Loading required package: MASS</code></pre>
<pre><code>## 
## Attaching package: &#39;MASS&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:patchwork&#39;:
## 
##     area</code></pre>
<pre><code>## The following object is masked from &#39;package:dplyr&#39;:
## 
##     select</code></pre>
<pre><code>## Loading required package: NADA</code></pre>
<pre><code>## Loading required package: survival</code></pre>
<pre><code>## 
## Attaching package: &#39;NADA&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     cor</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     cor</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     cor</code></pre>
<pre><code>## Loading required package: truncnorm</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="log-ratio-methods-1.html#cb101-1"></a>norm_ALDEx2 &lt;-<span class="st"> </span><span class="cf">function</span>(ps, group_var) {</span>
<span id="cb101-2"><a href="log-ratio-methods-1.html#cb101-2"></a>    raw &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">otu_table</span>(ps))</span>
<span id="cb101-3"><a href="log-ratio-methods-1.html#cb101-3"></a>    groups &lt;-<span class="st"> </span><span class="kw">get_variable</span>(ps, group_var)</span>
<span id="cb101-4"><a href="log-ratio-methods-1.html#cb101-4"></a>    <span class="co"># Run ALDEx2 function</span></span>
<span id="cb101-5"><a href="log-ratio-methods-1.html#cb101-5"></a>    aldex.clr.res &lt;-</span>
<span id="cb101-6"><a href="log-ratio-methods-1.html#cb101-6"></a><span class="st">        </span><span class="kw">aldex.clr</span>(</span>
<span id="cb101-7"><a href="log-ratio-methods-1.html#cb101-7"></a>            <span class="dt">reads =</span> raw,</span>
<span id="cb101-8"><a href="log-ratio-methods-1.html#cb101-8"></a>            <span class="dt">conds =</span> groups,</span>
<span id="cb101-9"><a href="log-ratio-methods-1.html#cb101-9"></a>            <span class="dt">mc.samples =</span> <span class="dv">128</span></span>
<span id="cb101-10"><a href="log-ratio-methods-1.html#cb101-10"></a>        )</span>
<span id="cb101-11"><a href="log-ratio-methods-1.html#cb101-11"></a>    <span class="co"># Extract each clr MC draw</span></span>
<span id="cb101-12"><a href="log-ratio-methods-1.html#cb101-12"></a>    mc_draws &lt;-<span class="st"> </span><span class="kw">getMonteCarloInstances</span>(aldex.clr.res)</span>
<span id="cb101-13"><a href="log-ratio-methods-1.html#cb101-13"></a></span>
<span id="cb101-14"><a href="log-ratio-methods-1.html#cb101-14"></a>    <span class="co"># Average each draw for the &#39;normalized&#39; clr read table</span></span>
<span id="cb101-15"><a href="log-ratio-methods-1.html#cb101-15"></a>    dat.normed &lt;-</span>
<span id="cb101-16"><a href="log-ratio-methods-1.html#cb101-16"></a><span class="st">        </span><span class="kw">matrix</span>(<span class="dv">0</span>,</span>
<span id="cb101-17"><a href="log-ratio-methods-1.html#cb101-17"></a>               <span class="dt">nrow =</span> <span class="kw">nrow</span>(mc_draws[[<span class="dv">1</span>]]) ,</span>
<span id="cb101-18"><a href="log-ratio-methods-1.html#cb101-18"></a>               <span class="dt">ncol =</span> <span class="kw">length</span>(mc_draws))</span>
<span id="cb101-19"><a href="log-ratio-methods-1.html#cb101-19"></a>    <span class="cf">for</span> (col <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(dat.normed)) {</span>
<span id="cb101-20"><a href="log-ratio-methods-1.html#cb101-20"></a>        dat.normed[, col] &lt;-<span class="st"> </span><span class="kw">apply</span>(mc_draws[[col]], <span class="dv">1</span>, mean)</span>
<span id="cb101-21"><a href="log-ratio-methods-1.html#cb101-21"></a>    }</span>
<span id="cb101-22"><a href="log-ratio-methods-1.html#cb101-22"></a>    <span class="kw">rownames</span>(dat.normed) &lt;-<span class="st"> </span><span class="kw">rownames</span>(mc_draws[[<span class="dv">1</span>]])</span>
<span id="cb101-23"><a href="log-ratio-methods-1.html#cb101-23"></a>    <span class="kw">colnames</span>(dat.normed) &lt;-<span class="st"> </span><span class="kw">names</span>(mc_draws)</span>
<span id="cb101-24"><a href="log-ratio-methods-1.html#cb101-24"></a>    </span>
<span id="cb101-25"><a href="log-ratio-methods-1.html#cb101-25"></a>    otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(dat.normed, T)</span>
<span id="cb101-26"><a href="log-ratio-methods-1.html#cb101-26"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb101-27"><a href="log-ratio-methods-1.html#cb101-27"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb101-28"><a href="log-ratio-methods-1.html#cb101-28"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb101-29"><a href="log-ratio-methods-1.html#cb101-29"></a>    ps_aldex2 &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu, sam, tax, phy)</span>
<span id="cb101-30"><a href="log-ratio-methods-1.html#cb101-30"></a>    <span class="kw">return</span>(ps_aldex2)</span>
<span id="cb101-31"><a href="log-ratio-methods-1.html#cb101-31"></a>}</span>
<span id="cb101-32"><a href="log-ratio-methods-1.html#cb101-32"></a></span>
<span id="cb101-33"><a href="log-ratio-methods-1.html#cb101-33"></a>k_aldex2 &lt;-<span class="st"> </span><span class="kw">norm_ALDEx2</span>(k_raw, <span class="dt">group_var =</span> <span class="st">&quot;DIAGNOSIS&quot;</span>)</span></code></pre></div>
<pre><code>## operating in serial mode</code></pre>
<pre><code>## computing center with all features</code></pre>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references hanging-indent">
<div id="ref-fernandes2014">
<p>Fernandes, Andrew D, Jennifer NS Reid, Jean M Macklaim, Thomas A McMurrough, David R Edgell, and Gregory B Gloor. 2014. “Unifying the Analysis of High-Throughput Sequencing Datasets: Characterizing Rna-Seq, 16S rRNA Gene Sequencing and Selective Growth Experiments by Compositional Data Analysis.” <em>Microbiome</em> 2 (May): 15. <a href="https://doi.org/10.1186/2049-2618-2-15">https://doi.org/10.1186/2049-2618-2-15</a>.</p>
</div>
<div id="ref-kaul2017">
<p>Kaul, Abhishek, Siddhartha Mandal, Ori Davidov, and Shyamal D. Peddada. 2017. “Analysis of Microbiome Data in the Presence of Excess Zeros.” <em>Frontiers in Microbiology</em> 0. <a href="https://doi.org/10.3389/fmicb.2017.02114">https://doi.org/10.3389/fmicb.2017.02114</a>.</p>
</div>
<div id="ref-lin2020">
<p>Lin, Huang, and Shyamal Das Peddada. 2020. “Analysis of Compositions of Microbiomes with Bias Correction.” <em>Nature Communications</em> 11 (1): 3514. <a href="https://doi.org/10.1038/s41467-020-17041-7">https://doi.org/10.1038/s41467-020-17041-7</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="wrench.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="comparisions.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MB_normalization_review.pdf", "MB_normalization_review.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
