<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 TMM (edgeR) | Tutorial on Normalization of Microbiome Data</title>
  <meta name="description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 TMM (edgeR) | Tutorial on Normalization of Microbiome Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 TMM (edgeR) | Tutorial on Normalization of Microbiome Data" />
  
  <meta name="twitter:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="deseq.html"/>
<link rel="next" href="cumulative-sum-scaling-css.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">MB Normalization</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#the-importance-of-normalization"><i class="fa fa-check"></i><b>1.1</b> The importance of normalization</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-compositional-nature-of-microbiome-data"><i class="fa fa-check"></i><b>1.2</b> The compositional nature of microbiome data</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#log-ratio-transformation-methods"><i class="fa fa-check"></i><b>1.2.1</b> Log ratio transformation methods</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>2</b> Importing Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-data.html"><a href="importing-data.html#global-patterns"><i class="fa fa-check"></i><b>2.1</b> Global Patterns</a></li>
<li class="chapter" data-level="2.2" data-path="importing-data.html"><a href="importing-data.html#colorectal-cancer"><i class="fa fa-check"></i><b>2.2</b> Colorectal cancer</a></li>
<li class="chapter" data-level="2.3" data-path="importing-data.html"><a href="importing-data.html#pre-processing-quality-control-and-filtering"><i class="fa fa-check"></i><b>2.3</b> Pre-processing Quality Control and Filtering</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html"><i class="fa fa-check"></i><b>3</b> Total Sum scaling (TSS)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#about-tss"><i class="fa fa-check"></i><b>3.1</b> About TSS</a></li>
<li class="chapter" data-level="3.2" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-implementation"><i class="fa fa-check"></i><b>3.2</b> TSS Implementation</a></li>
<li class="chapter" data-level="3.3" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-on-global-patterns"><i class="fa fa-check"></i><b>3.3</b> TSS on Global Patterns</a></li>
<li class="chapter" data-level="3.4" data-path="total-sum-scaling-tss.html"><a href="total-sum-scaling-tss.html#tss-on-kostic-data"><i class="fa fa-check"></i><b>3.4</b> TSS on Kostic data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="rarefying.html"><a href="rarefying.html"><i class="fa fa-check"></i><b>4</b> Rarefying</a>
<ul>
<li class="chapter" data-level="4.1" data-path="rarefying.html"><a href="rarefying.html#about-rarefying"><i class="fa fa-check"></i><b>4.1</b> About Rarefying</a></li>
<li class="chapter" data-level="4.2" data-path="rarefying.html"><a href="rarefying.html#rarefying-implementation"><i class="fa fa-check"></i><b>4.2</b> Rarefying implementation</a></li>
<li class="chapter" data-level="4.3" data-path="rarefying.html"><a href="rarefying.html#rarefying-on-global-patterns"><i class="fa fa-check"></i><b>4.3</b> Rarefying on Global Patterns</a></li>
<li class="chapter" data-level="4.4" data-path="rarefying.html"><a href="rarefying.html#rarefying-on-kostic-data"><i class="fa fa-check"></i><b>4.4</b> Rarefying on Kostic Data</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="deseq.html"><a href="deseq.html"><i class="fa fa-check"></i><b>5</b> DESeq</a>
<ul>
<li class="chapter" data-level="5.1" data-path="deseq.html"><a href="deseq.html#about-deseq"><i class="fa fa-check"></i><b>5.1</b> About DESeq</a></li>
<li class="chapter" data-level="5.2" data-path="deseq.html"><a href="deseq.html#deseq-implementation"><i class="fa fa-check"></i><b>5.2</b> DESeq Implementation</a></li>
<li class="chapter" data-level="5.3" data-path="deseq.html"><a href="deseq.html#deseq-on-global-patterns"><i class="fa fa-check"></i><b>5.3</b> DESeq on Global Patterns</a></li>
<li class="chapter" data-level="5.4" data-path="deseq.html"><a href="deseq.html#deseq-on-kostic-data"><i class="fa fa-check"></i><b>5.4</b> DESeq on Kostic data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tmm-edger.html"><a href="tmm-edger.html"><i class="fa fa-check"></i><b>6</b> TMM (edgeR)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="tmm-edger.html"><a href="tmm-edger.html#edger-tmm-implementation"><i class="fa fa-check"></i><b>6.1</b> EdgeR TMM implementation</a></li>
<li class="chapter" data-level="6.2" data-path="tmm-edger.html"><a href="tmm-edger.html#tmm-on-global-patterns"><i class="fa fa-check"></i><b>6.2</b> TMM on Global Patterns</a></li>
<li class="chapter" data-level="6.3" data-path="tmm-edger.html"><a href="tmm-edger.html#tmm-on-kostic-data"><i class="fa fa-check"></i><b>6.3</b> TMM on Kostic data</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html"><i class="fa fa-check"></i><b>7</b> Cumulative sum scaling (CSS)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-implementation"><i class="fa fa-check"></i><b>7.1</b> CSS implementation</a></li>
<li class="chapter" data-level="7.2" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-on-global-patterns"><i class="fa fa-check"></i><b>7.2</b> CSS on Global Patterns</a></li>
<li class="chapter" data-level="7.3" data-path="cumulative-sum-scaling-css.html"><a href="cumulative-sum-scaling-css.html#css-on-kostic-data"><i class="fa fa-check"></i><b>7.3</b> CSS on Kostic data</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="gmpr.html"><a href="gmpr.html"><i class="fa fa-check"></i><b>8</b> GMPR</a>
<ul>
<li class="chapter" data-level="8.1" data-path="gmpr.html"><a href="gmpr.html#gmpr-implementation"><i class="fa fa-check"></i><b>8.1</b> GMPR Implementation</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="wrench.html"><a href="wrench.html"><i class="fa fa-check"></i><b>9</b> Wrench</a>
<ul>
<li class="chapter" data-level="9.1" data-path="wrench.html"><a href="wrench.html#wrench-implementation"><i class="fa fa-check"></i><b>9.1</b> Wrench Implementation</a></li>
<li class="chapter" data-level="9.2" data-path="wrench.html"><a href="wrench.html#wrench-on-global-patterns"><i class="fa fa-check"></i><b>9.2</b> Wrench on Global Patterns</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="log-ratio-methods.html"><a href="log-ratio-methods.html"><i class="fa fa-check"></i><b>10</b> Log Ratio Methods</a>
<ul>
<li class="chapter" data-level="10.1" data-path="log-ratio-methods.html"><a href="log-ratio-methods.html#ancom-ii"><i class="fa fa-check"></i><b>10.1</b> ANCOM II</a></li>
<li class="chapter" data-level="10.2" data-path="log-ratio-methods.html"><a href="log-ratio-methods.html#ancom-bc"><i class="fa fa-check"></i><b>10.2</b> ANCOM BC</a></li>
<li class="chapter" data-level="10.3" data-path="log-ratio-methods.html"><a href="log-ratio-methods.html#aldex2"><i class="fa fa-check"></i><b>10.3</b> ALDEx2</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="comparisions.html"><a href="comparisions.html"><i class="fa fa-check"></i><b>11</b> Comparisions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial on Normalization of Microbiome Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="tmm-edger" class="section level1" number="6">
<h1><span class="header-section-number">Chapter 6</span> TMM (edgeR)</h1>
<p>TMM (Trimmed median of m-values) is another method borrowed from RNA-Seq analysis, and implemented in <code>edgeR</code> <span class="citation">(Robinson and Oshlack <a href="#ref-robinson2010" role="doc-biblioref">2010</a>)</span>. This method uses, or calculates a reference sample, and compares all other samples to the reference sample. The size factor is the mean of the log-ratios after excluding the highest count taxa and taxa with the largest fold change. As taxa with zero counts are excluded, a pseudo count is needed. Additionally, there is the <code>TMMwsp</code> option which is encouraged as it is more robust to zero counts. Positive counts are reused to increase the number of features when we compared. The singleton positive counts are paired up in decreasing order of size and then a modified TMM method is applied to the re-ordered libraries.</p>
<div id="edger-tmm-implementation" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> EdgeR TMM implementation</h2>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="tmm-edger.html#cb47-1"></a>norm_TMM &lt;-<span class="st"> </span><span class="cf">function</span>(physeq, <span class="dt">group =</span> <span class="dv">1</span>, <span class="dt">method=</span><span class="st">&quot;TMM&quot;</span>, <span class="dt">pseudocount =</span> <span class="dv">1</span>, ...){</span>
<span id="cb47-2"><a href="tmm-edger.html#cb47-2"></a>    <span class="kw">require</span>(<span class="st">&quot;edgeR&quot;</span>, <span class="dt">quietly =</span> T)</span>
<span id="cb47-3"><a href="tmm-edger.html#cb47-3"></a>    <span class="kw">require</span>(<span class="st">&quot;phyloseq&quot;</span>, <span class="dt">quietly =</span> T)</span>
<span id="cb47-4"><a href="tmm-edger.html#cb47-4"></a>    <span class="co"># Enforce orientation.</span></span>
<span id="cb47-5"><a href="tmm-edger.html#cb47-5"></a>    <span class="cf">if</span>( <span class="op">!</span><span class="kw">taxa_are_rows</span>(physeq) ){ physeq &lt;-<span class="st"> </span><span class="kw">t</span>(physeq) }</span>
<span id="cb47-6"><a href="tmm-edger.html#cb47-6"></a>    x =<span class="st"> </span><span class="kw">as</span>(<span class="kw">otu_table</span>(physeq), <span class="st">&quot;matrix&quot;</span>)</span>
<span id="cb47-7"><a href="tmm-edger.html#cb47-7"></a>    <span class="co"># Add one to protect against overflow, log(0) issues.</span></span>
<span id="cb47-8"><a href="tmm-edger.html#cb47-8"></a>    x =<span class="st"> </span>x <span class="op">+</span><span class="st"> </span>pseudocount</span>
<span id="cb47-9"><a href="tmm-edger.html#cb47-9"></a>    <span class="co"># Check `group` argument</span></span>
<span id="cb47-10"><a href="tmm-edger.html#cb47-10"></a>    <span class="cf">if</span>( <span class="kw">identical</span>(<span class="kw">all.equal</span>(<span class="kw">length</span>(group), <span class="dv">1</span>), <span class="ot">TRUE</span>) <span class="op">&amp;</span><span class="st"> </span><span class="kw">nsamples</span>(physeq) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span> ){</span>
<span id="cb47-11"><a href="tmm-edger.html#cb47-11"></a>        <span class="co"># Assume that group was a sample variable name (must be categorical)</span></span>
<span id="cb47-12"><a href="tmm-edger.html#cb47-12"></a>        group =<span class="st"> </span><span class="kw">get_variable</span>(physeq, group)</span>
<span id="cb47-13"><a href="tmm-edger.html#cb47-13"></a>    }</span>
<span id="cb47-14"><a href="tmm-edger.html#cb47-14"></a>    <span class="co"># Define gene annotations (`genes`) as tax_table</span></span>
<span id="cb47-15"><a href="tmm-edger.html#cb47-15"></a>    taxonomy =<span class="st"> </span><span class="kw">tax_table</span>(physeq, <span class="dt">errorIfNULL=</span><span class="ot">FALSE</span>)</span>
<span id="cb47-16"><a href="tmm-edger.html#cb47-16"></a>    <span class="cf">if</span>( <span class="op">!</span><span class="kw">is.null</span>(taxonomy) ){</span>
<span id="cb47-17"><a href="tmm-edger.html#cb47-17"></a>        taxonomy =<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">as</span>(taxonomy, <span class="st">&quot;matrix&quot;</span>))</span>
<span id="cb47-18"><a href="tmm-edger.html#cb47-18"></a>    } </span>
<span id="cb47-19"><a href="tmm-edger.html#cb47-19"></a>    <span class="co"># Now turn into a DGEList</span></span>
<span id="cb47-20"><a href="tmm-edger.html#cb47-20"></a>    y =<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">counts=</span>x, <span class="dt">group=</span>group, <span class="dt">genes=</span>taxonomy, <span class="dt">remove.zeros =</span> <span class="ot">TRUE</span>)</span>
<span id="cb47-21"><a href="tmm-edger.html#cb47-21"></a>    <span class="co"># Calculate the normalization factors</span></span>
<span id="cb47-22"><a href="tmm-edger.html#cb47-22"></a>    d =<span class="st"> </span>edgeR<span class="op">::</span><span class="kw">calcNormFactors</span>(y, <span class="dt">method=</span>method)</span>
<span id="cb47-23"><a href="tmm-edger.html#cb47-23"></a>    <span class="co"># Check for division by zero inside `calcNormFactors`</span></span>
<span id="cb47-24"><a href="tmm-edger.html#cb47-24"></a>    <span class="cf">if</span>( <span class="op">!</span><span class="kw">all</span>(<span class="kw">is.finite</span>(d<span class="op">$</span>samples<span class="op">$</span>norm.factors)) ){</span>
<span id="cb47-25"><a href="tmm-edger.html#cb47-25"></a>        <span class="kw">stop</span>(<span class="st">&quot;Something wrong with edgeR::calcNormFactors on this data,</span></span>
<span id="cb47-26"><a href="tmm-edger.html#cb47-26"></a><span class="st">         non-finite $norm.factors, consider changing `method` argument&quot;</span>)</span>
<span id="cb47-27"><a href="tmm-edger.html#cb47-27"></a>    }</span>
<span id="cb47-28"><a href="tmm-edger.html#cb47-28"></a>    scalingFactor &lt;-<span class="st"> </span>d<span class="op">$</span>samples<span class="op">$</span>norm.factors <span class="op">*</span><span class="st"> </span>d<span class="op">$</span>samples<span class="op">$</span>lib.size <span class="op">/</span><span class="st"> </span><span class="fl">1e6</span></span>
<span id="cb47-29"><a href="tmm-edger.html#cb47-29"></a>    dataNormalized &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(<span class="kw">otu_table</span>(physeq)) <span class="op">/</span><span class="st"> </span>scalingFactor)</span>
<span id="cb47-30"><a href="tmm-edger.html#cb47-30"></a>    <span class="co">#dataNormalized &lt;- cpm(d)</span></span>
<span id="cb47-31"><a href="tmm-edger.html#cb47-31"></a>    otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(dataNormalized, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb47-32"><a href="tmm-edger.html#cb47-32"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(physeq, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb47-33"><a href="tmm-edger.html#cb47-33"></a>    sam<span class="op">$</span>scaling_factor &lt;-<span class="st"> </span>scalingFactor</span>
<span id="cb47-34"><a href="tmm-edger.html#cb47-34"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(physeq, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb47-35"><a href="tmm-edger.html#cb47-35"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(physeq, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb47-36"><a href="tmm-edger.html#cb47-36"></a>    ps_edgeR &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu,sam,tax,phy)</span>
<span id="cb47-37"><a href="tmm-edger.html#cb47-37"></a>    </span>
<span id="cb47-38"><a href="tmm-edger.html#cb47-38"></a>    <span class="kw">return</span>(ps_edgeR)</span>
<span id="cb47-39"><a href="tmm-edger.html#cb47-39"></a>    </span>
<span id="cb47-40"><a href="tmm-edger.html#cb47-40"></a>}</span></code></pre></div>
</div>
<div id="tmm-on-global-patterns" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> TMM on Global Patterns</h2>
<p>Perform normalization:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="tmm-edger.html#cb48-1"></a>gp_tmm &lt;-<span class="st"> </span><span class="kw">norm_TMM</span>(gp_raw)</span></code></pre></div>
<p>View PCoA plots</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="tmm-edger.html#cb49-1"></a><span class="kw">plot_ordination</span>(gp_tmm, phyloseq<span class="op">::</span><span class="kw">ordinate</span>(gp_tmm, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) , <span class="dt">color =</span> <span class="st">&quot;SampleType&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw data&quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>View how TMM normalization changes distance metrics differently than raw counts.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="tmm-edger.html#cb50-1"></a><span class="kw">plot_norm_changes</span>(gp_tmm, gp_raw,</span>
<span id="cb50-2"><a href="tmm-edger.html#cb50-2"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;Raw&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;TMM&quot;</span>,</span>
<span id="cb50-3"><a href="tmm-edger.html#cb50-3"></a>                  <span class="dt">title =</span> <span class="st">&quot;Distance metric comparision between TMM normalization and Raw counts &quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
</div>
<div id="tmm-on-kostic-data" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> TMM on Kostic data</h2>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="tmm-edger.html#cb51-1"></a>k_tmm &lt;-<span class="st"> </span><span class="kw">norm_TMM</span>(k_raw)</span>
<span id="cb51-2"><a href="tmm-edger.html#cb51-2"></a><span class="kw">plot_ordination</span>(</span>
<span id="cb51-3"><a href="tmm-edger.html#cb51-3"></a>    k_tmm,</span>
<span id="cb51-4"><a href="tmm-edger.html#cb51-4"></a>    phyloseq<span class="op">::</span><span class="kw">ordinate</span>(k_tmm, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) ,</span>
<span id="cb51-5"><a href="tmm-edger.html#cb51-5"></a>    <span class="dt">color =</span> <span class="st">&quot;DIAGNOSIS&quot;</span>,</span>
<span id="cb51-6"><a href="tmm-edger.html#cb51-6"></a>    <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw data&quot;</span></span>
<span id="cb51-7"><a href="tmm-edger.html#cb51-7"></a>) <span class="op">/</span></span>
<span id="cb51-8"><a href="tmm-edger.html#cb51-8"></a><span class="st">    </span><span class="kw">plot_ordination</span>(</span>
<span id="cb51-9"><a href="tmm-edger.html#cb51-9"></a>        k_tmm,</span>
<span id="cb51-10"><a href="tmm-edger.html#cb51-10"></a>        phyloseq<span class="op">::</span><span class="kw">ordinate</span>(k_tmm, <span class="st">&quot;PCoA&quot;</span>, <span class="st">&quot;bray&quot;</span>) ,</span>
<span id="cb51-11"><a href="tmm-edger.html#cb51-11"></a>        <span class="dt">color =</span> <span class="st">&quot;depth&quot;</span>,</span>
<span id="cb51-12"><a href="tmm-edger.html#cb51-12"></a>        <span class="dt">title =</span> <span class="st">&quot;PCoA on Raw data&quot;</span></span>
<span id="cb51-13"><a href="tmm-edger.html#cb51-13"></a>    )</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/edgeR-kostic-1.png" width="672" /></p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="tmm-edger.html#cb52-1"></a><span class="kw">plot_norm_changes</span>(k_tmm, k_raw,</span>
<span id="cb52-2"><a href="tmm-edger.html#cb52-2"></a>                  <span class="dt">x_lab =</span> <span class="st">&quot;Raw&quot;</span>, <span class="dt">y_lab =</span> <span class="st">&quot;TMM&quot;</span>,</span>
<span id="cb52-3"><a href="tmm-edger.html#cb52-3"></a>                  <span class="dt">title =</span> <span class="st">&quot;Distance metric comparision between TMM normalization and Raw counts &quot;</span>)</span></code></pre></div>
<p><img src="MB_normalization_review_files/figure-html/edgeR-kostic-2.png" width="672" /></p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references hanging-indent">
<div id="ref-robinson2010">
<p>Robinson, Mark D., and Alicia Oshlack. 2010. “A Scaling Normalization Method for Differential Expression Analysis of Rna-Seq Data.” <em>Genome Biology</em> 11 (3): R25. <a href="https://doi.org/10.1186/gb-2010-11-3-r25">https://doi.org/10.1186/gb-2010-11-3-r25</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="deseq.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="cumulative-sum-scaling-css.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MB_normalization_review.pdf", "MB_normalization_review.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
