<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Wrench | Tutorial on Normalization of Microbiome Data</title>
  <meta name="description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Wrench | Tutorial on Normalization of Microbiome Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Wrench | Tutorial on Normalization of Microbiome Data" />
  
  <meta name="twitter:description" content="A tutorial on normalization methods for microbiome data and discussion of thier uses and performance" />
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="library-size-and-composition-scaling-methods.html"/>
<link rel="next" href="log-ratio-transformation-methods.html"/>
<script src="libs/header-attrs-2.9/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">MB Normalization</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#the-importance-of-normalization"><i class="fa fa-check"></i><b>1.1</b> The importance of normalization</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#the-compositional-nature-of-microbiome-data"><i class="fa fa-check"></i><b>1.2</b> The compositional nature of microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="importing-data.html"><a href="importing-data.html"><i class="fa fa-check"></i><b>2</b> Importing Data</a>
<ul>
<li class="chapter" data-level="2.1" data-path="importing-data.html"><a href="importing-data.html#global-patterns"><i class="fa fa-check"></i><b>2.1</b> Global Patterns</a></li>
<li class="chapter" data-level="2.2" data-path="importing-data.html"><a href="importing-data.html#colorectal-cancer"><i class="fa fa-check"></i><b>2.2</b> Colorectal cancer</a></li>
<li class="chapter" data-level="2.3" data-path="importing-data.html"><a href="importing-data.html#pre-processing-quality-control-and-filtering"><i class="fa fa-check"></i><b>2.3</b> Pre-processing Quality Control and Filtering</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html"><i class="fa fa-check"></i><b>3</b> Count Scaling Normalization methods</a>
<ul>
<li class="chapter" data-level="3.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#total-sum-scaling-tss"><i class="fa fa-check"></i><b>3.1</b> Total Sum scaling (TSS)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#about-tss"><i class="fa fa-check"></i><b>3.1.1</b> About TSS</a></li>
<li class="chapter" data-level="3.1.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#tss-r-code"><i class="fa fa-check"></i><b>3.1.2</b> TSS R code</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#cumulative-sum-scaling-css"><i class="fa fa-check"></i><b>3.2</b> Cumulative sum scaling (CSS)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#about-css"><i class="fa fa-check"></i><b>3.2.1</b> About CSS</a></li>
<li class="chapter" data-level="3.2.2" data-path="count-scaling-normalization-methods.html"><a href="count-scaling-normalization-methods.html#css-r-code"><i class="fa fa-check"></i><b>3.2.2</b> CSS R code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="subsampling-methods.html"><a href="subsampling-methods.html"><i class="fa fa-check"></i><b>4</b> Subsampling Methods</a>
<ul>
<li class="chapter" data-level="4.1" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying"><i class="fa fa-check"></i><b>4.1</b> Rarefying</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="subsampling-methods.html"><a href="subsampling-methods.html#about-rarefying"><i class="fa fa-check"></i><b>4.1.1</b> About Rarefying</a></li>
<li class="chapter" data-level="4.1.2" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-walkthrough"><i class="fa fa-check"></i><b>4.1.2</b> Rarefying walkthrough</a></li>
<li class="chapter" data-level="4.1.3" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-r-code"><i class="fa fa-check"></i><b>4.1.3</b> Rarefying R code</a></li>
<li class="chapter" data-level="4.1.4" data-path="subsampling-methods.html"><a href="subsampling-methods.html#rarefying-on-kostic-data"><i class="fa fa-check"></i><b>4.1.4</b> Rarefying on Kostic Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html"><i class="fa fa-check"></i><b>5</b> Library Size and Composition Scaling Methods</a>
<ul>
<li class="chapter" data-level="5.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#deseq"><i class="fa fa-check"></i><b>5.1</b> DESeq</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-deseq"><i class="fa fa-check"></i><b>5.1.1</b> About DESeq</a></li>
<li class="chapter" data-level="5.1.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#deseq-r-code"><i class="fa fa-check"></i><b>5.1.2</b> DESeq R Code</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr"><i class="fa fa-check"></i><b>5.2</b> GMPR</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-gmpr"><i class="fa fa-check"></i><b>5.2.1</b> About GMPR</a></li>
<li class="chapter" data-level="5.2.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr-walkthough"><i class="fa fa-check"></i><b>5.2.2</b> GMPR Walkthough</a></li>
<li class="chapter" data-level="5.2.3" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#gmpr-r-code"><i class="fa fa-check"></i><b>5.2.3</b> GMPR R Code</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#tmm-edger"><i class="fa fa-check"></i><b>5.3</b> TMM (edgeR)</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#about-tmm"><i class="fa fa-check"></i><b>5.3.1</b> About TMM</a></li>
<li class="chapter" data-level="5.3.2" data-path="library-size-and-composition-scaling-methods.html"><a href="library-size-and-composition-scaling-methods.html#edger-tmm-t-code"><i class="fa fa-check"></i><b>5.3.2</b> EdgeR TMM T code</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="wrench.html"><a href="wrench.html"><i class="fa fa-check"></i><b>6</b> Wrench</a></li>
<li class="chapter" data-level="7" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html"><i class="fa fa-check"></i><b>7</b> Log Ratio Transformation Methods</a>
<ul>
<li class="chapter" data-level="7.1" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#ancom-ii"><i class="fa fa-check"></i><b>7.1</b> ANCOM II</a></li>
<li class="chapter" data-level="7.2" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#ancom-bc"><i class="fa fa-check"></i><b>7.2</b> ANCOM BC</a></li>
<li class="chapter" data-level="7.3" data-path="log-ratio-transformation-methods.html"><a href="log-ratio-transformation-methods.html#aldex2"><i class="fa fa-check"></i><b>7.3</b> ALDEx2</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="comparisions.html"><a href="comparisions.html"><i class="fa fa-check"></i><b>8</b> Comparisions</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial on Normalization of Microbiome Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="wrench" class="section level1" number="6">
<h1><span class="header-section-number">Chapter 6</span> Wrench</h1>
<p>Wrench is a recent normalization method developed for microbiome data <span class="citation">(Kumar et al. <a href="#ref-kumar2018" role="doc-biblioref">2018</a>)</span> that uses an empirical Bayes Normalization approach. This method includes compositional bias correction for sparse datasets. This method uses a hurdle log-normal distribution to estimate the normalization factors (the location estimate for the group). For this method, we assume abundances are drawn from a hurdle Log-Gaussian distribution, and the scaling factor used is essentially the location estimate for the group.</p>
<!-- When a differential abundance analysis is performed on this data, fold changes of null features, those not differentially abundant in the absolute scale, are intimately tied to those of features that are perturbed in their absolute abundances, making the former appear differentially abundant.                                           -->
<!-- "In the absence of such control features, effective correc- tion for compositional bias can still be hoped for, as it can be shown that this correction amounts to resolv- ing a linear technical bias [13]. This fact allows one to exploit several widely used non- spike-in normalization approaches [13, 15–17], which approximate the afore- mentioned spike-in strategy by assuming that most fea- tures do not change on average across samples/conditions. For the same reason, such an interpretation can also be given to approaches like centered logarithmic transforms (CLR) from the theory of compositional data, which many analysts favor when working with relative abundances [18–24]. In this paper, we analyze the behavior of these existing scaling normalization techniques in light of com- positional bias. This method develops a compositional bias correction technique for sparse count data based on an empirical Bayes approach that borrows information across features and samples. The scaling factor is the average of ratios of relative abundances across taxa." -->
<!-- ## Wrench Implementation -->
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="wrench.html#cb70-1"></a>norm_wrench &lt;-<span class="st"> </span><span class="cf">function</span>(ps, group_col){</span>
<span id="cb70-2"><a href="wrench.html#cb70-2"></a>    <span class="kw">require</span>(Wrench, <span class="dt">quietly =</span> T)</span>
<span id="cb70-3"><a href="wrench.html#cb70-3"></a>    <span class="cf">if</span>( <span class="kw">identical</span>(<span class="kw">all.equal</span>(<span class="kw">length</span>(group_col), <span class="dv">1</span>), <span class="ot">TRUE</span>) <span class="op">&amp;</span><span class="st"> </span><span class="kw">nsamples</span>(ps) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span> ){</span>
<span id="cb70-4"><a href="wrench.html#cb70-4"></a>        <span class="co"># Assume that group was a sample variable name (must be categorical)</span></span>
<span id="cb70-5"><a href="wrench.html#cb70-5"></a>        group =<span class="st"> </span><span class="kw">get_variable</span>(ps, group_col)</span>
<span id="cb70-6"><a href="wrench.html#cb70-6"></a>    }</span>
<span id="cb70-7"><a href="wrench.html#cb70-7"></a>    otu_tab &lt;-<span class="st"> </span><span class="kw">otu_table</span>(ps)</span>
<span id="cb70-8"><a href="wrench.html#cb70-8"></a>    W &lt;-<span class="st"> </span><span class="kw">wrench</span>(otu_tab, group)</span>
<span id="cb70-9"><a href="wrench.html#cb70-9"></a>    </span>
<span id="cb70-10"><a href="wrench.html#cb70-10"></a>    compositionalFactors &lt;-<span class="st"> </span>W<span class="op">$</span>ccf</span>
<span id="cb70-11"><a href="wrench.html#cb70-11"></a>    normalizationFactors &lt;-<span class="st"> </span>W<span class="op">$</span>nf</span>
<span id="cb70-12"><a href="wrench.html#cb70-12"></a>    </span>
<span id="cb70-13"><a href="wrench.html#cb70-13"></a>    normed_otu &lt;-<span class="st"> </span>otu_tab<span class="op">/</span>(normalizationFactors<span class="op">/</span><span class="fl">1e6</span>)</span>
<span id="cb70-14"><a href="wrench.html#cb70-14"></a>    otu &lt;-<span class="st"> </span><span class="kw">otu_table</span>(normed_otu, <span class="dt">taxa_are_rows =</span> T)</span>
<span id="cb70-15"><a href="wrench.html#cb70-15"></a>    sam &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;sam_data&quot;</span>)</span>
<span id="cb70-16"><a href="wrench.html#cb70-16"></a>    sam<span class="op">$</span>scaling_factor &lt;-<span class="st"> </span>normalizationFactors</span>
<span id="cb70-17"><a href="wrench.html#cb70-17"></a>    tax &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;tax_table&quot;</span>)</span>
<span id="cb70-18"><a href="wrench.html#cb70-18"></a>    phy &lt;-<span class="st"> </span><span class="kw">access</span>(ps, <span class="st">&quot;phy_tree&quot;</span>)</span>
<span id="cb70-19"><a href="wrench.html#cb70-19"></a>    ps_wrench &lt;-<span class="st"> </span><span class="kw">phyloseq</span>(otu,sam,tax,phy)</span>
<span id="cb70-20"><a href="wrench.html#cb70-20"></a>    </span>
<span id="cb70-21"><a href="wrench.html#cb70-21"></a>    <span class="kw">return</span>(ps_wrench)</span>
<span id="cb70-22"><a href="wrench.html#cb70-22"></a>}</span></code></pre></div>
<!-- ## Wrench on Global Patterns -->
<!-- ```{r, eval = F} -->
<!-- gp_wrench <- norm_wrench(gp_raw, group_col = "SampleType") -->
<!-- ``` -->
<!-- ```{r, message = F, results = "hide", eval = F, echo = F} -->
<!-- gp_wrench <- norm_wrench(gp_raw, group_col = "SampleType") -->
<!-- # warnings, authors suggest robust averaging  -->
<!-- # Is this because we have such distinct groups?  -->
<!-- ``` -->
<!-- # ```{r} -->
<!-- # norm_ms_wrench <- function(ps, group_col){ -->
<!-- #   require(metagenomeSeq, quietly = T) -->
<!-- #   browser() -->
<!-- #   ps.metaG <- phyloseq_to_metagenomeSeq(ps) -->
<!-- #   p_stat = cumNormStatFast(ps.metaG) -->
<!-- #   ps.metaG = cumNorm(ps.metaG, p = p_stat) -->
<!-- #    -->
<!-- #   if( identical(all.equal(length(group_col), 1), TRUE) & nsamples(ps) > 1 ){ -->
<!-- #         # Assume that group was a sample variable name (must be categorical) -->
<!-- #         group = get_variable(ps, group_col) -->
<!-- #     } -->
<!-- #    -->
<!-- #   res_wrench <- wrenchNorm(ps.metaG, condition = group) -->
<!-- #    -->
<!-- #   ps.metaG.norm <- MRcounts(res_wrench, norm = T) -->
<!-- #  -->
<!-- #   otu <- otu_table(ps.metaG.norm, taxa_are_rows = T) -->
<!-- #   sam <- access(ps, "sam_data") -->
<!-- #   sam$scaling_factor <- normFactors(ps.metaG)/1e6 -->
<!-- #   tax <- access(ps, "tax_table") -->
<!-- #   phy <- access(ps, "phy_tree") -->
<!-- #   ps_CSS <- phyloseq(otu,sam,tax,phy) -->
<!-- #   return(ps_CSS) -->
<!-- # } -->
<!-- # ``` -->

</div>
<h3>References</h3>
<div id="refs" class="references hanging-indent">
<div id="ref-kumar2018">
<p>Kumar, M. Senthil, Eric V. Slud, Kwame Okrah, Stephanie C. Hicks, Sridhar Hannenhalli, and Héctor Corrada Bravo. 2018. “Analysis and Correction of Compositional Bias in Sparse Sequencing Count Data.” <em>BMC Genomics</em> 19 (1): 799. <a href="https://doi.org/10.1186/s12864-018-5160-5">https://doi.org/10.1186/s12864-018-5160-5</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="library-size-and-composition-scaling-methods.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="log-ratio-transformation-methods.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["MB_normalization_review.pdf", "MB_normalization_review.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
